// @cloudbase/functions/dist/cjs/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerFunctions = void 0;
var utilities_1 = require("@cloudbase/utilities");
var cloudrun_1 = require("@cloudbase/cloudrun");
var getSdkName = utilities_1.constants.getSdkName, ERRORS = utilities_1.constants.ERRORS, COMMUNITY_SITE_URL = utilities_1.constants.COMMUNITY_SITE_URL;
var execCallback = utilities_1.utils.execCallback;
var catchErrorsDecorator = utilities_1.helpers.catchErrorsDecorator;
var COMPONENT_NAME = 'functions';
function isWedaFunc(functionName) {
    return (functionName === 'lcap-common-service'
        || functionName === 'lcap-datasource-test'
        || functionName.startsWith('lcap-data')
        || functionName.startsWith('lowcode-datasource')
        || functionName === 'cloudbase_module');
}
function isCloudBackendFunc(functionName) {
    return functionName === 'lowcode-wx-datasource';
}
var CloudbaseFunctions = (function () {
    function CloudbaseFunctions() {
    }
    CloudbaseFunctions.prototype.callFunction = function (options, callback, customReqOpts) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function () {
            var name, data, query, parse, search, type, _d, path, method, _e, header, jsonData, requestId, result, _f, request, config, action, params, res, toGateWay, response, result, e_1;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        name = options.name, data = options.data, query = options.query, parse = options.parse, search = options.search, type = options.type, _d = options.path, path = _d === void 0 ? '' : _d, method = options.method, _e = options.header, header = _e === void 0 ? {} : _e;
                        if (!name) {
                            throw new Error(JSON.stringify({
                                code: ERRORS.INVALID_PARAMS,
                                msg: "[".concat(COMPONENT_NAME, ".callFunction] invalid function name"),
                            }));
                        }
                        try {
                            jsonData = data ? JSON.stringify(data) : '';
                        }
                        catch (e) {
                            throw new Error(JSON.stringify({
                                code: ERRORS.INVALID_PARAMS,
                                msg: "[".concat(COMPONENT_NAME, ".callFunction] invalid data"),
                            }));
                        }
                        if (!(type === 'cloudrun')) return [3, 2];
                        requestId = utilities_1.utils.generateRequestId();
                        return [4, cloudrun_1.requestContainer.call(this, {
                                name: name,
                                data: jsonData,
                                path: path,
                                method: method,
                                header: __assign(__assign({}, header), { 'X-Request-Id': requestId }),
                            })];
                    case 1:
                        result = _g.sent();
                        return [2, {
                                result: result,
                                requestId: requestId,
                            }];
                    case 2:
                        _f = this, request = _f.request, config = _f.config;
                        action = 'functions.invokeFunction';
                        params = {
                            inQuery: query,
                            parse: parse,
                            search: search,
                            function_name: name,
                            request_data: jsonData,
                        };
                        res = {};
                        toGateWay = config.endPointMode === 'GATEWAY' && !isWedaFunc(name) && !isCloudBackendFunc(name);
                        _g.label = 3;
                    case 3:
                        _g.trys.push([3, 8, , 9]);
                        if (!toGateWay) return [3, 5];
                        return [4, request.gateWay(__assign(__assign({}, options), { path: 'functions' }), customReqOpts)];
                    case 4:
                        response = _g.sent();
                        res = __assign({ requestId: response.requestId }, (((_a = response.data) === null || _a === void 0 ? void 0 : _a.code)
                            ? response.data
                            : {
                                data: {
                                    response_data: response.data,
                                },
                            }));
                        return [3, 7];
                    case 5: return [4, request.send(action, params, {
                            endPointMode: 'CLOUD_API',
                            defaultQuery: ((_b = data === null || data === void 0 ? void 0 : data.params) === null || _b === void 0 ? void 0 : _b.action) ? { action: (_c = data === null || data === void 0 ? void 0 : data.params) === null || _c === void 0 ? void 0 : _c.action } : {},
                        }, customReqOpts)];
                    case 6:
                        res = _g.sent();
                        _g.label = 7;
                    case 7:
                        if (res.code) {
                            return [2, execCallback(callback, null, res)];
                        }
                        result = res.data.response_data;
                        if (parse || toGateWay) {
                            return [2, execCallback(callback, null, {
                                    result: result,
                                    requestId: res.requestId,
                                })];
                        }
                        try {
                            result = JSON.parse(res.data.response_data);
                            return [2, execCallback(callback, null, {
                                    result: result,
                                    requestId: res.requestId,
                                })];
                        }
                        catch (e) {
                            execCallback(callback, new Error("[".concat(getSdkName(), "][").concat(ERRORS.INVALID_PARAMS, "][").concat(COMPONENT_NAME, ".callFunction] response data must be json")));
                        }
                        return [3, 9];
                    case 8:
                        e_1 = _g.sent();
                        execCallback(callback, e_1);
                        return [3, 9];
                    case 9: return [2];
                }
            });
        });
    };
    __decorate([
        catchErrorsDecorator({
            customInfo: {
                className: 'Cloudbase',
                methodName: 'callFunction',
            },
            title: '函数调用失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 callFunction() 的语法或参数是否正确',
                '  2 - 当前环境下是否存在此函数',
                '  3 - 函数安全规则是否限制了当前登录状态访问',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Function, Object]),
        __metadata("design:returntype", Promise)
    ], CloudbaseFunctions.prototype, "callFunction", null);
    return CloudbaseFunctions;
}());
var cloudbaseFunctions = new CloudbaseFunctions();
var component = {
    name: COMPONENT_NAME,
    entity: {
        callFunction: cloudbaseFunctions.callFunction,
    },
};
try {
    cloudbase.registerComponent(component);
}
catch (e) { }
function registerFunctions(app) {
    try {
        app.registerComponent(component);
    }
    catch (e) {
        console.warn(e);
    }
}
exports.registerFunctions = registerFunctions;
