// @cloudbase/oauth/dist/cjs/utils/urlSearchParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MyURLSearchParams = (function () {
    function MyURLSearchParams(init) {
        this.params = {};
        if (typeof init === 'string') {
            this.parse(init);
        }
        else if (init && typeof init === 'object') {
            for (var key in init) {
                if (Object.prototype.hasOwnProperty.call(init, key)) {
                    this.append(key, init[key]);
                }
            }
        }
    }
    MyURLSearchParams.prototype.parse = function (str) {
        var _this = this;
        var params = str.split('&');
        params.forEach(function (param) {
            var _a = param.split('=').map(decodeURIComponent), key = _a[0], value = _a[1];
            _this.append(key, value);
        });
    };
    MyURLSearchParams.prototype.append = function (key, value) {
        if (this.params[key]) {
            this.params[key] = this.params[key].concat([value]);
        }
        else {
            this.params[key] = [value];
        }
    };
    MyURLSearchParams.prototype.get = function (key) {
        return this.params[key] ? this.params[key][0] : null;
    };
    MyURLSearchParams.prototype.getAll = function (key) {
        return this.params[key] || [];
    };
    MyURLSearchParams.prototype.delete = function (key) {
        delete this.params[key];
    };
    MyURLSearchParams.prototype.has = function (key) {
        return Object.prototype.hasOwnProperty.call(this.params, key);
    };
    MyURLSearchParams.prototype.set = function (key, value) {
        this.params[key] = [value];
    };
    MyURLSearchParams.prototype.toString = function () {
        var items = [];
        var _loop_1 = function (key) {
            if (Object.prototype.hasOwnProperty.call(this_1.params, key)) {
                this_1.params[key].forEach(function (value) {
                    items.push("".concat(encodeURIComponent(key), "=").concat(encodeURIComponent(value)));
                });
            }
        };
        var this_1 = this;
        for (var key in this.params) {
            _loop_1(key);
        }
        return items.join('&');
    };
    return MyURLSearchParams;
}());
exports.default = MyURLSearchParams;
