// @cloudbase/oauth/dist/cjs/auth/auth-error.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthError = exports.AuthErrorCategory = void 0;
var consts_1 = require("./consts");
var AuthErrorCategory;
(function (AuthErrorCategory) {
    AuthErrorCategory["PROVIDER_NOT_ENABLED"] = "PROVIDER_NOT_ENABLED";
    AuthErrorCategory["INVALID_CREDENTIALS"] = "INVALID_CREDENTIALS";
    AuthErrorCategory["USER_NOT_FOUND"] = "USER_NOT_FOUND";
    AuthErrorCategory["USER_STATUS_ABNORMAL"] = "USER_STATUS_ABNORMAL";
    AuthErrorCategory["SERVICE_ERROR"] = "SERVICE_ERROR";
    AuthErrorCategory["INVALID_PARAMS"] = "INVALID_PARAMS";
    AuthErrorCategory["AUTH_METHOD_MISMATCH"] = "AUTH_METHOD_MISMATCH";
    AuthErrorCategory["UNKNOWN"] = "UNKNOWN";
})(AuthErrorCategory = exports.AuthErrorCategory || (exports.AuthErrorCategory = {}));
var CONSOLE_AUTH_URL = 'https://tcb.cloud.tencent.com/dev?envId=#/identity/login-manage';
var DOCS_URL = 'https://docs.cloudbase.net/api-reference/webv3-next/initialization';
function inferErrorCategory(message, status) {
    if (!message && !status)
        return AuthErrorCategory.UNKNOWN;
    try {
        var msg = (message || '').toLowerCase();
        if (status === consts_1.ErrorType.FAILED_PRECONDITION
            || status === consts_1.ErrorType.PROVIDER_NOT_ENABLED
            || status === consts_1.ErrorType.LOGIN_METHOD_DISABLED
            || msg.includes('身份源')
            || msg.includes('开启')
            || msg.includes('identity provider')
            || msg.includes('provider not enabled')
            || msg.includes('login method')
            || msg.includes('登录方式')
            || msg.includes('not enabled')
            || msg.includes('未开启')
            || msg.includes('未启用')
            || msg.includes('请联系开发者')
            || msg.includes('anonymous login is not enabled')
            || msg.includes('匿名登录未开启')) {
            return AuthErrorCategory.PROVIDER_NOT_ENABLED;
        }
        if (status === consts_1.ErrorType.INVALID_PASSWORD
            || status === consts_1.ErrorType.INVALID_GRANT
            || status === consts_1.ErrorType.INVALID_VERIFICATION_CODE
            || status === consts_1.ErrorType.INVALID_USERNAME_OR_PASSWORD
            || status === consts_1.ErrorType.INVALID_CREDENTIALS
            || status === consts_1.ErrorType.WRONG_PASSWORD
            || status === consts_1.ErrorType.UNAUTHENTICATED
            || status === consts_1.ErrorType.INVALID_TWO_FACTOR
            || status === consts_1.ErrorType.INVALID_TWO_FACTOR_RECOVERY
            || status === consts_1.ErrorType.PASSWORD_NOT_SET
            || msg.includes('密码不正确')
            || msg.includes('password')
            || msg.includes('invalid_password')
            || msg.includes('凭据')
            || msg.includes('credentials')
            || msg.includes('验证码')
            || msg.includes('verification')
            || msg.includes('invalid_verification_code')
            || msg.includes('invalid_username_or_password')
            || msg.includes('username or password')
            || msg.includes('用户名或密码')
            || msg.includes('wrong password')
            || msg.includes('incorrect password')) {
            return AuthErrorCategory.INVALID_CREDENTIALS;
        }
        if (status === consts_1.ErrorType.NOT_FOUND
            || status === consts_1.ErrorType.USER_NOT_FOUND
            || msg.includes('用户不存在')
            || msg.includes('not found')
            || msg.includes('user not found')
            || msg.includes('不存在')
            || msg.includes('no such user')
            || msg.includes('user does not exist')) {
            return AuthErrorCategory.USER_NOT_FOUND;
        }
        if (status === consts_1.ErrorType.USER_BLOCKED
            || status === consts_1.ErrorType.USER_PENDING
            || status === consts_1.ErrorType.UNDER_REVIEW
            || status === consts_1.ErrorType.INVALID_STATUS
            || msg.includes('被封禁')
            || msg.includes('blocked')
            || msg.includes('pending')
            || msg.includes('审核')
            || msg.includes('review')
            || msg.includes('locked')
            || msg.includes('已锁定')) {
            return AuthErrorCategory.USER_STATUS_ABNORMAL;
        }
        if (status === consts_1.ErrorType.SERVER_ERROR
            || status === consts_1.ErrorType.UNAVAILABLE
            || status === consts_1.ErrorType.TEMPORARILY_UNAVAILABLE
            || status === consts_1.ErrorType.INTERNAL
            || status === consts_1.ErrorType.UNREACHABLE
            || status === consts_1.ErrorType.DEADLINE_EXCEEDED
            || status === consts_1.ErrorType.DATA_LOSS
            || msg.includes('network')
            || msg.includes('timeout')
            || msg.includes('服务')
            || msg.includes('timed out')
            || msg.includes('超时')) {
            return AuthErrorCategory.SERVICE_ERROR;
        }
        if (status === consts_1.ErrorType.INVALID_ARGUMENT
            || status === consts_1.ErrorType.INVALID_REQUEST
            || status === consts_1.ErrorType.INVALID_SCOPE
            || status === consts_1.ErrorType.INVALID_REQUEST_URI
            || status === consts_1.ErrorType.INVALID_REQUEST_OBJECT) {
            return AuthErrorCategory.INVALID_PARAMS;
        }
        if (status === consts_1.ErrorType.PERMISSION_DENIED
            || status === consts_1.ErrorType.ACCESS_DENIED
            || status === consts_1.ErrorType.UNAUTHORIZED_CLIENT) {
            return AuthErrorCategory.PROVIDER_NOT_ENABLED;
        }
        return AuthErrorCategory.UNKNOWN;
    }
    catch (error) {
        return status;
    }
}
function generateHelpMessage(category, context) {
    var methodName = (context === null || context === void 0 ? void 0 : context.method) || 'signIn';
    var methodToProvider = {
        signInWithPassword: '「用户名密码登录」',
        signInAnonymously: '「匿名登录」',
        signInWithOtp: '「短信验证码登录」或「邮箱验证码登录」',
        signUp: '「短信验证码登录」或「邮箱验证码登录」',
        signInWithOAuth: '对应的第三方身份源',
        signInWithIdToken: '对应的第三方身份源',
        signInWithOpenId: '「微信小程序 OpenID 登录」',
        signInWithPhoneAuth: '「微信小程序手机号授权登录」',
        signInWithCustomTicket: '「自定义登录」',
    };
    var requiredProvider = methodToProvider[methodName] || '对应的登录方式';
    switch (category) {
        case AuthErrorCategory.PROVIDER_NOT_ENABLED:
            return ('[CloudBase Auth] 登录方式未开启\n'
                + '\n'
                + "  \u5F53\u524D\u8C03\u7528\u7684 ".concat(methodName, "() \u6240\u9700\u7684\u767B\u5F55\u65B9\u5F0F\u5C1A\u672A\u5728\u4E91\u5F00\u53D1\u63A7\u5236\u53F0\u542F\u7528\u3002\n")
                + "  \u9700\u8981\u5F00\u542F\u7684\u767B\u5F55\u65B9\u5F0F\uFF1A".concat(requiredProvider, "\n")
                + '\n'
                + '  请按以下步骤开启：\n'
                + "  1. \u6253\u5F00\u4E91\u5F00\u53D1\u63A7\u5236\u53F0\uFF1A".concat(CONSOLE_AUTH_URL, "\n")
                + '  2. 选择对应的云开发环境\n'
                + '  3. 进入「登录授权」页面，在「登录方式」或「身份源列表」中开启所需的登录方式\n'
                + '\n'
                + '  各登录方法与所需登录方式的对应关系：\n'
                + '     - signInWithPassword → 「用户名密码登录」\n'
                + '     - signInAnonymously → 「匿名登录」\n'
                + '     - signInWithOtp/signUp → 「短信验证码登录」或「邮箱验证码登录」\n'
                + '     - signInWithOAuth/signInWithIdToken → 对应的第三方身份源\n'
                + '\n'
                + '  如使用 CLI/MCP 工具，可通过 auth 工具查询和配置身份源状态。\n'
                + "  \u66F4\u591A\u5E2E\u52A9\u8BF7\u53C2\u8003\u6587\u6863\uFF1A".concat(DOCS_URL));
        case AuthErrorCategory.INVALID_CREDENTIALS:
            return ('[CloudBase Auth] 凭据验证失败\n'
                + '\n'
                + "  \u8C03\u7528 ".concat(methodName, "() \u65F6\u51ED\u636E\u6821\u9A8C\u672A\u901A\u8FC7\u3002\n")
                + '\n'
                + '  请按优先级逐项排查：\n'
                + '  1. 确认用户名/邮箱/手机号是否拼写正确、格式完整\n'
                + '  2. 确认密码是否正确（注意大小写、特殊字符、前后空格）\n'
                + '  3. 确认用户账号确实存在（可通过 auth.isUsernameRegistered() 检查）\n'
                + '  4. 确认凭据参数来源可靠（环境变量、配置文件中的值是否已正确设置）\n'
                + '  5. 如使用加密登录（is_encrypt: true），请确认公钥版本正确\n'
                + '\n'
                + '  如忘记密码，可通过 auth.resetPasswordForEmail() 重置密码。\n'
                + '\n'
                + '  提示：如果是 AI agent / 自动化测试场景，请确认 ENV_ID、USERNAME、PASSWORD\n'
                + '  等环境变量已正确传入，且对应用户已在目标环境中注册。');
        case AuthErrorCategory.USER_NOT_FOUND:
            return ('[CloudBase Auth] 用户不存在\n'
                + '\n'
                + '  请检查以下项目：\n'
                + '  1. 确认用户名/邮箱/手机号是否正确\n'
                + '  2. 确认是在正确的云开发环境（envId）下操作\n'
                + '  3. 如果是新用户，请先通过 auth.signUp() 注册\n'
                + "  4. \u4E5F\u53EF\u5728\u4E91\u5F00\u53D1\u63A7\u5236\u53F0\u624B\u52A8\u521B\u5EFA\u7528\u6237\uFF1A".concat(CONSOLE_AUTH_URL, "\n")
                + '\n'
                + '  提示：不同环境（开发/测试/生产）的用户库是隔离的，请确认 envId 正确。');
        case AuthErrorCategory.USER_STATUS_ABNORMAL:
            return ('[CloudBase Auth] 用户状态异常\n'
                + '\n'
                + '  当前用户账号可能处于以下状态之一：\n'
                + '  - 账号已被封禁\n'
                + '  - 账号正在审核中\n'
                + '  - 账号已被锁定（多次密码错误）\n'
                + '\n'
                + "  \u8BF7\u8054\u7CFB\u7BA1\u7406\u5458\u5728\u4E91\u5F00\u53D1\u63A7\u5236\u53F0\u68C0\u67E5\u7528\u6237\u72B6\u6001\uFF1A".concat(CONSOLE_AUTH_URL));
        case AuthErrorCategory.SERVICE_ERROR:
            return ('[CloudBase Auth] 服务异常\n'
                + '\n'
                + '  请检查以下项目：\n'
                + '  1. 网络连接是否正常\n'
                + '  2. 云开发环境是否正常运行\n'
                + '  3. 稍后重试，如持续失败请联系技术支持\n'
                + '\n'
                + '  提示：如果是超时错误，请检查网络环境和请求频率。');
        case AuthErrorCategory.INVALID_PARAMS:
            return ('[CloudBase Auth] 参数错误\n'
                + '\n'
                + '  请检查传入的参数是否符合要求，参考 API 文档确认各参数的格式和约束。\n'
                + "  \u6587\u6863\u5730\u5740\uFF1A".concat(DOCS_URL));
        case AuthErrorCategory.AUTH_METHOD_MISMATCH:
            return ('[CloudBase Auth] 登录方式不匹配\n'
                + '\n'
                + "  \u8C03\u7528 ".concat(methodName, "() \u65F6\u9047\u5230\u9519\u8BEF\uFF0C\u53EF\u80FD\u539F\u56E0\uFF1A\n")
                + '  - 匿名登录失败后尝试了用户名密码登录，但凭据不正确或用户不存在\n'
                + '  - 应使用匿名登录但传入了密码参数\n'
                + '  - 应使用密码登录但未传入正确的凭据\n'
                + '\n'
                + '  建议：\n'
                + '  1. 明确当前场景需要的登录方式：\n'
                + '     - 无需用户注册 → auth.signInAnonymously()\n'
                + '     - 有用户名和密码 → auth.signInWithPassword({ username, password })\n'
                + '     - 第三方平台 → auth.signInWithOAuth({ provider })\n'
                + '  2. 不要在同一流程中混合使用不同的登录方式\n'
                + '  3. 每种登录方式需要在控制台开启对应的身份源\n'
                + "  4. \u63A7\u5236\u53F0\u5730\u5740\uFF1A".concat(CONSOLE_AUTH_URL));
        default:
            return ('[CloudBase Auth] 操作失败\n'
                + '\n'
                + '  如果问题持续存在，建议：\n'
                + "  1. \u68C0\u67E5\u4E91\u5F00\u53D1\u63A7\u5236\u53F0\u7684\u8BA4\u8BC1\u914D\u7F6E\uFF1A".concat(CONSOLE_AUTH_URL, "\n")
                + '  2. 到官方问答社区提问：https://cnb.cool/tencent/cloud/cloudbase/community/-/issues\n'
                + "  3. \u67E5\u9605\u6587\u6863\uFF1A".concat(DOCS_URL));
    }
}
function generateLoginMethodHint(method) {
    if (!method)
        return '';
    var hints = {
        signInWithPassword: '当前使用「用户名密码登录」，需确保：\n'
            + '  ✓ 控制台已开启「用户名密码登录」\n'
            + '  ✓ 用户已注册且凭据正确\n'
            + '  如果不需要用户系统，建议改用 auth.signInAnonymously()',
        signInAnonymously: '当前使用「匿名登录」，需确保：\n'
            + '  ✓ 控制台已开启「匿名登录」\n'
            + '  注意：匿名用户无需用户名密码，适用于快速体验或临时访问场景',
        signInWithOtp: '当前使用「OTP 验证码登录」，需确保：\n' + '  ✓ 控制台已开启「短信验证码登录」或「邮箱验证码登录」',
        signInWithOAuth: '当前使用「第三方 OAuth 登录」，需确保：\n' + '  ✓ 控制台已配置并开启对应的第三方身份源',
    };
    return hints[method] || '';
}
var AuthError = (function (_super) {
    __extends(AuthError, _super);
    function AuthError(error, context) {
        var _this = this;
        var message = error.error_description || error.message;
        _this = _super.call(this, message) || this;
        _this.__isAuthError = true;
        _this.name = 'AuthError';
        _this.status = error.error;
        _this.code = error.error_code;
        _this.requestId = error.requestId;
        _this.category = inferErrorCategory(message, _this.status);
        _this.helpMessage = generateHelpMessage(_this.category, context);
        _this.loginMethodHint = generateLoginMethodHint(context === null || context === void 0 ? void 0 : context.method);
        return _this;
    }
    return AuthError;
}(Error));
exports.AuthError = AuthError;
