// @cloudbase/oauth/dist/cjs/auth/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_NODE_ACCESS_SCOPE = exports.OAUTH_TYPE = exports.EVENTS = exports.AUTH_STATE_CHANGED_TYPE = exports.LOGIN_STATE_CHANGED_TYPE = exports.ErrorType = exports.VerificationUsages = exports.ApiUrlsV2 = exports.ApiUrls = exports.AUTH_API_PREFIX = void 0;
exports.AUTH_API_PREFIX = '/auth';
var ApiUrls;
(function (ApiUrls) {
    ApiUrls["AUTH_SIGN_UP_URL"] = "/v1/signup";
    ApiUrls["AUTH_TOKEN_URL"] = "/v1/token";
    ApiUrls["AUTH_REVOKE_URL"] = "/v1/revoke";
    ApiUrls["WEDA_USER_URL"] = "/v1/plugin/weda/userinfo";
    ApiUrls["AUTH_RESET_PASSWORD"] = "/v1/reset";
    ApiUrls["AUTH_GET_DEVICE_CODE"] = "/v1/device/code";
    ApiUrls["CHECK_USERNAME"] = "/v1/checkUsername";
    ApiUrls["CHECK_IF_USER_EXIST"] = "/v1/checkIfUserExist";
    ApiUrls["GET_PROVIDER_TYPE"] = "/v1/mgr/provider/providerSubType";
    ApiUrls["AUTH_SIGN_IN_URL"] = "/v1/signin";
    ApiUrls["AUTH_SIGN_IN_ANONYMOUSLY_URL"] = "/v1/signin/anonymously";
    ApiUrls["AUTH_SIGN_IN_WITH_PROVIDER_URL"] = "/v1/signin/with/provider";
    ApiUrls["AUTH_SIGN_IN_WITH_WECHAT_URL"] = "/v1/signin/wechat/noauth";
    ApiUrls["AUTH_SIGN_IN_CUSTOM"] = "/v1/signin/custom";
    ApiUrls["PROVIDER_TOKEN_URL"] = "/v1/provider/token";
    ApiUrls["PROVIDER_URI_URL"] = "/v1/provider/uri";
    ApiUrls["USER_ME_URL"] = "/v1/user/me";
    ApiUrls["AUTH_SIGNOUT_URL"] = "/v1/user/signout";
    ApiUrls["USER_QUERY_URL"] = "/v1/user/query";
    ApiUrls["USER_PRIFILE_URL"] = "/v1/user/profile";
    ApiUrls["USER_BASIC_EDIT_URL"] = "/v1/user/basic/edit";
    ApiUrls["USER_TRANS_BY_PROVIDER_URL"] = "/v1/user/trans/by/provider";
    ApiUrls["PROVIDER_LIST"] = "/v1/user/provider";
    ApiUrls["PROVIDER_BIND_URL"] = "/v1/user/provider/bind";
    ApiUrls["PROVIDER_UNBIND_URL"] = "/v1/user/provider";
    ApiUrls["CHECK_PWD_URL"] = "/v1/user/sudo";
    ApiUrls["SUDO_URL"] = "/v1/user/sudo";
    ApiUrls["BIND_CONTACT_URL"] = "/v1/user/contact";
    ApiUrls["AUTH_SET_PASSWORD"] = "/v1/user/password";
    ApiUrls["AUTHORIZE_DEVICE_URL"] = "/v1/user/device/authorize";
    ApiUrls["AUTHORIZE_URL"] = "/v1/user/authorize";
    ApiUrls["AUTHORIZE_INFO_URL"] = "/v1/user/authorize/info";
    ApiUrls["AUTHORIZED_DEVICES_DELETE_URL"] = "/v1/user/authorized/devices/";
    ApiUrls["AUTH_REVOKE_ALL_URL"] = "/v1/user/revoke/all";
    ApiUrls["GET_USER_BEHAVIOR_LOG"] = "/v1/user/security/history";
    ApiUrls["VERIFICATION_URL"] = "/v1/verification";
    ApiUrls["VERIFY_URL"] = "/v1/verification/verify";
    ApiUrls["CAPTCHA_DATA_URL"] = "/v1/captcha/data";
    ApiUrls["VERIFY_CAPTCHA_DATA_URL"] = "/v1/captcha/data/verify";
    ApiUrls["GET_CAPTCHA_URL"] = "/v1/captcha/init";
    ApiUrls["GET_MINIPROGRAM_QRCODE"] = "/v1/qrcode/generate";
    ApiUrls["GET_MINIPROGRAM_QRCODE_STATUS"] = "/v1/qrcode/get/status";
})(ApiUrls = exports.ApiUrls || (exports.ApiUrls = {}));
var ApiUrlsV2;
(function (ApiUrlsV2) {
    ApiUrlsV2["AUTH_SIGN_IN_URL"] = "/v2/signin/username";
    ApiUrlsV2["AUTH_TOKEN_URL"] = "/v2/token";
    ApiUrlsV2["USER_ME_URL"] = "/v2/user/me";
    ApiUrlsV2["VERIFY_URL"] = "/v2/signin/verificationcode";
    ApiUrlsV2["AUTH_SIGN_IN_WITH_PROVIDER_URL"] = "/v2/signin/with/provider";
    ApiUrlsV2["AUTH_PUBLIC_KEY"] = "/v2/signin/publichkey";
    ApiUrlsV2["AUTH_RESET_PASSWORD"] = "/v2/signin/password/update";
})(ApiUrlsV2 = exports.ApiUrlsV2 || (exports.ApiUrlsV2 = {}));
var VerificationUsages;
(function (VerificationUsages) {
    VerificationUsages["REGISTER"] = "REGISTER";
    VerificationUsages["SIGN_IN"] = "SIGN_IN";
    VerificationUsages["PASSWORD_RESET"] = "PASSWORD_RESET";
    VerificationUsages["EMAIL_ADDRESS_CHANGE"] = "EMAIL_ADDRESS_CHANGE";
    VerificationUsages["PHONE_NUMBER_CHANGE"] = "PHONE_NUMBER_CHANGE";
})(VerificationUsages = exports.VerificationUsages || (exports.VerificationUsages = {}));
var ErrorType;
(function (ErrorType) {
    ErrorType["UNREACHABLE"] = "unreachable";
    ErrorType["LOCAL"] = "local";
    ErrorType["CANCELLED"] = "cancelled";
    ErrorType["UNKNOWN"] = "unknown";
    ErrorType["UNAUTHENTICATED"] = "unauthenticated";
    ErrorType["RESOURCE_EXHAUSTED"] = "resource_exhausted";
    ErrorType["FAILED_PRECONDITION"] = "failed_precondition";
    ErrorType["INVALID_ARGUMENT"] = "invalid_argument";
    ErrorType["DEADLINE_EXCEEDED"] = "deadline_exceeded";
    ErrorType["NOT_FOUND"] = "not_found";
    ErrorType["ALREADY_EXISTS"] = "already_exists";
    ErrorType["PERMISSION_DENIED"] = "permission_denied";
    ErrorType["ABORTED"] = "aborted";
    ErrorType["OUT_OF_RANGE"] = "out_of_range";
    ErrorType["UNIMPLEMENTED"] = "unimplemented";
    ErrorType["INTERNAL"] = "internal";
    ErrorType["UNAVAILABLE"] = "unavailable";
    ErrorType["DATA_LOSS"] = "data_loss";
    ErrorType["INVALID_PASSWORD"] = "invalid_password";
    ErrorType["PASSWORD_NOT_SET"] = "password_not_set";
    ErrorType["INVALID_STATUS"] = "invalid_status";
    ErrorType["USER_PENDING"] = "user_pending";
    ErrorType["USER_BLOCKED"] = "user_blocked";
    ErrorType["INVALID_VERIFICATION_CODE"] = "invalid_verification_code";
    ErrorType["TWO_FACTOR_REQUIRED"] = "two_factor_required";
    ErrorType["INVALID_TWO_FACTOR"] = "invalid_two_factor";
    ErrorType["INVALID_TWO_FACTOR_RECOVERY"] = "invalid_two_factor_recovery";
    ErrorType["UNDER_REVIEW"] = "under_review";
    ErrorType["INVALID_REQUEST"] = "invalid_request";
    ErrorType["UNAUTHORIZED_CLIENT"] = "unauthorized_client";
    ErrorType["ACCESS_DENIED"] = "access_denied";
    ErrorType["UNSUPPORTED_RESPONSE_TYPE"] = "unsupported_response_type";
    ErrorType["INVALID_SCOPE"] = "invalid_scope";
    ErrorType["INVALID_GRANT"] = "invalid_grant";
    ErrorType["SERVER_ERROR"] = "server_error";
    ErrorType["TEMPORARILY_UNAVAILABLE"] = "temporarily_unavailable";
    ErrorType["INTERACTION_REQUIRED"] = "interaction_required";
    ErrorType["LOGIN_REQUIRED"] = "login_required";
    ErrorType["ACCOUNT_SELECTION_REQUIRED"] = "account_selection_required";
    ErrorType["CONSENT_REQUIRED"] = "consent_required";
    ErrorType["INVALID_REQUEST_URI"] = "invalid_request_uri";
    ErrorType["INVALID_REQUEST_OBJECT"] = "invalid_request_object";
    ErrorType["REQUEST_NOT_SUPPORTED"] = "request_not_supported";
    ErrorType["REQUEST_URI_NOT_SUPPORTED"] = "request_uri_not_supported";
    ErrorType["REGISTRATION_NOT_SUPPORTED"] = "registration_not_supported";
    ErrorType["CAPTCHA_REQUIRED"] = "captcha_required";
    ErrorType["CAPTCHA_INVALID"] = "captcha_invalid";
    ErrorType["PROVIDER_NOT_ENABLED"] = "provider_not_enabled";
    ErrorType["LOGIN_METHOD_DISABLED"] = "login_method_disabled";
    ErrorType["INVALID_USERNAME_OR_PASSWORD"] = "invalid_username_or_password";
    ErrorType["INVALID_CREDENTIALS"] = "invalid_credentials";
    ErrorType["WRONG_PASSWORD"] = "wrong_password";
    ErrorType["USER_NOT_FOUND"] = "user_not_found";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
exports.LOGIN_STATE_CHANGED_TYPE = {
    SIGN_OUT: 'sign_out',
    SIGN_IN: 'sign_in',
    CREDENTIALS_ERROR: 'credentials_error',
};
exports.AUTH_STATE_CHANGED_TYPE = {
    SIGNED_OUT: 'SIGNED_OUT',
    SIGNED_IN: 'SIGNED_IN',
    INITIAL_SESSION: 'INITIAL_SESSION',
    PASSWORD_RECOVERY: 'PASSWORD_RECOVERY',
    TOKEN_REFRESHED: 'TOKEN_REFRESHED',
    USER_UPDATED: 'USER_UPDATED',
    BIND_IDENTITY: 'BIND_IDENTITY',
};
exports.EVENTS = {
    LOGIN_STATE_CHANGED: 'loginStateChanged',
    AUTH_STATE_CHANGED: 'authStateChanged',
};
exports.OAUTH_TYPE = {
    SIGN_IN: 'sign_in',
    BIND_IDENTITY: 'bind_identity',
};
exports.DEFAULT_NODE_ACCESS_SCOPE = 'node-sdk-scope';
