// @cloudbase/oauth/dist/cjs/captcha/captcha-dom.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openURIWithCallback = void 0;
var events_1 = require("@cloudbase/utilities/dist/cjs/libs/events");
var util_1 = require("@cloudbase/utilities/dist/cjs/libs/util");
var debounce = function (func, delay) {
    var timeoutId;
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        clearTimeout(timeoutId);
        timeoutId = setTimeout(function () { return func.apply(null, args); }, delay);
    };
};
var eventBus = new events_1.CloudbaseEventEmitter();
var CAPTCHA_EVENT = {
    CAPTCHA_DATA_CHANGE: 'captchaDataChange',
    RESOLVE_CAPTCHA_DATA: 'resolveCaptchaData',
};
var CAPTCHA_STYLES = {
    overlay: 'position:fixed;top:0;left:0;width:100%;height:100%;background:rgba(0,0,0,0.5);z-index:999',
    modal: 'position:absolute;top:50%;left:50%;transform:translate(-50%,-50%);width:320px;padding:20px;background:#fff;border-radius:8px;box-shadow:0 0 10px rgba(0,0,0,0.2);z-index:1000',
    prompt: 'margin-bottom:15px',
    captchaWrap: 'display:flex;align-items:center;justify-content:space-between;margin-bottom:15px',
    captchaImg: 'width:100%;height:auto;border:1px solid #eee',
    refreshBtn: 'padding:10px 8px;background:#007bff;color:#fff;border:none;border-radius:4px;cursor:pointer;flex:1;white-space:nowrap;margin-left:10px',
    input: 'width:100%;padding:14px 16px;margin-bottom:15px;box-sizing:border-box;border:2px solid #e8ecf4;border-radius:10px;font-size:15px',
    confirmBtn: 'width:100%;padding:10px;background:#007bff;color:#fff;border:none;border-radius:4px;cursor:pointer',
    loading: 'background:#6c757d;cursor:not-allowed',
};
var genCaptchaDom = function (_a) {
    var oauthInstance = _a.oauthInstance, captchaState = _a.captchaState;
    var CAPTCHA_IMG_ID = 'captcha-image';
    var refreshCaptcha = function () { return __awaiter(void 0, void 0, void 0, function () {
        var result, error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4, oauthInstance.createCaptchaData({ state: captchaState.state })];
                case 1:
                    result = _a.sent();
                    captchaState = __assign(__assign({}, captchaState), { captchaData: result.data, token: result.token });
                    document.getElementById(CAPTCHA_IMG_ID).src = result.data;
                    return [3, 3];
                case 2:
                    error_1 = _a.sent();
                    console.error('刷新图形验证码失败', error_1);
                    return [3, 3];
                case 3: return [2];
            }
        });
    }); };
    var createElement = function (tag, styles, text) {
        var el = document.createElement(tag);
        if (styles)
            el.style.cssText = styles;
        if (text)
            el.textContent = text;
        return el;
    };
    var overlay = createElement('div', CAPTCHA_STYLES.overlay);
    var modal = createElement('div', CAPTCHA_STYLES.modal);
    modal.appendChild(createElement('p', CAPTCHA_STYLES.prompt, '请输入你看到的字符：'));
    var captchaWrap = createElement('div', CAPTCHA_STYLES.captchaWrap);
    var captchaImg = createElement('img', CAPTCHA_STYLES.captchaImg);
    captchaImg.id = CAPTCHA_IMG_ID;
    captchaImg.src = captchaState.captchaData;
    captchaWrap.appendChild(captchaImg);
    var refreshBtn = createElement('button', CAPTCHA_STYLES.refreshBtn, '刷新');
    refreshBtn.onclick = debounce(function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    refreshBtn.textContent = '加载中...';
                    refreshBtn.disabled = true;
                    refreshBtn.style.cssText = "".concat(CAPTCHA_STYLES.refreshBtn, ";").concat(CAPTCHA_STYLES.loading);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, , 3, 4]);
                    return [4, refreshCaptcha()];
                case 2:
                    _a.sent();
                    return [3, 4];
                case 3:
                    refreshBtn.textContent = '刷新';
                    refreshBtn.disabled = false;
                    refreshBtn.style.cssText = CAPTCHA_STYLES.refreshBtn;
                    return [7];
                case 4: return [2];
            }
        });
    }); }, 500);
    captchaWrap.appendChild(refreshBtn);
    modal.appendChild(captchaWrap);
    var input = createElement('input', CAPTCHA_STYLES.input);
    input.type = 'text';
    input.placeholder = '输入字符';
    input.maxLength = 4;
    modal.appendChild(input);
    var errorMsg = createElement('div', 'color:#dc3545;font-size:12px;margin-bottom:10px;display:none;padding:8px;background:#f8d7da;border:1px solid #f5c6cb;border-radius:4px');
    modal.appendChild(errorMsg);
    var showError = function (message) {
        errorMsg.textContent = message;
        errorMsg.style.display = 'block';
    };
    var confirmBtn = createElement('button', CAPTCHA_STYLES.confirmBtn, '确定');
    confirmBtn.onclick = debounce(function () { return __awaiter(void 0, void 0, void 0, function () {
        var inputValue, verifyResult, error_2, errorMessage;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    inputValue = input.value.trim();
                    if (!inputValue) {
                        showError('请输入字符');
                        return [2];
                    }
                    errorMsg.style.display = 'none';
                    confirmBtn.textContent = '验证中...';
                    confirmBtn.disabled = true;
                    confirmBtn.style.cssText = "".concat(CAPTCHA_STYLES.confirmBtn, ";").concat(CAPTCHA_STYLES.loading);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 5, 6]);
                    return [4, oauthInstance.verifyCaptchaData({
                            token: captchaState.token,
                            key: inputValue,
                        })];
                case 2:
                    verifyResult = _a.sent();
                    eventBus.fire(CAPTCHA_EVENT.RESOLVE_CAPTCHA_DATA, verifyResult);
                    console.log('图形验证码校验成功');
                    document.body.removeChild(overlay);
                    document.body.removeChild(modal);
                    return [3, 6];
                case 3:
                    error_2 = _a.sent();
                    console.error('图形验证码校验失败', error_2);
                    errorMessage = error_2.error_description || '图形验证码校验失败';
                    showError(errorMessage);
                    input.value = '';
                    input.focus();
                    return [4, refreshCaptcha()];
                case 4:
                    _a.sent();
                    return [3, 6];
                case 5:
                    confirmBtn.textContent = '确定';
                    confirmBtn.disabled = false;
                    confirmBtn.style.cssText = CAPTCHA_STYLES.confirmBtn;
                    return [7];
                case 6: return [2];
            }
        });
    }); }, 500);
    modal.appendChild(confirmBtn);
    document.body.appendChild(overlay);
    document.body.appendChild(modal);
};
var openURIWithCallback = function (url, oauthInstance) { return __awaiter(void 0, void 0, void 0, function () {
    var _a, captchaData, state, token;
    return __generator(this, function (_b) {
        _a = (0, util_1.parseCaptcha)(url), captchaData = _a.captchaData, state = _a.state, token = _a.token;
        genCaptchaDom({ oauthInstance: oauthInstance, captchaState: { captchaData: captchaData, state: state, token: token } });
        return [2, new Promise(function (resolve) {
                console.log('等待图形验证码校验结果...');
                eventBus.on(CAPTCHA_EVENT.RESOLVE_CAPTCHA_DATA, function (res) {
                    resolve(res === null || res === void 0 ? void 0 : res.data);
                });
            })];
    });
}); };
exports.openURIWithCallback = openURIWithCallback;
