// @cloudbase/database/src/transaction/query.ts
import { Transaction } from './index'
import { EJSON } from 'bson'
import { OrderByDirection, ErrorCode } from '../constant'
import { Validate } from '../validate'
import { QuerySerializer } from '../serializer/query'
import { UpdateSerializer } from '../serializer/update'
import { Util } from '../util'

const GET_DOC = 'database.getInTransaction'
const UPDATE_DOC = 'database.updateDocInTransaction'
const DELETE_DOC = 'database.deleteDocInTransaction'

interface QueryOrder {
  field?: string
  direction?: 'asc' | 'desc'
}

interface QueryOption {
  limit?: number
  offset?: number
  projection?: Object
}

interface TransactionAPI {
  send(interfaceName: string, param?: any)
}

interface GetResult {
  data: any[]
  requestId: string
  total?: number
  limit?: number
  offset?: number
}

interface UpdateResult {
  requestId: string
  updated: number
}

interface DeleteResult {
  requestId: string
  deleted: number | string
}

interface TcbError {
  requestId: string
  code: string
  message: string
  stack?: string
}

/**
 * 事务查询模块
 *
 * 支持 where 条件查询在事务中使用
 */
export class Query {
  /**
   * Collection name
   *
   * @internal
   */
  protected _coll: string

  protected _transaction: Transaction

  /**
   * 过滤条件
   *
   * @internal
   */
  private _fieldFilters: Object

  /**
   * 排序条件
   *
   * @internal
   */
  private _fieldOrders: QueryOrder[]

  /**
   * 查询条件
   *
   * @internal
   */
  private _queryOptions: QueryOption

  /**
   * 请求实例
   *
   * @internal
   */
  private _request: TransactionAPI

  /**
   * 事务ID
   *
   * @internal
   */
  private _transactionId: string

  constructor(
    transaction: Transaction,
    coll: string,
    fieldFilters?: Object,
    fieldOrders?: QueryOrder[],
    queryOptions?: QueryOption
  ) {
    this._coll = coll
    this._transaction = transaction
    this._fieldFilters = fieldFilters
    this._fieldOrders = fieldOrders || []
    this._queryOptions = queryOptions || {}
    this._request = this._transaction.getRequestMethod()
    this._transactionId = this._transaction.getTransactionId()
  }

  /**
   * 查询条件
   *
   * @param query - 查询条件对象
   */
  public where(query: object): Query {
    if (Object.prototype.toString.call(query).slice(8, -1) !== 'Object') {
      throw Error(ErrorCode.QueryParamTypeError)
    }

    const keys = Object.keys(query)
    const checkFlag = keys.some(item => {
      return query[item] !== undefined
    })

    if (keys.length && !checkFlag) {
      throw Error(ErrorCode.QueryParamValueError)
    }

    return new Query(
      this._transaction,
      this._coll,
      QuerySerializer.encode(query),
      this._fieldOrders,
      this._queryOptions
    )
  }

  /**
   * 设置排序方式
   *
   * @param fieldPath    - 字段路径
   * @param directionStr - 排序方式
   */
  public orderBy(fieldPath: string, directionStr: OrderByDirection): Query {
    Validate.isFieldPath(fieldPath)
    Validate.isFieldOrder(directionStr)

    const newOrder: QueryOrder = {
      field: fieldPath,
      direction: directionStr
    }
    const combinedOrders = this._fieldOrders.concat(newOrder)

    return new Query(
      this._transaction,
      this._coll,
      this._fieldFilters,
      combinedOrders,
      this._queryOptions
    )
  }

  /**
   * 设置查询条数
   *
   * @param limit - 限制条数
   */
  public limit(limit: number): Query {
    Validate.isInteger('limit', limit)

    let option = { ...this._queryOptions }
    option.limit = limit

    return new Query(
      this._transaction,
      this._coll,
      this._fieldFilters,
      this._fieldOrders,
      option
    )
  }

  /**
   * 设置偏移量
   *
   * @param offset - 偏移量
   */
  public skip(offset: number): Query {
    Validate.isInteger('offset', offset)

    let option = { ...this._queryOptions }
    option.offset = offset

    return new Query(
      this._transaction,
      this._coll,
      this._fieldFilters,
      this._fieldOrders,
      option
    )
  }

  /**
   * 指定要返回的字段
   *
   * @param projection
   */
  public field(projection: any): Query {
    for (let k in projection) {
      if (projection[k]) {
        if (typeof projection[k] !== 'object') {
          projection[k] = 1
        }
      } else {
        projection[k] = 0
      }
    }

    let option = { ...this._queryOptions }
    option.projection = projection

    return new Query(
      this._transaction,
      this._coll,
      this._fieldFilters,
      this._fieldOrders,
      option
    )
  }

  /**
   * 发起请求获取文档列表（事务内）
   */
  public async get(): Promise<GetResult> {
    let newOrder = []
    if (this._fieldOrders) {
      this._fieldOrders.forEach(order => {
        newOrder.push(order)
      })
    }

    let param: any = {
      collectionName: this._coll,
      transactionId: this._transactionId
    }

    if (this._fieldFilters) {
      param.query = this._fieldFilters
    }
    if (newOrder.length > 0) {
      param.order = newOrder
    }
    if (this._queryOptions.offset) {
      param.offset = this._queryOptions.offset
    }
    if (this._queryOptions.limit) {
      param.limit = this._queryOptions.limit < 1000 ? this._queryOptions.limit : 1000
    } else {
      param.limit = 100
    }
    if (this._queryOptions.projection) {
      param.projection = this._queryOptions.projection
    }

    const res = await this._request.send(GET_DOC, param)
    if (res.code) throw res

    const data = res.data !== 'null' ? EJSON.parse(res.data) : null
    const documents = data ? (Array.isArray(data) ? Util.formatResDocumentData(data) : [Util.formatField(data)]) : []

    const result: GetResult = {
      data: documents,
      requestId: res.requestId
    }
    if (res.total !== undefined) result.total = res.total
    if (res.limit !== undefined) result.limit = res.limit
    if (res.offset !== undefined) result.offset = res.offset

    return result
  }

  /**
   * 获取总数（事务内）
   */
  public async count(): Promise<{ requestId: string; total: number }> {
    let param: any = {
      collectionName: this._coll,
      transactionId: this._transactionId
    }

    if (this._fieldFilters) {
      param.query = this._fieldFilters
    }

    const res = await this._request.send(GET_DOC, param)
    if (res.code) throw res

    const data = res.data !== 'null' ? EJSON.parse(res.data) : null
    const total = Array.isArray(data) ? data.length : (data ? 1 : 0)

    return {
      requestId: res.requestId,
      total: res.total !== undefined ? res.total : total
    }
  }

  /**
   * 发起请求批量更新文档（事务内）
   *
   * @param data - 更新数据
   */
  public async update(data: Object): Promise<UpdateResult> {
    if (!data || typeof data !== 'object') {
      throw new Error('参数必需是非空对象')
    }

    if (data.hasOwnProperty('_id')) {
      throw new Error('不能更新_id的值')
    }

    const param = {
      collectionName: this._coll,
      transactionId: this._transactionId,
      query: this._fieldFilters,
      data: EJSON.stringify(
        UpdateSerializer.encode(data),
        { relaxed: false }
      )
    }

    const res = await this._request.send(UPDATE_DOC, param)
    if (res.code) throw res

    return {
      ...res,
      updated: EJSON.parse(res.updated)
    }
  }

  /**
   * 条件删除文档（事务内）
   */
  public async remove(): Promise<DeleteResult> {
    if (Object.keys(this._queryOptions).length > 0) {
      console.warn('`offset`, `limit` and `projection` are not supported in remove() operation')
    }
    if (this._fieldOrders.length > 0) {
      console.warn('`orderBy` is not supported in remove() operation')
    }

    const param = {
      collectionName: this._coll,
      transactionId: this._transactionId,
      query: this._fieldFilters
    }

    const res: DeleteResult | TcbError = await this._request.send(DELETE_DOC, param)
    if ((res as TcbError).code) throw res

    return {
      ...res,
      deleted: EJSON.parse((res as DeleteResult).deleted as any) as any
    }
  }
}
