// @cloudbase/mysql/dist/cjs/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerMySQL = void 0;
var wx_cloud_client_sdk_1 = require("@cloudbase/wx-cloud-client-sdk");
function getEntity(cloudbase, options, isPg) {
    var _this = this;
    var cloudbaseFetch = cloudbase.request.fetch.bind(cloudbase.request);
    var _a = options || {}, _b = _a.instance, instance = _b === void 0 ? isPg ? undefined : 'default' : _b, _c = _a.database, database = _c === void 0 ? isPg ? undefined : cloudbase.config.env : _c;
    var fetch = function (url, options) { return __awaiter(_this, void 0, void 0, function () {
        var headers, result, data, e_1, res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    headers = {};
                    if (options.headers instanceof Headers) {
                        options.headers.forEach(function (value, key) {
                            headers[key] = value;
                        });
                    }
                    else {
                        headers = options.headers || {};
                    }
                    [
                        ['X-Db-Instance', instance],
                        ['Accept-Profile', database],
                        ['Content-Profile', database],
                    ].forEach(function (_a) {
                        var key = _a[0], value = _a[1];
                        if (!headers[key] && !headers[key.toLocaleLowerCase()] && value) {
                            headers[key] = value;
                        }
                    });
                    return [4, cloudbaseFetch(__assign(__assign({ url: url.toString() }, options), { headers: __assign({ 'Content-Type': 'application/json' }, headers), shouldThrowOnError: false }))];
                case 1:
                    result = _a.sent();
                    data = {};
                    _a.label = 2;
                case 2:
                    _a.trys.push([2, 4, , 5]);
                    return [4, result.data];
                case 3:
                    data = _a.sent();
                    return [3, 5];
                case 4:
                    e_1 = _a.sent();
                    if (!e_1.message.includes('Unexpected end of JSON input')) {
                        throw e_1;
                    }
                    return [3, 5];
                case 5:
                    res = {
                        ok: (result === null || result === void 0 ? void 0 : result.statusCode) >= 200 && (result === null || result === void 0 ? void 0 : result.statusCode) < 300,
                        status: (result === null || result === void 0 ? void 0 : result.statusCode) || 200,
                        statusText: (result === null || result === void 0 ? void 0 : result.statusMessage) || 'OK',
                        json: function () { return Promise.resolve(data || {}); },
                        text: function () { return Promise.resolve(typeof data === 'string' ? data : JSON.stringify(data || {})); },
                        headers: result === null || result === void 0 ? void 0 : result.header,
                    };
                    return [2, res];
            }
        });
    }); };
    var _d = cloudbase.getEndPointWithKey('GATEWAY'), BASE_URL = _d.BASE_URL, PROTOCOL = _d.PROTOCOL;
    var mysqlBaseUrl = "".concat(PROTOCOL).concat(BASE_URL, "/rdb/rest");
    var entity = isPg
        ? (0, wx_cloud_client_sdk_1.generatePGClient)(cloudbase, {
            pgBaseUrl: mysqlBaseUrl,
            fetch: fetch,
        })
        : (0, wx_cloud_client_sdk_1.generateMySQLClient)(cloudbase, {
            mysqlBaseUrl: mysqlBaseUrl,
            fetch: fetch,
        });
    return entity;
}
var component = {
    name: 'mysql',
    entity: {
        mysql: function (options) {
            return getEntity(this, options, false)(options);
        },
    },
};
var rdbComponent = {
    name: 'rdb',
    entity: {
        rdb: function (options) {
            return getEntity(this, options, true)(options);
        },
    },
};
try {
    cloudbase.registerComponent(component);
    cloudbase.registerComponent(rdbComponent);
}
catch (e) { }
function registerMySQL(app) {
    try {
        app.registerComponent(component);
        app.registerComponent(rdbComponent);
    }
    catch (e) {
        console.warn(e);
    }
}
exports.registerMySQL = registerMySQL;
try {
    window.registerMySQL = registerMySQL;
}
catch (e) { }
__exportStar(require("@cloudbase/wx-cloud-client-sdk"), exports);
