// @cloudbase/auth/dist/cjs/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerAuth = exports.AuthError = exports.AuthErrorCategory = exports.AnonymousAuthProvider = exports.CustomAuthProvider = exports.WeixinAuthProvider = exports.Auth = exports.generateAuthInstance = exports.LoginState = exports.User = void 0;
var oauth_1 = require("@cloudbase/oauth");
var adapter_1 = require("./adapter");
var utilities_1 = require("./utilities");
var utils_1 = require("./utils");
var v1_compat_1 = require("./v1-compat");
Object.defineProperty(exports, "WeixinAuthProvider", { enumerable: true, get: function () { return v1_compat_1.WeixinAuthProvider; } });
Object.defineProperty(exports, "CustomAuthProvider", { enumerable: true, get: function () { return v1_compat_1.CustomAuthProvider; } });
Object.defineProperty(exports, "AnonymousAuthProvider", { enumerable: true, get: function () { return v1_compat_1.AnonymousAuthProvider; } });
var utilities_2 = require("@cloudbase/utilities");
var isBrowser = function () { return typeof window !== 'undefined' && typeof document !== 'undefined'; };
var COMPONENT_NAME = 'auth';
var onCredentialsError = function (eventBus) { return function (params) {
    eventBus.fire(oauth_1.EVENTS.LOGIN_STATE_CHANGED, __assign(__assign({}, params), { eventType: oauth_1.LOGIN_STATE_CHANGED_TYPE.CREDENTIALS_ERROR }));
}; };
var User = exports.User = (function () {
    function User(options) {
        var cache = options.cache, oauthInstance = options.oauthInstance;
        this.cache = cache;
        this.oauthInstance = oauthInstance;
        this.setUserInfo();
    }
    User.prototype.checkLocalInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.uid = this.getLocalUserInfo('uid');
                this.gender = this.getLocalUserInfo('gender');
                this.picture = this.getLocalUserInfo('picture');
                this.email = this.getLocalUserInfo('email');
                this.emailVerified = this.getLocalUserInfo('email_verified');
                this.phoneNumber = this.getLocalUserInfo('phone_number');
                this.username = this.getLocalUserInfo('username');
                this.name = this.getLocalUserInfo('name');
                this.birthdate = this.getLocalUserInfo('birthdate');
                this.zoneinfo = this.getLocalUserInfo('zoneinfo');
                this.locale = this.getLocalUserInfo('locale');
                this.sub = this.getLocalUserInfo('sub');
                this.createdFrom = this.getLocalUserInfo('created_from');
                this.providers = this.getLocalUserInfo('providers');
                return [2];
            });
        });
    };
    User.prototype.checkLocalInfoAsync = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        _a = this;
                        return [4, this.getLocalUserInfoAsync('uid')];
                    case 1:
                        _a.uid = _e.sent();
                        _b = this;
                        return [4, this.getLocalUserInfoAsync('gender')];
                    case 2:
                        _b.gender = _e.sent();
                        this.picture = this.getLocalUserInfo('picture');
                        _c = this;
                        return [4, this.getLocalUserInfoAsync('email')];
                    case 3:
                        _c.email = _e.sent();
                        this.emailVerified = this.getLocalUserInfo('email_verified');
                        this.phoneNumber = this.getLocalUserInfo('phone_number');
                        _d = this;
                        return [4, this.getLocalUserInfoAsync('username')];
                    case 4:
                        _d.username = _e.sent();
                        this.name = this.getLocalUserInfo('name');
                        this.birthdate = this.getLocalUserInfo('birthdate');
                        this.zoneinfo = this.getLocalUserInfo('zoneinfo');
                        this.locale = this.getLocalUserInfo('locale');
                        this.sub = this.getLocalUserInfo('sub');
                        this.createdFrom = this.getLocalUserInfo('created_from');
                        this.providers = this.getLocalUserInfo('providers');
                        return [2];
                }
            });
        });
    };
    User.prototype.update = function (userInfo) {
        return __awaiter(this, void 0, void 0, function () {
            var newUserInfo;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.authApi.setUserProfile(__assign({}, userInfo))];
                    case 1:
                        newUserInfo = _a.sent();
                        this.setLocalUserInfo(newUserInfo);
                        return [2];
                }
            });
        });
    };
    User.prototype.updateUserBasicInfo = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.authApi.updateUserBasicInfo(__assign({}, params))];
                    case 1:
                        _a.sent();
                        return [4, this.refresh()];
                    case 2:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    User.prototype.updatePassword = function (newPassword, oldPassword) {
        return this.oauthInstance.authApi.updatePasswordByOld({
            old_password: oldPassword,
            new_password: newPassword,
        });
    };
    User.prototype.updateUsername = function (username) {
        if (typeof username !== 'string') {
            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'username must be a string');
        }
        return this.update({
            username: username,
        });
    };
    User.prototype.refresh = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var newUserInfo;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.authApi.getUserInfo(params)];
                    case 1:
                        newUserInfo = _a.sent();
                        if (newUserInfo.code === 'INVALID_ACCESS_TOKEN') {
                            this.setLocalUserInfo({});
                            throw newUserInfo;
                        }
                        this.setLocalUserInfo(newUserInfo);
                        return [2, newUserInfo];
                }
            });
        });
    };
    User.prototype.getLocalUserInfo = function (key) {
        var userInfoKey = this.cache.keys.userInfoKey;
        var userInfo = this.cache.getStore(userInfoKey);
        if (!key)
            return userInfo || {};
        return userInfo[key];
    };
    User.prototype.setLocalUserInfo = function (userInfo) {
        var userInfoKey = this.cache.keys.userInfoKey;
        this.cache.setStore(userInfoKey, userInfo);
        this.setUserInfo();
    };
    User.prototype.getLocalUserInfoAsync = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            var userInfoKey, userInfo;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        userInfoKey = this.cache.keys.userInfoKey;
                        return [4, this.cache.getStoreAsync(userInfoKey)];
                    case 1:
                        userInfo = _a.sent();
                        return [2, userInfo[key]];
                }
            });
        });
    };
    User.prototype.setUserInfo = function () {
        var _this = this;
        var userInfoKey = this.cache.keys.userInfoKey;
        var userInfo = this.cache.getStore(userInfoKey);
        [
            'uid',
            'email',
            'name',
            'gender',
            'picture',
            'email_verified',
            'phone_number',
            'birthdate',
            'zoneinfo',
            'locale',
            'sub',
            'created_from',
            'providers',
            'username',
            'created_at',
            'open_id',
        ].forEach(function (infoKey) {
            _this[infoKey] = userInfo[infoKey];
        });
    };
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '更新用户信息失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 User.update() 的语法或参数是否正确',
                '  2 - 用户信息中是否包含非法值',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], User.prototype, "update", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '更新密码失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 User.updatePassword() 的语法或参数是否正确',
                '  3 - 新密码中是否包含非法字符',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String, String]),
        __metadata("design:returntype", void 0)
    ], User.prototype, "updatePassword", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '更新用户名失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 User.updateUsername() 的语法或参数是否正确',
                '  2 - 当前环境是否开通了用户名密码登录',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], User.prototype, "updateUsername", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '刷新本地用户信息失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 User.refresh() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], User.prototype, "refresh", null);
    return User;
}());
(0, v1_compat_1.applyUserV1Compat)(User);
var LoginState = (function () {
    function LoginState(options) {
        var envId = options.envId, cache = options.cache, oauthInstance = options.oauthInstance;
        if (!envId) {
            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'envId is not defined');
        }
        this.cache = cache;
        this.oauthInstance = oauthInstance;
        this.user = new User({
            cache: this.cache,
            oauthInstance: oauthInstance,
        });
    }
    LoginState.prototype.checkLocalState = function () {
        var _a;
        this.oauthLoginState = (_a = this.oauthInstance) === null || _a === void 0 ? void 0 : _a.authApi.hasLoginStateSync();
        this.user.checkLocalInfo();
    };
    LoginState.prototype.checkLocalStateAsync = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4, ((_a = this.oauthInstance) === null || _a === void 0 ? void 0 : _a.authApi.getLoginState())];
                    case 1:
                        _b.sent();
                        return [4, this.user.checkLocalInfoAsync()];
                    case 2:
                        _b.sent();
                        return [2];
                }
            });
        });
    };
    return LoginState;
}());
exports.LoginState = LoginState;
var Auth = (function (_super) {
    __extends(Auth, _super);
    function Auth(config) {
        var _this = _super.call(this) || this;
        _this.listeners = new Map();
        _this.hasListenerSetUp = false;
        _this.config = config;
        _this.oauthInstance = config.oauthInstance;
        _this.cache = config.cache;
        _this.init();
        _this.setAccessKey();
        return _this;
    }
    Auth.prototype.signInAnonymously = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var loginState, _a, _b, session, error_1, authError;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 4, , 5]);
                        return [4, this.oauthInstance.authApi.signInAnonymously(params)];
                    case 1:
                        _c.sent();
                        return [4, this.createLoginState()];
                    case 2:
                        loginState = _c.sent();
                        return [4, this.getSession()];
                    case 3:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, __assign(__assign({}, loginState), { data: { user: session.user, session: session }, error: null })];
                    case 4:
                        error_1 = _c.sent();
                        authError = new oauth_1.AuthError(error_1, { method: 'signInAnonymously' });
                        if (authError.category === oauth_1.AuthErrorCategory.PROVIDER_NOT_ENABLED) {
                            console.error(authError.helpMessage);
                            console.error('[CloudBase Auth] 提示：匿名登录需要在控制台开启「匿名登录」。\n'
                                + '  如果您需要使用用户名密码登录，请改用 auth.signInWithPassword({ username, password })。\n'
                                + '  不要在匿名登录失败后自动回退到密码登录，这两种方式的使用场景不同。');
                        }
                        else {
                            console.warn("[CloudBase Auth] signInAnonymously \u5931\u8D25 [".concat(authError.category, "]: ").concat(authError.message));
                            if (authError.loginMethodHint) {
                                console.warn(authError.loginMethodHint);
                            }
                        }
                        return [2, { data: {}, error: authError }];
                    case 5: return [2];
                }
            });
        });
    };
    Auth.prototype.signUp = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var verificationInfo_1, error_2;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(params.phone_number || params.verification_code || params.verification_token || params.provider_token)) return [3, 2];
                        if (params.phone_number) {
                            params.phone_number = this.formatPhone(params.phone_number);
                        }
                        return [4, this.oauthInstance.authApi.signUp(params)];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState()];
                    case 2:
                        _a.trys.push([2, 4, , 5]);
                        this.validateAtLeastOne(params, [['email'], ['phone']], 'You must provide either an email or phone number');
                        return [4, this.getVerification(params.email ? { email: params.email } : { phone_number: this.formatPhone(params.phone) })];
                    case 3:
                        verificationInfo_1 = _a.sent();
                        return [2, {
                                data: {
                                    verifyOtp: function (_a) {
                                        var token = _a.token, _b = _a.messageId, messageId = _b === void 0 ? verificationInfo_1.verification_id : _b;
                                        return __awaiter(_this, void 0, void 0, function () {
                                            var verificationTokenRes, data, _c, _d, session, error_3;
                                            return __generator(this, function (_e) {
                                                switch (_e.label) {
                                                    case 0:
                                                        _e.trys.push([0, 8, , 9]);
                                                        return [4, this.verify({
                                                                verification_id: messageId || verificationInfo_1.verification_id,
                                                                verification_code: token,
                                                            })];
                                                    case 1:
                                                        verificationTokenRes = _e.sent();
                                                        if (!verificationInfo_1.is_user) return [3, 3];
                                                        return [4, this.signIn({
                                                                username: params.email || this.formatPhone(params.phone),
                                                                verification_token: verificationTokenRes.verification_token,
                                                            })];
                                                    case 2:
                                                        _e.sent();
                                                        return [3, 6];
                                                    case 3:
                                                        data = JSON.parse(JSON.stringify(params));
                                                        delete data.email;
                                                        delete data.phone;
                                                        return [4, this.oauthInstance.authApi.signUp(__assign(__assign(__assign({}, data), (params.email ? { email: params.email } : { phone_number: this.formatPhone(params.phone) })), { verification_token: verificationTokenRes.verification_token, verification_code: token }))];
                                                    case 4:
                                                        _e.sent();
                                                        return [4, this.createLoginState()];
                                                    case 5:
                                                        _e.sent();
                                                        _e.label = 6;
                                                    case 6: return [4, this.getSession()];
                                                    case 7:
                                                        _c = (_e.sent()).data, _d = _c === void 0 ? {} : _c, session = _d.session;
                                                        return [2, { data: { user: session.user, session: session }, error: null }];
                                                    case 8:
                                                        error_3 = _e.sent();
                                                        return [2, { data: {}, error: new oauth_1.AuthError(error_3, { method: 'signUp' }) }];
                                                    case 9: return [2];
                                                }
                                            });
                                        });
                                    },
                                },
                                error: null,
                            }];
                    case 4:
                        error_2 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_2, { method: 'signUp' }) }];
                    case 5: return [2];
                }
            });
        });
    };
    Auth.prototype.signOut = function (params) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var userInfoKey, res, error_4;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        userInfoKey = this.cache.keys.userInfoKey;
                        return [4, this.oauthInstance.authApi.signOut(params)];
                    case 1:
                        res = _c.sent();
                        return [4, this.cache.removeStoreAsync(userInfoKey)];
                    case 2:
                        _c.sent();
                        this.setAccessKey();
                        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.LOGIN_STATE_CHANGED, { eventType: oauth_1.LOGIN_STATE_CHANGED_TYPE.SIGN_OUT });
                        (_b = this.config.eventBus) === null || _b === void 0 ? void 0 : _b.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_OUT });
                        return [2, __assign(__assign({}, res), { data: {}, error: null })];
                    case 3:
                        error_4 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_4, { method: 'signOut' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.onAuthStateChange = function (callback) {
        var _this = this;
        if (!this.hasListenerSetUp) {
            this.setupListeners();
            this.hasListenerSetUp = true;
        }
        var id = Math.random().toString(36);
        if (!this.listeners.has(id)) {
            this.listeners.set(id, new Set());
        }
        this.listeners.get(id).add(callback);
        var subscription = {
            id: id,
            callback: callback,
            unsubscribe: function () {
                var callbacks = _this.listeners.get(id);
                if (callbacks) {
                    callbacks.delete(callback);
                    if (callbacks.size === 0) {
                        _this.listeners.delete(id);
                    }
                }
            },
        };
        return {
            data: { subscription: subscription },
        };
    };
    Auth.prototype.signInWithPassword = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, session, error_5, authError;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        this.validateAtLeastOne(params, [['username'], ['email'], ['phone']], 'You must provide either username, email, or phone');
                        this.validateParams(params, {
                            password: { required: true, message: 'Password is required' },
                        });
                        return [4, this.signIn(__assign({ username: params.username || params.email || this.formatPhone(params.phone), password: params.password }, (params.is_encrypt ? { isEncrypt: true, version: 'v2' } : {})))];
                    case 1:
                        _c.sent();
                        return [4, this.getSession()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_5 = _c.sent();
                        authError = new oauth_1.AuthError(error_5, { method: 'signInWithPassword' });
                        if (authError.category === oauth_1.AuthErrorCategory.PROVIDER_NOT_ENABLED) {
                            console.error(authError.helpMessage);
                        }
                        else if (authError.category === oauth_1.AuthErrorCategory.INVALID_CREDENTIALS) {
                            console.warn(authError.helpMessage);
                        }
                        else if (authError.category === oauth_1.AuthErrorCategory.USER_NOT_FOUND) {
                            console.warn(authError.helpMessage);
                        }
                        else {
                            console.warn("[CloudBase Auth] signInWithPassword \u5931\u8D25 [".concat(authError.category, "]: ").concat(authError.message));
                        }
                        if (authError.loginMethodHint) {
                            console.info(authError.loginMethodHint);
                        }
                        return [2, { data: {}, error: authError }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.signInWithIdToken = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, session, error_6;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        this.validateParams(params, {
                            token: { required: true, message: 'Token is required' },
                        });
                        return [4, this.signInWithProvider({
                                provider_token: params.token,
                            })];
                    case 1:
                        _c.sent();
                        return [4, this.getSession()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_6 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_6, { method: 'signInWithIdToken' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.signInWithOtp = function (params) {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function () {
            var verificationInfo_2, error_7;
            var _this = this;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        if (!((_a = params.options) === null || _a === void 0 ? void 0 : _a.emailRedirectTo) && (((_b = params.options) === null || _b === void 0 ? void 0 : _b.shouldCreateUser) === undefined || !!((_c = params.options) === null || _c === void 0 ? void 0 : _c.shouldCreateUser))) {
                            return [2, this.signUp({
                                    email: params.email,
                                    phone: params.phone,
                                })];
                        }
                        _e.label = 1;
                    case 1:
                        _e.trys.push([1, 3, , 4]);
                        this.validateAtLeastOne(params, [['email'], ['phone']], 'You must provide either an email or phone number');
                        return [4, this.getVerification(__assign({ email_redirect_to: (_d = params.options) === null || _d === void 0 ? void 0 : _d.emailRedirectTo }, (params.email ? { email: params.email } : { phone_number: this.formatPhone(params.phone) })))];
                    case 2:
                        verificationInfo_2 = _e.sent();
                        return [2, {
                                data: {
                                    user: null,
                                    session: null,
                                    verifyOtp: function (_a) {
                                        var token = _a.token, _b = _a.messageId, messageId = _b === void 0 ? verificationInfo_2.verification_id : _b;
                                        return __awaiter(_this, void 0, void 0, function () {
                                            return __generator(this, function (_c) {
                                                return [2, this.verifyOtp({
                                                        email: params.email,
                                                        phone: params.phone,
                                                        token: token,
                                                        messageId: messageId,
                                                    })];
                                            });
                                        });
                                    },
                                },
                                error: null,
                            }];
                    case 3:
                        error_7 = _e.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_7, { method: 'signInWithOtp' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.verifyOAuth = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var data, code, state, cacheData, provider, token, res, localSearch, error_8;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        data = {};
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 7, , 8]);
                        code = (params === null || params === void 0 ? void 0 : params.code) || utilities_2.utils.getQuery('code');
                        state = (params === null || params === void 0 ? void 0 : params.state) || utilities_2.utils.getQuery('state');
                        if (!code) {
                            return [2, { data: {}, error: new oauth_1.AuthError({ message: 'Code is required' }) }];
                        }
                        if (!state) {
                            return [2, { data: {}, error: new oauth_1.AuthError({ message: 'State is required' }) }];
                        }
                        cacheData = (0, utils_1.getBrowserSession)(state);
                        data.type = cacheData === null || cacheData === void 0 ? void 0 : cacheData.type;
                        provider = (params === null || params === void 0 ? void 0 : params.provider) || (cacheData === null || cacheData === void 0 ? void 0 : cacheData.provider) || utilities_2.utils.getQuery('provider');
                        if (!provider) {
                            return [2, { data: data, error: new oauth_1.AuthError({ message: 'Provider is required' }) }];
                        }
                        return [4, this.grantProviderToken({
                                provider_id: provider,
                                provider_redirect_uri: location.origin + location.pathname,
                                provider_code: code,
                            })];
                    case 2:
                        token = (_a.sent()).provider_token;
                        res = void 0;
                        if (!(cacheData.type === oauth_1.OAUTH_TYPE.BIND_IDENTITY)) return [3, 4];
                        return [4, this.oauthInstance.authApi.toBindIdentity({ provider_token: token, provider: provider, fireEvent: true })];
                    case 3:
                        res = _a.sent();
                        return [3, 6];
                    case 4: return [4, this.signInWithIdToken({
                            token: token,
                        })];
                    case 5:
                        res = _a.sent();
                        res.data = __assign(__assign({}, data), res.data);
                        _a.label = 6;
                    case 6:
                        localSearch = new URLSearchParams(location === null || location === void 0 ? void 0 : location.search);
                        localSearch.delete('code');
                        localSearch.delete('state');
                        res.data.redirectUrl = (0, utils_1.addUrlSearch)((cacheData === null || cacheData === void 0 ? void 0 : cacheData.search) === undefined ? "?".concat(localSearch.toString()) : cacheData === null || cacheData === void 0 ? void 0 : cacheData.search, (cacheData === null || cacheData === void 0 ? void 0 : cacheData.hash) || location.hash);
                        (0, utils_1.removeBrowserSession)(state);
                        return [2, res];
                    case 7:
                        error_8 = _a.sent();
                        return [2, { data: data, error: new oauth_1.AuthError(error_8, { method: 'verifyOAuth' }) }];
                    case 8: return [2];
                }
            });
        });
    };
    Auth.prototype.signInWithOAuth = function (params) {
        var _a, _b, _c, _d, _e;
        return __awaiter(this, void 0, void 0, function () {
            var href, urlObject, provider_redirect_uri, state, uri, decodedUri, finalUri, url_1, error_9;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _f.trys.push([0, 2, , 3]);
                        this.validateParams(params, {
                            provider: { required: true, message: 'Provider is required' },
                        });
                        href = ((_a = params.options) === null || _a === void 0 ? void 0 : _a.redirectTo) || location.href;
                        urlObject = new URL(href);
                        provider_redirect_uri = urlObject.origin + urlObject.pathname;
                        state = ((_b = params.options) === null || _b === void 0 ? void 0 : _b.state) || "prd-".concat(params.provider, "-").concat(Math.random().toString(36)
                            .slice(2));
                        return [4, this.genProviderRedirectUri({
                                provider_id: params.provider,
                                provider_redirect_uri: provider_redirect_uri,
                                state: state,
                            })];
                    case 1:
                        uri = (_f.sent()).uri;
                        decodedUri = decodeURIComponent(uri);
                        finalUri = decodedUri;
                        if ((_c = params.options) === null || _c === void 0 ? void 0 : _c.queryParams) {
                            url_1 = new URL(decodedUri);
                            Object.entries(params.options.queryParams).forEach(function (_a) {
                                var key = _a[0], value = _a[1];
                                url_1.searchParams.set(key, value);
                            });
                            finalUri = url_1.toString();
                        }
                        (0, utils_1.saveToBrowserSession)(state, {
                            provider: params.provider,
                            search: urlObject.search,
                            hash: urlObject.hash,
                            type: ((_d = params.options) === null || _d === void 0 ? void 0 : _d.type) || oauth_1.OAUTH_TYPE.SIGN_IN,
                        });
                        if (isBrowser() && !((_e = params.options) === null || _e === void 0 ? void 0 : _e.skipBrowserRedirect)) {
                            window.location.assign(finalUri);
                        }
                        return [2, { data: { url: finalUri, provider: params.provider }, error: null }];
                    case 2:
                        error_9 = _f.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_9, { method: 'signInWithOAuth' }) }];
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.getClaims = function () {
        return __awaiter(this, void 0, void 0, function () {
            var accessToken, parsedToken, error_10;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4, this.getAccessToken()];
                    case 1:
                        accessToken = (_a.sent()).accessToken;
                        parsedToken = (0, oauth_1.weAppJwtDecodeAll)(accessToken);
                        return [2, { data: parsedToken, error: null }];
                    case 2:
                        error_10 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_10, { method: 'getClaims' }) }];
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.resetPasswordForEmail = function (emailOrPhone, options) {
        return __awaiter(this, void 0, void 0, function () {
            var redirectTo_1, isEmail_1, verificationParams, formattedPhone, verificationInfo_3, error_11;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateParams({ emailOrPhone: emailOrPhone }, {
                            emailOrPhone: { required: true, message: 'Email or phone is required' },
                        });
                        redirectTo_1 = (options || {}).redirectTo;
                        isEmail_1 = emailOrPhone.includes('@');
                        verificationParams = void 0;
                        if (isEmail_1) {
                            verificationParams = { email: emailOrPhone };
                        }
                        else {
                            formattedPhone = this.formatPhone(emailOrPhone);
                            verificationParams = { phone_number: formattedPhone };
                        }
                        return [4, this.getVerification(verificationParams)];
                    case 1:
                        verificationInfo_3 = _a.sent();
                        return [2, {
                                data: {
                                    updateUser: function (attributes) { return __awaiter(_this, void 0, void 0, function () {
                                        var verificationTokenRes, res, error_12;
                                        var _a;
                                        return __generator(this, function (_b) {
                                            switch (_b.label) {
                                                case 0:
                                                    this.validateParams(attributes, {
                                                        nonce: { required: true, message: 'Nonce is required' },
                                                        password: { required: true, message: 'Password is required' },
                                                    });
                                                    _b.label = 1;
                                                case 1:
                                                    _b.trys.push([1, 5, , 6]);
                                                    return [4, this.verify({
                                                            verification_id: verificationInfo_3.verification_id,
                                                            verification_code: attributes.nonce,
                                                        })];
                                                case 2:
                                                    verificationTokenRes = _b.sent();
                                                    return [4, this.oauthInstance.authApi.resetPassword({
                                                            email: isEmail_1 ? emailOrPhone : undefined,
                                                            phone_number: !isEmail_1 ? emailOrPhone : undefined,
                                                            new_password: attributes.password,
                                                            verification_token: verificationTokenRes.verification_token,
                                                        })];
                                                case 3:
                                                    _b.sent();
                                                    (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, {
                                                        event: oauth_1.AUTH_STATE_CHANGED_TYPE.PASSWORD_RECOVERY,
                                                    });
                                                    return [4, this.signInWithPassword({
                                                            email: isEmail_1 ? emailOrPhone : undefined,
                                                            phone: !isEmail_1 ? emailOrPhone : undefined,
                                                            password: attributes.password,
                                                        })];
                                                case 4:
                                                    res = _b.sent();
                                                    if (redirectTo_1 && isBrowser()) {
                                                        window.location.assign(redirectTo_1);
                                                    }
                                                    return [2, res];
                                                case 5:
                                                    error_12 = _b.sent();
                                                    return [2, { data: {}, error: new oauth_1.AuthError(error_12, { method: 'resetPasswordForEmail' }) }];
                                                case 6: return [2];
                                            }
                                        });
                                    }); },
                                },
                                error: null,
                            }];
                    case 2:
                        error_11 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_11, { method: 'resetPasswordForEmail' }) }];
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.resetPasswordForOld = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, session, error_13;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        return [4, this.oauthInstance.authApi.updatePasswordByOld({
                                old_password: params.old_password,
                                new_password: params.new_password,
                            })];
                    case 1:
                        _c.sent();
                        return [4, this.getSession()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_13 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_13, { method: 'resetPasswordForOld' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.verifyOtp = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var type, _a, _b, session, error_14;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 6, , 7]);
                        type = params.type;
                        this.validateParams(params, {
                            token: { required: true, message: 'Token is required' },
                            messageId: { required: true, message: 'messageId is required' },
                        });
                        if (!['phone_change', 'email_change'].includes(type)) return [3, 2];
                        return [4, this.verify({
                                verification_id: params.messageId,
                                verification_code: params.token,
                            })];
                    case 1:
                        _c.sent();
                        return [3, 4];
                    case 2: return [4, this.signInWithUsername({
                            verificationInfo: { verification_id: params.messageId, is_user: true },
                            verificationCode: params.token,
                            username: params.email || this.formatPhone(params.phone) || '',
                            loginType: params.email ? 'email' : 'phone',
                        })];
                    case 3:
                        _c.sent();
                        _c.label = 4;
                    case 4: return [4, this.getSession()];
                    case 5:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 6:
                        error_14 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_14, { method: 'verifyOtp' }) }];
                    case 7: return [2];
                }
            });
        });
    };
    Auth.prototype.getSession = function () {
        return __awaiter(this, void 0, void 0, function () {
            var credentials, _a, _b, user, error_15;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        return [4, this.oauthInstance.oauth2client.getCredentials()];
                    case 1:
                        credentials = _c.sent();
                        if (!credentials || credentials.scope === 'accessKey') {
                            return [2, { data: { session: null }, error: null }];
                        }
                        return [4, this.getUser(false)];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, user = _b.user;
                        return [2, { data: { session: __assign(__assign({}, credentials), { user: user }), user: user }, error: null }];
                    case 3:
                        error_15 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_15, { method: 'getSession' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.refreshSession = function (refresh_token) {
        return __awaiter(this, void 0, void 0, function () {
            var credentials, newTokens, _a, _b, user, error_16;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 4, , 5]);
                        return [4, this.oauthInstance.oauth2client.localCredentials.getCredentials()];
                    case 1:
                        credentials = _c.sent();
                        credentials.refresh_token = refresh_token || credentials.refresh_token;
                        return [4, this.oauthInstance.oauth2client.refreshToken(credentials)];
                    case 2:
                        newTokens = _c.sent();
                        return [4, this.getUser(false)];
                    case 3:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, user = _b.user;
                        return [2, { data: { user: user, session: __assign(__assign({}, newTokens), { user: user }) }, error: null }];
                    case 4:
                        error_16 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_16, { method: 'refreshSession' }) }];
                    case 5: return [2];
                }
            });
        });
    };
    Auth.prototype.getUser = function (isRefresh) {
        if (isRefresh === void 0) { isRefresh = true; }
        return __awaiter(this, void 0, void 0, function () {
            var user, _a, error_17;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        _a = this.convertToUser;
                        return [4, this.getCurrentUser(isRefresh)];
                    case 1:
                        user = _a.apply(this, [_b.sent()]);
                        return [2, { data: { user: user }, error: null }];
                    case 2:
                        error_17 = _b.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_17, { method: 'getUser' }) }];
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.refreshUser = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, session, error_18;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 3, , 4]);
                        return [4, this.currentUser.refresh()];
                    case 1:
                        _c.sent();
                        return [4, this.getSession()];
                    case 2:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_18 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_18, { method: 'refreshUser' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.updateUser = function (params) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var hasValue, email, phone, restParams, needsEmailVerification_1, needsPhoneVerification, extraRes, verificationParams, verificationType_1, formattedPhone, verificationInfo_4, _b, user, error_19;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 8, , 9]);
                        hasValue = Object.keys(params).some(function (key) { return params[key] !== undefined && params[key] !== null && params[key] !== ''; });
                        if (!hasValue) {
                            throw new oauth_1.AuthError({ message: 'At least one field must be provided for update' });
                        }
                        email = params.email, phone = params.phone, restParams = __rest(params, ["email", "phone"]);
                        needsEmailVerification_1 = email !== undefined;
                        needsPhoneVerification = phone !== undefined;
                        extraRes = {};
                        if (!(needsEmailVerification_1 || needsPhoneVerification)) return [3, 4];
                        verificationParams = void 0;
                        if (needsEmailVerification_1) {
                            verificationParams = { email: params.email };
                            verificationType_1 = 'email_change';
                        }
                        else {
                            formattedPhone = this.formatPhone(params.phone);
                            verificationParams = { phone_number: formattedPhone };
                            verificationType_1 = 'phone_change';
                        }
                        return [4, this.getVerification(verificationParams)];
                    case 1:
                        verificationInfo_4 = _c.sent();
                        _b = Object.keys(restParams).length > 0;
                        if (!_b) return [3, 3];
                        return [4, this.updateUserBasicInfo(restParams)];
                    case 2:
                        _b = (_c.sent());
                        _c.label = 3;
                    case 3:
                        _b;
                        extraRes = {
                            messageId: verificationInfo_4.verification_id,
                            verifyOtp: function (verifyParams) { return __awaiter(_this, void 0, void 0, function () {
                                var user_1, error_20;
                                var _a;
                                return __generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            _b.trys.push([0, 11, , 12]);
                                            if (!(verifyParams.email && params.email === verifyParams.email)) return [3, 3];
                                            return [4, this.verifyOtp({
                                                    type: 'email_change',
                                                    email: params.email,
                                                    token: verifyParams.token,
                                                    messageId: verificationInfo_4.verification_id,
                                                })];
                                        case 1:
                                            _b.sent();
                                            return [4, this.updateUserBasicInfo({ email: params.email })];
                                        case 2:
                                            _b.sent();
                                            return [3, 9];
                                        case 3:
                                            if (!(verifyParams.phone && params.phone === verifyParams.phone)) return [3, 6];
                                            return [4, this.verifyOtp({
                                                    type: 'phone_change',
                                                    phone: params.phone,
                                                    token: verifyParams.token,
                                                    messageId: verificationInfo_4.verification_id,
                                                })];
                                        case 4:
                                            _b.sent();
                                            return [4, this.updateUserBasicInfo({ phone: this.formatPhone(params.phone) })];
                                        case 5:
                                            _b.sent();
                                            return [3, 9];
                                        case 6: return [4, this.verifyOtp({
                                                type: verificationType_1,
                                                email: needsEmailVerification_1 ? params.email : undefined,
                                                phone: !needsEmailVerification_1 ? params.phone : undefined,
                                                token: verifyParams.token,
                                                messageId: verificationInfo_4.verification_id,
                                            })];
                                        case 7:
                                            _b.sent();
                                            return [4, this.updateUserBasicInfo(params)];
                                        case 8:
                                            _b.sent();
                                            _b.label = 9;
                                        case 9: return [4, this.getUser(false)];
                                        case 10:
                                            user_1 = (_b.sent()).data.user;
                                            (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.USER_UPDATED });
                                            return [2, { data: { user: user_1 }, error: null }];
                                        case 11:
                                            error_20 = _b.sent();
                                            return [2, { data: {}, error: new oauth_1.AuthError(error_20, { method: 'updateUser' }) }];
                                        case 12: return [2];
                                    }
                                });
                            }); },
                        };
                        return [3, 6];
                    case 4: return [4, this.updateUserBasicInfo(params)];
                    case 5:
                        _c.sent();
                        _c.label = 6;
                    case 6: return [4, this.getUser(false)];
                    case 7:
                        user = (_c.sent()).data.user;
                        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.USER_UPDATED });
                        return [2, { data: __assign({ user: user }, extraRes), error: null }];
                    case 8:
                        error_19 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_19, { method: 'updateUser' }) }];
                    case 9: return [2];
                }
            });
        });
    };
    Auth.prototype.getUserIdentities = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var providers, error_21;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        return [4, this.oauthInstance.authApi.getProviders()];
                    case 1:
                        providers = _b.sent();
                        return [2, {
                                data: {
                                    identities: (_a = providers === null || providers === void 0 ? void 0 : providers.data) === null || _a === void 0 ? void 0 : _a.filter(function (v) { return !!v.bind; }),
                                },
                                error: null,
                            }];
                    case 2:
                        error_21 = _b.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_21, { method: 'getUserIdentities' }) }];
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.linkIdentity = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var error_22;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateParams(params, {
                            provider: { required: true, message: 'Provider is required' },
                        });
                        return [4, this.signInWithOAuth({
                                provider: params.provider,
                                options: {
                                    type: oauth_1.OAUTH_TYPE.BIND_IDENTITY,
                                },
                            })];
                    case 1:
                        _a.sent();
                        return [2, { data: { provider: params.provider }, error: null }];
                    case 2:
                        error_22 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_22, { method: 'linkIdentity' }) }];
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.unlinkIdentity = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var error_23;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateParams(params, {
                            provider: { required: true, message: 'Provider is required' },
                        });
                        return [4, this.oauthInstance.authApi.unbindProvider({ provider_id: params.provider })];
                    case 1:
                        _a.sent();
                        return [2, { data: {}, error: null }];
                    case 2:
                        error_23 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_23, { method: 'unlinkIdentity' }) }];
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.reauthenticate = function () {
        return __awaiter(this, void 0, void 0, function () {
            var user, userInfo_1, verificationInfo_5, error_24;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4, this.getUser(false)];
                    case 1:
                        user = (_a.sent()).data.user;
                        this.validateAtLeastOne(user, [['email', 'phone']], 'You must provide either an email or phone number');
                        userInfo_1 = user.email ? { email: user.email } : { phone_number: this.formatPhone(user.phone) };
                        return [4, this.getVerification(userInfo_1)];
                    case 2:
                        verificationInfo_5 = _a.sent();
                        return [2, {
                                data: {
                                    updateUser: function (attributes) { return __awaiter(_this, void 0, void 0, function () {
                                        var verificationTokenRes, sudoRes, _a, _b, session, error_25;
                                        return __generator(this, function (_c) {
                                            switch (_c.label) {
                                                case 0:
                                                    this.validateParams(attributes, {
                                                        nonce: { required: true, message: 'Nonce is required' },
                                                    });
                                                    _c.label = 1;
                                                case 1:
                                                    _c.trys.push([1, 9, , 10]);
                                                    if (!attributes.password) return [3, 5];
                                                    return [4, this.verify({
                                                            verification_id: verificationInfo_5.verification_id,
                                                            verification_code: attributes.nonce,
                                                        })];
                                                case 2:
                                                    verificationTokenRes = _c.sent();
                                                    return [4, this.oauthInstance.authApi.sudo({
                                                            verification_token: verificationTokenRes.verification_token,
                                                        })];
                                                case 3:
                                                    sudoRes = _c.sent();
                                                    return [4, this.oauthInstance.authApi.setPassword({
                                                            new_password: attributes.password,
                                                            sudo_token: sudoRes.sudo_token,
                                                        })];
                                                case 4:
                                                    _c.sent();
                                                    return [3, 7];
                                                case 5: return [4, this.signInWithUsername(__assign(__assign({ verificationInfo: verificationInfo_5, verificationCode: attributes.nonce }, userInfo_1), { loginType: userInfo_1.email ? 'email' : 'phone' }))];
                                                case 6:
                                                    _c.sent();
                                                    _c.label = 7;
                                                case 7: return [4, this.getSession()];
                                                case 8:
                                                    _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                                                    return [2, { data: { user: session.user, session: session }, error: null }];
                                                case 9:
                                                    error_25 = _c.sent();
                                                    return [2, { data: {}, error: new oauth_1.AuthError(error_25, { method: 'reauthenticate' }) }];
                                                case 10: return [2];
                                            }
                                        });
                                    }); },
                                },
                                error: null,
                            }];
                    case 3:
                        error_24 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_24, { method: 'reauthenticate' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.resend = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var target, data, verificationId, error_26;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.validateAtLeastOne(params, [['email'], ['phone']], 'You must provide either an email or phone number');
                        target = params.type === 'signup' ? 'ANY' : 'USER';
                        data = { target: target };
                        if ('email' in params) {
                            data.email = params.email;
                        }
                        if ('phone' in params) {
                            data.phone_number = this.formatPhone(params.phone);
                        }
                        return [4, this.oauthInstance.authApi.getVerification(data)];
                    case 1:
                        verificationId = (_a.sent()).verification_id;
                        return [2, {
                                data: { messageId: verificationId },
                                error: null,
                            }];
                    case 2:
                        error_26 = _a.sent();
                        return [2, {
                                data: {},
                                error: new oauth_1.AuthError(error_26, { method: 'resend' }),
                            }];
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.setSession = function (params) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var _b, _c, session, error_27;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _d.trys.push([0, 3, , 4]);
                        this.validateParams(params, {
                            access_token: { required: true, message: 'Access token is required' },
                            refresh_token: { required: true, message: 'Refresh token is required' },
                        });
                        return [4, this.oauthInstance.oauth2client.refreshToken(params, { throwError: true })];
                    case 1:
                        _d.sent();
                        return [4, this.getSession()];
                    case 2:
                        _b = (_d.sent()).data, _c = _b === void 0 ? {} : _b, session = _c.session;
                        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_IN });
                        return [2, { data: { user: session.user, session: session }, error: null }];
                    case 3:
                        error_27 = _d.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_27, { method: 'setSession' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.exchangeCodeForSession = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2];
            });
        });
    };
    Auth.prototype.deleteUser = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var sudo_token, error_28;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        this.validateParams(params, {
                            password: { required: true, message: 'Password is required' },
                        });
                        return [4, this.oauthInstance.authApi.sudo(params)];
                    case 1:
                        sudo_token = (_a.sent()).sudo_token;
                        return [4, this.oauthInstance.authApi.deleteMe({ sudo_token: sudo_token })];
                    case 2:
                        _a.sent();
                        return [2, { data: {}, error: null }];
                    case 3:
                        error_28 = _a.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_28, { method: 'deleteUser' }) }];
                    case 4: return [2];
                }
            });
        });
    };
    Auth.prototype.toDefaultLoginPage = function (params) {
        if (params === void 0) { params = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var configVersion, query, redirectUri, urlObj, loginPage;
            return __generator(this, function (_a) {
                try {
                    configVersion = params.config_version || 'env';
                    query = Object.keys(params.query || {})
                        .map(function (key) { return "".concat(key, "=").concat(params.query[key]); })
                        .join('&');
                    if (utilities_1.adapterForWxMp.isMatch()) {
                        wx.navigateTo({ url: "/packages/$wd_system/pages/login/index".concat(query ? "?".concat(query) : '') });
                    }
                    else {
                        redirectUri = params.redirect_uri || window.location.href;
                        urlObj = new URL(redirectUri);
                        loginPage = "".concat(urlObj.origin, "/__auth/?app_id=").concat(params.app_id || '', "&env_id=").concat(this.config.env, "&client_id=").concat(this.config.clientId || this.config.env, "&config_version=").concat(configVersion, "&redirect_uri=").concat(encodeURIComponent(redirectUri)).concat(query ? "&".concat(query) : '');
                        window.location.href = loginPage;
                    }
                    return [2, { data: {}, error: null }];
                }
                catch (error) {
                    return [2, { data: {}, error: new oauth_1.AuthError(error, { method: 'toDefaultLoginPage' }) }];
                }
                return [2];
            });
        });
    };
    Auth.prototype.signInWithCustomTicket = function (getTickFn) {
        return __awaiter(this, void 0, void 0, function () {
            var loginState, _a, _b, session, error_29;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (getTickFn) {
                            this.setCustomSignFunc(getTickFn);
                        }
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 5, , 6]);
                        return [4, this.oauthInstance.authApi.signInWithCustomTicket()];
                    case 2:
                        _c.sent();
                        return [4, this.createLoginState()];
                    case 3:
                        loginState = _c.sent();
                        return [4, this.getSession()];
                    case 4:
                        _a = (_c.sent()).data, _b = _a === void 0 ? {} : _a, session = _b.session;
                        return [2, __assign(__assign({}, loginState), { data: { user: session.user, session: session }, error: null })];
                    case 5:
                        error_29 = _c.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_29, { method: 'signInWithCustomTicket' }) }];
                    case 6: return [2];
                }
            });
        });
    };
    Auth.prototype.signInWithOpenId = function (_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.useWxCloud, useWxCloud = _c === void 0 ? true : _c;
        return __awaiter(this, void 0, void 0, function () {
            var wxInfo, mainFunc, loginState, _d, _e, session, error_30;
            var _this = this;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        if (!utilities_1.adapterForWxMp.isMatch()) {
                            throw Error('wx api undefined');
                        }
                        wxInfo = wx.getAccountInfoSync().miniProgram;
                        mainFunc = function (code) { return __awaiter(_this, void 0, void 0, function () {
                            var result, credentials, error_31;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        result = undefined;
                                        credentials = undefined;
                                        _a.label = 1;
                                    case 1:
                                        _a.trys.push([1, 4, , 5]);
                                        return [4, this.oauthInstance.authApi.grantProviderToken({
                                                provider_id: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                provider_code: code,
                                                provider_params: {
                                                    provider_code_type: 'open_id',
                                                    appid: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                },
                                            }, useWxCloud)];
                                    case 2:
                                        result = _a.sent();
                                        if ((result === null || result === void 0 ? void 0 : result.error_code) || !result.provider_token) {
                                            throw result;
                                        }
                                        return [4, this.oauthInstance.authApi.signInWithProvider({ provider_token: result.provider_token }, useWxCloud)];
                                    case 3:
                                        credentials = _a.sent();
                                        if (credentials === null || credentials === void 0 ? void 0 : credentials.error_code) {
                                            throw credentials;
                                        }
                                        return [3, 5];
                                    case 4:
                                        error_31 = _a.sent();
                                        throw error_31;
                                    case 5: return [4, this.oauthInstance.oauth2client.setCredentials(credentials)];
                                    case 6:
                                        _a.sent();
                                        return [2];
                                }
                            });
                        }); };
                        _f.label = 1;
                    case 1:
                        _f.trys.push([1, 5, , 6]);
                        return [4, new Promise(function (resolve, reject) {
                                wx.login({
                                    success: function (res) { return __awaiter(_this, void 0, void 0, function () {
                                        var error_32;
                                        return __generator(this, function (_a) {
                                            switch (_a.label) {
                                                case 0:
                                                    _a.trys.push([0, 2, , 3]);
                                                    return [4, mainFunc(res.code)];
                                                case 1:
                                                    _a.sent();
                                                    resolve(true);
                                                    return [3, 3];
                                                case 2:
                                                    error_32 = _a.sent();
                                                    reject(error_32);
                                                    return [3, 3];
                                                case 3: return [2];
                                            }
                                        });
                                    }); },
                                    fail: function (res) {
                                        var error = new Error(res === null || res === void 0 ? void 0 : res.errMsg);
                                        error.code = res === null || res === void 0 ? void 0 : res.errno;
                                        reject(error);
                                    },
                                });
                            })];
                    case 2:
                        _f.sent();
                        return [4, this.createLoginState()];
                    case 3:
                        loginState = _f.sent();
                        return [4, this.getSession()];
                    case 4:
                        _d = (_f.sent()).data, _e = _d === void 0 ? {} : _d, session = _e.session;
                        return [2, __assign(__assign({}, loginState), { data: { user: session.user, session: session }, error: null })];
                    case 5:
                        error_30 = _f.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_30, { method: 'signInWithOpenId' }) }];
                    case 6: return [2];
                }
            });
        });
    };
    Auth.prototype.signInWithPhoneAuth = function (_a) {
        var _b = _a.phoneCode, phoneCode = _b === void 0 ? '' : _b, _c = _a.useWxCloud, useWxCloud = _c === void 0 ? false : _c;
        return __awaiter(this, void 0, void 0, function () {
            var wxInfo, providerInfo, code, providerToken, signInRes, error_33, loginState, _d, _e, session;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        if (!utilities_1.adapterForWxMp.isMatch()) {
                            return [2, { data: {}, error: new oauth_1.AuthError({ message: 'wx api undefined' }) }];
                        }
                        wxInfo = wx.getAccountInfoSync().miniProgram;
                        providerInfo = {
                            provider_params: { provider_code_type: 'phone', code: phoneCode, appid: wxInfo.appId },
                            provider_id: wxInfo.appId,
                        };
                        return [4, wx.login()];
                    case 1:
                        code = (_f.sent()).code;
                        providerInfo.provider_code = code;
                        _f.label = 2;
                    case 2:
                        _f.trys.push([2, 5, , 6]);
                        return [4, this.oauthInstance.authApi.grantProviderToken(providerInfo, useWxCloud)];
                    case 3:
                        providerToken = _f.sent();
                        if (providerToken.error_code) {
                            throw providerToken;
                        }
                        return [4, this.oauthInstance.authApi.signInWithProvider({
                                provider_token: providerToken.provider_token,
                            }, useWxCloud)];
                    case 4:
                        signInRes = _f.sent();
                        if (signInRes === null || signInRes === void 0 ? void 0 : signInRes.error_code) {
                            throw signInRes;
                        }
                        return [3, 6];
                    case 5:
                        error_33 = _f.sent();
                        return [2, { data: {}, error: new oauth_1.AuthError(error_33, { method: 'signInWithPhoneAuth' }) }];
                    case 6: return [4, this.createLoginState()];
                    case 7:
                        loginState = _f.sent();
                        return [4, this.getSession()];
                    case 8:
                        _d = (_f.sent()).data, _e = _d === void 0 ? {} : _d, session = _e.session;
                        return [2, __assign(__assign({}, loginState), { data: { user: session.user, session: session }, error: null })];
                }
            });
        });
    };
    Auth.prototype.bindPhoneNumber = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.editContact(params)];
            });
        });
    };
    Auth.prototype.unbindProvider = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.unbindProvider(params)];
            });
        });
    };
    Auth.prototype.bindEmail = function (params) {
        return this.oauthInstance.authApi.editContact(params);
    };
    Auth.prototype.verify = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.verify(params)];
            });
        });
    };
    Auth.prototype.getVerification = function (params, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (params.phone_number) {
                    params.phone_number = this.formatPhone(params.phone_number);
                }
                return [2, this.oauthInstance.authApi.getVerification(params, options)];
            });
        });
    };
    Object.defineProperty(Auth.prototype, "currentUser", {
        get: function () {
            if (this.cache.mode === 'async') {
                (0, utilities_1.printWarn)(utilities_1.ERRORS.INVALID_OPERATION, 'current platform\'s storage is asynchronous, please use getCurrentUser instead');
                return;
            }
            var loginState = this.hasLoginState();
            if (loginState) {
                return loginState.user || null;
            }
            return null;
        },
        enumerable: false,
        configurable: true
    });
    Auth.prototype.getCurrentUser = function (isRefresh) {
        if (isRefresh === void 0) { isRefresh = true; }
        return __awaiter(this, void 0, void 0, function () {
            var loginState_1, error_34, loginState, userInfo;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        if (!isRefresh) return [3, 3];
                        return [4, this.getLoginState()];
                    case 1:
                        loginState_1 = _a.sent();
                        if (!loginState_1) return [3, 3];
                        return [4, loginState_1.user.refresh()];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [3, 5];
                    case 4:
                        error_34 = _a.sent();
                        return [3, 5];
                    case 5: return [4, this.getLoginState()];
                    case 6:
                        loginState = _a.sent();
                        if (!loginState) return [3, 8];
                        userInfo = loginState.user.getLocalUserInfo();
                        return [4, loginState.user.checkLocalInfoAsync()];
                    case 7:
                        _a.sent();
                        return [2, __assign(__assign({}, loginState.user), userInfo)];
                    case 8: return [2, null];
                }
            });
        });
    };
    Auth.prototype.signInAnonymouslyInWx = function (_a) {
        var _b = _a === void 0 ? {} : _a, useWxCloud = _b.useWxCloud;
        return __awaiter(this, void 0, void 0, function () {
            var wxInfo, mainFunc;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!utilities_1.adapterForWxMp.isMatch()) {
                            throw Error('wx api undefined');
                        }
                        wxInfo = wx.getAccountInfoSync().miniProgram;
                        mainFunc = function (code) { return __awaiter(_this, void 0, void 0, function () {
                            var result, credentials, error_35;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        result = undefined;
                                        credentials = undefined;
                                        _a.label = 1;
                                    case 1:
                                        _a.trys.push([1, 4, , 5]);
                                        return [4, this.oauthInstance.authApi.grantProviderToken({
                                                provider_id: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                provider_code: code,
                                                provider_params: {
                                                    provider_code_type: 'open_id',
                                                    appid: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                },
                                            }, useWxCloud)];
                                    case 2:
                                        result = _a.sent();
                                        if ((result === null || result === void 0 ? void 0 : result.error_code) || !result.provider_token) {
                                            throw result;
                                        }
                                        return [4, this.oauthInstance.authApi.signInAnonymously({ provider_token: result.provider_token }, useWxCloud)];
                                    case 3:
                                        credentials = _a.sent();
                                        if (credentials === null || credentials === void 0 ? void 0 : credentials.error_code) {
                                            throw credentials;
                                        }
                                        return [3, 5];
                                    case 4:
                                        error_35 = _a.sent();
                                        throw error_35;
                                    case 5: return [2];
                                }
                            });
                        }); };
                        return [4, new Promise(function (resolve, reject) {
                                wx.login({
                                    success: function (res) { return __awaiter(_this, void 0, void 0, function () {
                                        var error_36;
                                        return __generator(this, function (_a) {
                                            switch (_a.label) {
                                                case 0:
                                                    _a.trys.push([0, 2, , 3]);
                                                    return [4, mainFunc(res.code)];
                                                case 1:
                                                    _a.sent();
                                                    resolve(true);
                                                    return [3, 3];
                                                case 2:
                                                    error_36 = _a.sent();
                                                    reject(error_36);
                                                    return [3, 3];
                                                case 3: return [2];
                                            }
                                        });
                                    }); },
                                    fail: function (res) {
                                        var error = new Error(res === null || res === void 0 ? void 0 : res.errMsg);
                                        error.code = res === null || res === void 0 ? void 0 : res.errno;
                                        reject(error);
                                    },
                                });
                            })];
                    case 1:
                        _c.sent();
                        return [2, this.createLoginState(undefined, { asyncRefreshUser: true })];
                }
            });
        });
    };
    Auth.prototype.bindOpenId = function () {
        return __awaiter(this, void 0, void 0, function () {
            var wxInfo, mainFunc;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!utilities_1.adapterForWxMp.isMatch()) {
                            throw Error('wx api undefined');
                        }
                        wxInfo = wx.getAccountInfoSync().miniProgram;
                        mainFunc = function (code) { return __awaiter(_this, void 0, void 0, function () {
                            var result, error_37;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        result = undefined;
                                        _a.label = 1;
                                    case 1:
                                        _a.trys.push([1, 4, , 5]);
                                        return [4, this.oauthInstance.authApi.grantProviderToken({
                                                provider_id: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                provider_code: code,
                                                provider_params: {
                                                    provider_code_type: 'open_id',
                                                    appid: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                },
                                            })];
                                    case 2:
                                        result = _a.sent();
                                        if ((result === null || result === void 0 ? void 0 : result.error_code) || !result.provider_token) {
                                            throw result;
                                        }
                                        return [4, this.oauthInstance.authApi.bindWithProvider({ provider_token: result.provider_token })];
                                    case 3:
                                        _a.sent();
                                        return [3, 5];
                                    case 4:
                                        error_37 = _a.sent();
                                        throw error_37;
                                    case 5: return [2];
                                }
                            });
                        }); };
                        return [4, new Promise(function (resolve, reject) {
                                wx.login({
                                    success: function (res) { return __awaiter(_this, void 0, void 0, function () {
                                        var error_38;
                                        return __generator(this, function (_a) {
                                            switch (_a.label) {
                                                case 0:
                                                    _a.trys.push([0, 2, , 3]);
                                                    return [4, mainFunc(res.code)];
                                                case 1:
                                                    _a.sent();
                                                    resolve(true);
                                                    return [3, 3];
                                                case 2:
                                                    error_38 = _a.sent();
                                                    reject(error_38);
                                                    return [3, 3];
                                                case 3: return [2];
                                            }
                                        });
                                    }); },
                                    fail: function (res) {
                                        var error = new Error(res === null || res === void 0 ? void 0 : res.errMsg);
                                        error.code = res === null || res === void 0 ? void 0 : res.errno;
                                        reject(error);
                                    },
                                });
                            })];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    Auth.prototype.signInWithUnionId = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_39;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!utilities_1.adapterForWxMp.isMatch()) {
                            throw Error('wx api undefined');
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4, new Promise(function (resolve, reject) {
                                var wxInfo = wx.getAccountInfoSync().miniProgram;
                                wx.login({
                                    success: function (res) { return __awaiter(_this, void 0, void 0, function () {
                                        var providerId, result, providerToken, signInRes, error_40;
                                        return __generator(this, function (_a) {
                                            switch (_a.label) {
                                                case 0:
                                                    providerId = wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId;
                                                    _a.label = 1;
                                                case 1:
                                                    _a.trys.push([1, 4, , 5]);
                                                    return [4, this.oauthInstance.authApi.grantProviderToken({
                                                            provider_code: res.code,
                                                            provider_id: providerId,
                                                            provider_params: {
                                                                provider_code_type: 'union_id',
                                                                appid: wxInfo === null || wxInfo === void 0 ? void 0 : wxInfo.appId,
                                                            },
                                                        })];
                                                case 2:
                                                    result = _a.sent();
                                                    providerToken = result.provider_token;
                                                    if (!providerToken) {
                                                        reject(result);
                                                        return [2];
                                                    }
                                                    return [4, this.oauthInstance.authApi.signInWithProvider({
                                                            provider_id: providerId,
                                                            provider_token: providerToken,
                                                        })];
                                                case 3:
                                                    signInRes = _a.sent();
                                                    if (signInRes === null || signInRes === void 0 ? void 0 : signInRes.error_code) {
                                                        reject(signInRes);
                                                        return [2];
                                                    }
                                                    resolve(true);
                                                    return [3, 5];
                                                case 4:
                                                    error_40 = _a.sent();
                                                    reject(error_40);
                                                    return [3, 5];
                                                case 5: return [2];
                                            }
                                        });
                                    }); },
                                    fail: function (res) {
                                        var error = new Error(res === null || res === void 0 ? void 0 : res.errMsg);
                                        error.code = res === null || res === void 0 ? void 0 : res.errno;
                                        reject(error);
                                    },
                                });
                            })];
                    case 2:
                        _a.sent();
                        return [3, 4];
                    case 3:
                        error_39 = _a.sent();
                        throw error_39;
                    case 4: return [2, this.createLoginState()];
                }
            });
        });
    };
    Auth.prototype.signInWithSms = function (_a) {
        var _b = _a.verificationInfo, verificationInfo = _b === void 0 ? { verification_id: '', is_user: false } : _b, _c = _a.verificationCode, verificationCode = _c === void 0 ? '' : _c, _d = _a.phoneNum, phoneNum = _d === void 0 ? '' : _d, _e = _a.bindInfo, bindInfo = _e === void 0 ? undefined : _e;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_f) {
                try {
                    return [2, this.signInWithUsername({
                            verificationInfo: verificationInfo,
                            verificationCode: verificationCode,
                            bindInfo: bindInfo,
                            username: phoneNum,
                            loginType: 'sms',
                        })];
                }
                catch (error) {
                    throw error;
                }
                return [2];
            });
        });
    };
    Auth.prototype.signInWithEmail = function (_a) {
        var _b = _a.verificationInfo, verificationInfo = _b === void 0 ? { verification_id: '', is_user: false } : _b, _c = _a.verificationCode, verificationCode = _c === void 0 ? '' : _c, _d = _a.bindInfo, bindInfo = _d === void 0 ? undefined : _d, _e = _a.email, email = _e === void 0 ? '' : _e;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_f) {
                try {
                    return [2, this.signInWithUsername({
                            verificationInfo: verificationInfo,
                            verificationCode: verificationCode,
                            bindInfo: bindInfo,
                            username: email,
                            loginType: 'email',
                        })];
                }
                catch (error) {
                    throw error;
                }
                return [2];
            });
        });
    };
    Auth.prototype.setCustomSignFunc = function (getTickFn) {
        this.oauthInstance.authApi.setCustomSignFunc(getTickFn);
    };
    Auth.prototype.signIn = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.authApi.signIn(params)];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState(params)];
                }
            });
        });
    };
    Auth.prototype.setPassword = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.setPassword(params)];
            });
        });
    };
    Auth.prototype.isUsernameRegistered = function (username) {
        return __awaiter(this, void 0, void 0, function () {
            var exist;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof username !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'username must be a string');
                        }
                        return [4, this.oauthInstance.authApi.checkIfUserExist({ username: username })];
                    case 1:
                        exist = (_a.sent()).exist;
                        return [2, exist];
                }
            });
        });
    };
    Auth.prototype.hasLoginState = function () {
        var _a;
        if (this.cache.mode === 'async') {
            (0, utilities_1.printWarn)(utilities_1.ERRORS.INVALID_OPERATION, 'current platform\'s storage is asynchronous, please use getLoginState instead');
            return;
        }
        var oauthLoginState = (_a = this.oauthInstance) === null || _a === void 0 ? void 0 : _a.authApi.hasLoginStateSync();
        if (oauthLoginState) {
            var loginState = new LoginState({
                envId: this.config.env,
                cache: this.cache,
                oauthInstance: this.oauthInstance,
            });
            return loginState;
        }
        return null;
    };
    Auth.prototype.getLoginState = function () {
        return __awaiter(this, void 0, void 0, function () {
            var oauthLoginState, error_41, loginState;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        oauthLoginState = null;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4, this.oauthInstance.authApi.getLoginState()];
                    case 2:
                        oauthLoginState = _a.sent();
                        return [3, 4];
                    case 3:
                        error_41 = _a.sent();
                        return [2, null];
                    case 4:
                        if (oauthLoginState) {
                            loginState = new LoginState({
                                envId: this.config.env,
                                cache: this.cache,
                                oauthInstance: this.oauthInstance,
                            });
                            return [2, loginState];
                        }
                        return [2, null];
                }
            });
        });
    };
    Auth.prototype.getUserInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.getCurrentUser()];
            });
        });
    };
    Auth.prototype.getWedaUserInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.getWedaUserInfo()];
            });
        });
    };
    Auth.prototype.updateUserBasicInfo = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var loginState;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getLoginState()];
                    case 1:
                        loginState = _a.sent();
                        if (!loginState) return [3, 3];
                        return [4, loginState.user.updateUserBasicInfo(params)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2];
                }
            });
        });
    };
    Auth.prototype.getAuthHeader = function () {
        console.error('Auth.getAuthHeader API 已废弃');
        return {};
    };
    Auth.prototype.bindWithProvider = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.bindWithProvider(params)];
            });
        });
    };
    Auth.prototype.queryUser = function (queryObj) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.queryUserProfile(queryObj)];
            });
        });
    };
    Auth.prototype.getAccessToken = function () {
        return __awaiter(this, void 0, void 0, function () {
            var oauthAccessTokenRes;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.oauth2client.getAccessToken()];
                    case 1:
                        oauthAccessTokenRes = _a.sent();
                        return [2, {
                                accessToken: oauthAccessTokenRes,
                                env: this.config.env,
                            }];
                }
            });
        });
    };
    Auth.prototype.grantProviderToken = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.grantProviderToken(params)];
            });
        });
    };
    Auth.prototype.patchProviderToken = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.patchProviderToken(params)];
            });
        });
    };
    Auth.prototype.signInWithProvider = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.authApi.signInWithProvider(params)];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState(params)];
                }
            });
        });
    };
    Auth.prototype.signInWithWechat = function (params) {
        if (params === void 0) { params = {}; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.authApi.signInWithWechat(params)];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState(params)];
                }
            });
        });
    };
    Auth.prototype.grantToken = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.authApi.grantToken(params)];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState()];
                }
            });
        });
    };
    Auth.prototype.genProviderRedirectUri = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.genProviderRedirectUri(params)];
            });
        });
    };
    Auth.prototype.resetPassword = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.resetPassword(params)];
            });
        });
    };
    Auth.prototype.deviceAuthorize = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.deviceAuthorize(params)];
            });
        });
    };
    Auth.prototype.sudo = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.sudo(params)];
            });
        });
    };
    Auth.prototype.deleteMe = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.deleteMe(params)];
            });
        });
    };
    Auth.prototype.getProviders = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.getProviders()];
            });
        });
    };
    Auth.prototype.loginScope = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.loginScope()];
            });
        });
    };
    Auth.prototype.loginGroups = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.loginGroups()];
            });
        });
    };
    Auth.prototype.onLoginStateChanged = function (callback) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var loginState;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.on(oauth_1.EVENTS.LOGIN_STATE_CHANGED, function (params) { return __awaiter(_this, void 0, void 0, function () {
                            var loginState, _a;
                            var _b;
                            return __generator(this, function (_c) {
                                switch (_c.label) {
                                    case 0:
                                        if (!(((_b = params === null || params === void 0 ? void 0 : params.data) === null || _b === void 0 ? void 0 : _b.eventType) !== oauth_1.LOGIN_STATE_CHANGED_TYPE.CREDENTIALS_ERROR)) return [3, 2];
                                        return [4, this.getLoginState()];
                                    case 1:
                                        _a = _c.sent();
                                        return [3, 3];
                                    case 2:
                                        _a = {};
                                        _c.label = 3;
                                    case 3:
                                        loginState = _a;
                                        callback.call(this, __assign(__assign({}, params), loginState));
                                        return [2];
                                }
                            });
                        }); });
                        return [4, this.getLoginState()];
                    case 1:
                        loginState = _b.sent();
                        callback.call(this, loginState);
                        return [2];
                }
            });
        });
    };
    Auth.prototype.refreshTokenForce = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.refreshTokenForce(params)];
            });
        });
    };
    Auth.prototype.getCredentials = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.getCredentials()];
            });
        });
    };
    Auth.prototype.setCredentials = function (credentials) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.oauthInstance.oauth2client.setCredentials(credentials)];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    Auth.prototype.getProviderSubType = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.getProviderSubType()];
            });
        });
    };
    Auth.prototype.createCaptchaData = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.createCaptchaData(params)];
            });
        });
    };
    Auth.prototype.verifyCaptchaData = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.verifyCaptchaData(params)];
            });
        });
    };
    Auth.prototype.getMiniProgramQrCode = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.getMiniProgramCode(params)];
            });
        });
    };
    Auth.prototype.getMiniProgramQrCodeStatus = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.getMiniProgramQrCodeStatus(params)];
            });
        });
    };
    Auth.prototype.modifyPassword = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.modifyPassword(params)];
            });
        });
    };
    Auth.prototype.modifyPasswordWithoutLogin = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.modifyPasswordWithoutLogin(params)];
            });
        });
    };
    Auth.prototype.getUserBehaviorLog = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.oauthInstance.authApi.getUserBehaviorLog(params)];
            });
        });
    };
    Auth.prototype.signInWithUsername = function (_a) {
        var _b = _a.verificationInfo, verificationInfo = _b === void 0 ? { verification_id: '', is_user: false } : _b, _c = _a.verificationCode, verificationCode = _c === void 0 ? '' : _c, _d = _a.username, rawUsername = _d === void 0 ? '' : _d, _e = _a.bindInfo, bindInfo = _e === void 0 ? undefined : _e, _f = _a.loginType, loginType = _f === void 0 ? '' : _f;
        return __awaiter(this, void 0, void 0, function () {
            var verifyRes, verification_token, username, signUpParam, signInRes, bindRes, signUpRes, error_42;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        _g.trys.push([0, 8, , 9]);
                        return [4, this.oauthInstance.authApi.verify({
                                verification_id: verificationInfo.verification_id,
                                verification_code: verificationCode,
                            })];
                    case 1:
                        verifyRes = _g.sent();
                        if (verifyRes === null || verifyRes === void 0 ? void 0 : verifyRes.error_code) {
                            throw verifyRes;
                        }
                        verification_token = verifyRes.verification_token;
                        username = this.formatPhone(rawUsername);
                        signUpParam = { phone_number: username };
                        if (loginType === 'email') {
                            username = rawUsername;
                            signUpParam = { email: username };
                        }
                        if (!verificationInfo.is_user) return [3, 5];
                        return [4, this.oauthInstance.authApi.signIn({
                                username: username,
                                verification_token: verification_token,
                            })];
                    case 2:
                        signInRes = _g.sent();
                        if (signInRes === null || signInRes === void 0 ? void 0 : signInRes.error_code) {
                            throw signInRes;
                        }
                        if (!bindInfo) return [3, 4];
                        return [4, this.oauthInstance.authApi.bindWithProvider({
                                provider_token: bindInfo === null || bindInfo === void 0 ? void 0 : bindInfo.providerToken,
                            })];
                    case 3:
                        bindRes = _g.sent();
                        if (bindRes === null || bindRes === void 0 ? void 0 : bindRes.error_code) {
                            throw bindRes;
                        }
                        _g.label = 4;
                    case 4: return [3, 7];
                    case 5: return [4, this.oauthInstance.authApi.signUp(__assign(__assign({}, signUpParam), { verification_token: verification_token, provider_token: bindInfo === null || bindInfo === void 0 ? void 0 : bindInfo.providerId }))];
                    case 6:
                        signUpRes = _g.sent();
                        if (signUpRes === null || signUpRes === void 0 ? void 0 : signUpRes.error_code) {
                            throw signUpRes;
                        }
                        _g.label = 7;
                    case 7: return [2, this.createLoginState()];
                    case 8:
                        error_42 = _g.sent();
                        throw error_42;
                    case 9: return [2];
                }
            });
        });
    };
    Auth.prototype.createLoginState = function (params, options) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var loginState;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        loginState = new LoginState({
                            envId: this.config.env,
                            cache: this.cache,
                            oauthInstance: this.oauthInstance,
                        });
                        return [4, loginState.checkLocalStateAsync()];
                    case 1:
                        _c.sent();
                        if (!(options === null || options === void 0 ? void 0 : options.userInfo)) return [3, 2];
                        loginState.user.setLocalUserInfo(options.userInfo);
                        return [3, 5];
                    case 2:
                        if (!(options === null || options === void 0 ? void 0 : options.asyncRefreshUser)) return [3, 3];
                        loginState.user.refresh(params);
                        return [3, 5];
                    case 3: return [4, loginState.user.refresh(params)];
                    case 4:
                        _c.sent();
                        _c.label = 5;
                    case 5:
                        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.LOGIN_STATE_CHANGED, { eventType: oauth_1.LOGIN_STATE_CHANGED_TYPE.SIGN_IN });
                        (_b = this.config.eventBus) === null || _b === void 0 ? void 0 : _b.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_IN });
                        return [2, loginState];
                }
            });
        });
    };
    Auth.prototype.setAccessKey = function () {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var accessToken, scope;
            return __generator(this, function (_c) {
                accessToken = '';
                scope = '';
                if (this.config.accessKey) {
                    accessToken = this.config.accessKey;
                    scope = 'accessKey';
                }
                else if (((_a = this.config.auth) === null || _a === void 0 ? void 0 : _a.secretId) && ((_b = this.config.auth) === null || _b === void 0 ? void 0 : _b.secretKey)) {
                    accessToken = '';
                    scope = oauth_1.DEFAULT_NODE_ACCESS_SCOPE;
                }
                if (!scope) {
                    return [2];
                }
                try {
                    this.oauthInstance.oauth2client.setAccessKeyCredentials({
                        access_token: accessToken,
                        token_type: 'Bearer',
                        scope: scope,
                        expires_at: new Date(+new Date() + +new Date()),
                        expires_in: +new Date() + +new Date(),
                    });
                }
                catch (error) {
                    console.warn('accessKey error: ', error);
                }
                return [2];
            });
        });
    };
    Auth.prototype.formatPhone = function (phone) {
        if (!/\s+/.test(phone) && /^\+\d{1,3}\d+/.test(phone)) {
            return phone.replace(/^(\+\d{1,2})(\d+)$/, '$1 $2');
        }
        return /^\+\d{1,3}\s+/.test(phone) ? phone : "+86 ".concat(phone);
    };
    Auth.prototype.notifyListeners = function (event, session, info) {
        this.listeners.forEach(function (callbacks) {
            callbacks.forEach(function (callback) {
                try {
                    callback(event, session, info);
                }
                catch (error) {
                    console.error('Error in auth state change callback:', error);
                }
            });
        });
        return;
    };
    Auth.prototype.setupListeners = function () {
        var _this = this;
        var _a;
        (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.on(oauth_1.EVENTS.AUTH_STATE_CHANGED, function (params) { return __awaiter(_this, void 0, void 0, function () {
            var event, info, session;
            var _a, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        event = (_a = params === null || params === void 0 ? void 0 : params.data) === null || _a === void 0 ? void 0 : _a.event;
                        info = (_b = params === null || params === void 0 ? void 0 : params.data) === null || _b === void 0 ? void 0 : _b.info;
                        return [4, this.getSession()];
                    case 1:
                        session = (_c.sent()).data.session;
                        this.notifyListeners(event, session, info);
                        return [2];
                }
            });
        }); });
    };
    Auth.prototype.convertToUser = function (userInfo) {
        var _a, _b, _c, _d, _e, _f;
        if (!userInfo)
            return null;
        var email = (userInfo === null || userInfo === void 0 ? void 0 : userInfo.email) || '';
        var phone = (userInfo === null || userInfo === void 0 ? void 0 : userInfo.phone_number) || '';
        var userId = (userInfo === null || userInfo === void 0 ? void 0 : userInfo.sub) || (userInfo === null || userInfo === void 0 ? void 0 : userInfo.uid) || '';
        return {
            id: userId,
            aud: 'authenticated',
            role: (_b = (_a = userInfo.groups) === null || _a === void 0 ? void 0 : _a.map) === null || _b === void 0 ? void 0 : _b.call(_a, function (group) { return (typeof group === 'string' ? group : group.id); }),
            email: email || '',
            email_confirmed_at: (userInfo === null || userInfo === void 0 ? void 0 : userInfo.email_verified) ? userInfo.created_at : userInfo.created_at,
            phone: phone,
            phone_confirmed_at: phone ? userInfo.created_at : undefined,
            confirmed_at: userInfo.created_at,
            last_sign_in_at: userInfo.last_sign_in_at,
            app_metadata: {
                provider: ((_c = userInfo.loginType) === null || _c === void 0 ? void 0 : _c.toLowerCase()) || 'cloudbase',
                providers: [((_d = userInfo.loginType) === null || _d === void 0 ? void 0 : _d.toLowerCase()) || 'cloudbase'],
            },
            user_metadata: {
                name: userInfo === null || userInfo === void 0 ? void 0 : userInfo.name,
                picture: userInfo === null || userInfo === void 0 ? void 0 : userInfo.picture,
                username: userInfo === null || userInfo === void 0 ? void 0 : userInfo.username,
                gender: userInfo === null || userInfo === void 0 ? void 0 : userInfo.gender,
                locale: userInfo === null || userInfo === void 0 ? void 0 : userInfo.locale,
                uid: userInfo.uid,
                nickName: userInfo.nickName || (userInfo === null || userInfo === void 0 ? void 0 : userInfo.name),
                avatarUrl: userInfo.avatarUrl || userInfo.picture,
                location: userInfo.location,
                hasPassword: (_e = userInfo.hasPassword) !== null && _e !== void 0 ? _e : (userInfo.password === 'SET'),
                open_id: userInfo === null || userInfo === void 0 ? void 0 : userInfo.open_id,
            },
            identities: ((_f = userInfo === null || userInfo === void 0 ? void 0 : userInfo.providers) === null || _f === void 0 ? void 0 : _f.map(function (p) { return ({
                id: p.id || '',
                identity_id: p.id || '',
                user_id: userId,
                identity_data: {
                    provider_id: p.id,
                    provider_user_id: p.provider_user_id,
                    name: p.name,
                },
                provider: p.id || 'cloudbase',
                created_at: userInfo.created_at,
                updated_at: userInfo.updated_at,
                last_sign_in_at: userInfo.last_sign_in_at,
            }); })) || [],
            created_at: userInfo.created_at,
            updated_at: userInfo.updated_at,
            is_anonymous: userInfo.name === 'anonymous',
        };
    };
    Auth.prototype.validateParams = function (params, rules) {
        for (var _i = 0, _a = Object.entries(rules); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], rule = _b[1];
            if (rule.required && ((params === null || params === void 0 ? void 0 : params[key]) === undefined || (params === null || params === void 0 ? void 0 : params[key]) === null || (params === null || params === void 0 ? void 0 : params[key]) === '')) {
                throw new oauth_1.AuthError({ message: rule.message });
            }
        }
    };
    Auth.prototype.validateAtLeastOne = function (params, fieldGroups, message) {
        var hasValue = fieldGroups.some(function (group) { return group.some(function (field) { return (params === null || params === void 0 ? void 0 : params[field]) !== undefined && (params === null || params === void 0 ? void 0 : params[field]) !== null && (params === null || params === void 0 ? void 0 : params[field]) !== ''; }); });
        if (!hasValue) {
            throw new oauth_1.AuthError({ message: message });
        }
    };
    Auth.prototype.init = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var credentials, error_43;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        return [4, this.oauthInstance.oauth2client.localCredentials.getCredentials()];
                    case 1:
                        credentials = _b.sent();
                        if (credentials) {
                            (_a = this.config.eventBus) === null || _a === void 0 ? void 0 : _a.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, { event: oauth_1.AUTH_STATE_CHANGED_TYPE.INITIAL_SESSION });
                        }
                        return [3, 3];
                    case 2:
                        error_43 = _b.sent();
                        return [3, 3];
                    case 3: return [2, { error: null }];
                }
            });
        });
    };
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '绑定手机号失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().bindPhoneNumber() 的语法或参数是否正确',
                '  2 - 当前环境是否开通了短信验证码登录',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "bindPhoneNumber", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '解除三方绑定失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().unbindProvider() 的语法或参数是否正确',
                '  2 - 当前账户是否已经与此登录方式解绑',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "unbindProvider", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '绑定邮箱地址失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().bindEmail() 的语法或参数是否正确',
                '  2 - 当前环境是否开通了邮箱密码登录',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], Auth.prototype, "bindEmail", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '验证码验证失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().verify() 的语法或参数是否正确',
                '  2 - 当前环境是否开通了手机验证码/邮箱登录',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "verify", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '获取验证码失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().getVerification() 的语法或参数是否正确',
                '  2 - 当前环境是否开通了手机验证码/邮箱登录',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object, Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "getVerification", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '获取用户信息失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().getCurrentUser() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "getCurrentUser", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '小程序匿名登录失败',
            messages: [
                '请确认以下各项：',
                '  1 - 当前环境是否开启了匿名登录',
                '  2 - 调用 auth().signInAnonymouslyInWx() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "signInAnonymouslyInWx", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '小程序绑定OpenID失败',
            messages: [
                '请确认以下各项：',
                '  1 - 当前环境是否开启了小程序openId静默登录',
                '  2 - 调用 auth().bindOpenId() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "bindOpenId", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '小程序unionId静默登录失败',
            messages: [
                '请确认以下各项：',
                '  1 - 当前环境是否开启了小程序unionId静默登录',
                '  2 - 调用 auth().signInWithUnionId() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "signInWithUnionId", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '短信验证码登陆',
            messages: [
                '请确认以下各项：',
                '  1 - 当前环境是否开启了小程序短信验证码登陆',
                '  2 - 调用 auth().signInWithSms() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "signInWithSms", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '邮箱验证码登陆',
            messages: [
                '请确认以下各项：',
                '  1 - 当前环境是否开启了邮箱登陆',
                '  2 - 调用 auth().signInWithEmail() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "signInWithEmail", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '获取用户是否被占用失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().isUsernameRegistered() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "isUsernameRegistered", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '获取本地登录态失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().getLoginState() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "getLoginState", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '获取用户信息失败',
            messages: [
                '请确认以下各项：',
                '  1 - 是否已登录',
                '  2 - 调用 auth().getUserInfo() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "getUserInfo", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '获取微搭插件用户信息失败',
            messages: [
                '请确认以下各项：',
                '  1 - 是否已登录',
                '  2 - 调用 auth().getWedaUserInfo() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "getWedaUserInfo", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '绑定第三方登录方式失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().bindWithProvider() 的语法或参数是否正确',
                '  2 - 此账户是否已经绑定此第三方',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "bindWithProvider", null);
    __decorate([
        (0, utilities_1.catchErrorsDecorator)({
            title: '获取身份源类型',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().getProviderSubType() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(utilities_1.COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Auth.prototype, "getProviderSubType", null);
    return Auth;
}(v1_compat_1.AuthV1Compat));
exports.Auth = Auth;
function generateAuthInstance(config, options) {
    var _this = this;
    var _a, _b, _c, _d;
    var _e = config.region, region = _e === void 0 ? 'ap-shanghai' : _e, i18n = config.i18n, accessKey = config.accessKey, useWxCloud = config.useWxCloud;
    var platform = (options === null || options === void 0 ? void 0 : options.platform) || utilities_1.useDefaultAdapter.bind(options)();
    var runtime = platform.runtime, adapter = platform.adapter;
    var _f = options || {}, env = _f.env, clientId = _f.clientId, debug = _f.debug, cache = _f.cache, cloudbase = _f.app;
    var apiOrigin = (options || {}).apiOrigin;
    if (!apiOrigin) {
        apiOrigin = "https://".concat(env, ".").concat(region, ".tcb-api.tencentcloudapi.com");
    }
    var commonOpts = {
        env: env,
        clientId: clientId,
        i18n: i18n,
        accessKey: accessKey,
        useWxCloud: useWxCloud,
        eventBus: new utilities_1.CloudbaseEventEmitter(),
    };
    var oauthInstance = new oauth_1.CloudbaseOAuth((0, adapter_1.useAuthAdapter)(__assign(__assign({}, commonOpts), { apiOrigin: apiOrigin, apiPath: (config === null || config === void 0 ? void 0 : config.apiPath) || oauth_1.AUTH_API_PREFIX, storage: config === null || config === void 0 ? void 0 : config.storage, baseRequest: config === null || config === void 0 ? void 0 : config.baseRequest, request: config === null || config === void 0 ? void 0 : config.request, anonymousSignInFunc: config === null || config === void 0 ? void 0 : config.anonymousSignInFunc, captchaOptions: config === null || config === void 0 ? void 0 : config.captchaOptions, wxCloud: config === null || config === void 0 ? void 0 : config.wxCloud, adapter: adapter, onCredentialsError: onCredentialsError(commonOpts.eventBus), headers: __assign({ 'X-SDK-Version': "@cloudbase/js-sdk/".concat(config.sdkVersion) }, (config.headers || {})), detectSessionInUrl: config.detectSessionInUrl, debug: debug, auth: (_b = (_a = options.app) === null || _a === void 0 ? void 0 : _a.config) === null || _b === void 0 ? void 0 : _b.auth })));
    var authInstance = new Auth(__assign(__assign({}, commonOpts), { region: region, persistence: config.persistence, debug: debug, cache: cache || new utilities_1.CloudbaseCache({
            persistence: config.persistence,
            keys: { userInfoKey: "user_info_".concat(env) },
            platformInfo: platform,
        }), runtime: runtime || 'web', _fromApp: cloudbase, oauthInstance: oauthInstance, auth: (_d = (_c = options.app) === null || _c === void 0 ? void 0 : _c.config) === null || _d === void 0 ? void 0 : _d.auth }));
    oauthInstance.initializeSession(function (data, error) { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            if (!data)
                return [2];
            if (data.type === oauth_1.OAUTH_TYPE.SIGN_IN) {
                if (error) {
                    commonOpts.eventBus.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, {
                        event: oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_IN,
                        info: __assign(__assign({}, data), { error: error }),
                    });
                }
                else if (data.user) {
                    authInstance.createLoginState({}, { userInfo: data.user });
                }
            }
            else if (data.type === oauth_1.OAUTH_TYPE.BIND_IDENTITY) {
                commonOpts.eventBus.fire(oauth_1.EVENTS.AUTH_STATE_CHANGED, {
                    event: oauth_1.AUTH_STATE_CHANGED_TYPE.BIND_IDENTITY,
                    info: __assign(__assign({}, data), { error: error }),
                });
            }
            return [2];
        });
    }); });
    return { authInstance: authInstance, oauthInstance: oauthInstance };
}
exports.generateAuthInstance = generateAuthInstance;
var NAMESPACE = 'auth';
function levenshteinDistance(a, b) {
    var matrix = [];
    for (var i = 0; i <= a.length; i++) {
        matrix[i] = [i];
    }
    for (var j = 0; j <= b.length; j++) {
        matrix[0][j] = j;
    }
    for (var i = 1; i <= a.length; i++) {
        for (var j = 1; j <= b.length; j++) {
            var cost = a[i - 1].toLowerCase() === b[j - 1].toLowerCase() ? 0 : 1;
            matrix[i][j] = Math.min(matrix[i - 1][j] + 1, matrix[i][j - 1] + 1, matrix[i - 1][j - 1] + cost);
        }
    }
    return matrix[a.length][b.length];
}
function findSimilarMethod(target, methods) {
    var bestMatch = null;
    var bestDistance = Infinity;
    for (var _i = 0, methods_1 = methods; _i < methods_1.length; _i++) {
        var method = methods_1[_i];
        if (method.startsWith('_') || method === 'constructor')
            continue;
        var distance = levenshteinDistance(target, method);
        if (distance < bestDistance) {
            bestDistance = distance;
            bestMatch = method;
        }
    }
    return bestMatch;
}
var component = {
    name: COMPONENT_NAME,
    namespace: NAMESPACE,
    entity: function (config) {
        var auth = function (config) {
            var _a;
            if (this.authInstance && !config) {
                return this.authInstance;
            }
            config = config || {
                region: '',
                persistence: 'local',
                apiPath: oauth_1.AUTH_API_PREFIX,
            };
            var adapter = this.platform.adapter;
            var newPersistence = config.persistence || adapter.primaryStorage;
            if (newPersistence && newPersistence !== this.config.persistence) {
                this.updateConfig({ persistence: newPersistence });
            }
            var _b = generateAuthInstance(__assign(__assign({ wxCloud: this.config.wxCloud, storage: this.config.storage }, config), { persistence: this.config.persistence, i18n: this.config.i18n, accessKey: this.config.accessKey, useWxCloud: this.config.useWxCloud, sdkVersion: this.version, detectSessionInUrl: (_a = this.config.auth) === null || _a === void 0 ? void 0 : _a.detectSessionInUrl }), {
                env: this.config.env,
                clientId: this.config.clientId,
                apiOrigin: this.request.getBaseEndPoint(this.config.endPointMode || 'CLOUD_API'),
                platform: this.platform,
                cache: this.cache,
                app: this,
                debug: this.config.debug,
            }), authInstance = _b.authInstance, oauthInstance = _b.oauthInstance;
            this.oauthInstance = oauthInstance;
            this.authInstance = authInstance;
            return this.authInstance;
        };
        var authProto = auth.call(this, config);
        Object.assign(auth, authProto);
        Object.setPrototypeOf(auth, Object.getPrototypeOf(authProto));
        var authWithHint = typeof Proxy !== 'undefined'
            ? new Proxy(auth, {
                get: function (target, prop, receiver) {
                    var value = Reflect.get(target, prop, receiver);
                    if (value !== undefined || typeof prop !== 'string') {
                        return value;
                    }
                    var allKeys = Object.getOwnPropertyNames(Object.getPrototypeOf(target))
                        .concat(Object.keys(target))
                        .filter(function (k) { return typeof target[k] === 'function'; });
                    var suggestion = findSimilarMethod(prop, allKeys);
                    if (suggestion) {
                        console.warn("[CloudBase Auth] auth.".concat(prop, " is not a function. Did you mean: auth.").concat(suggestion, "() ?"));
                    }
                    return value;
                },
            })
            : auth;
        this[NAMESPACE] = authWithHint;
        return authWithHint;
    },
};
try {
    cloudbase.registerComponent(component);
}
catch (e) { }
var oauth_2 = require("@cloudbase/oauth");
Object.defineProperty(exports, "AuthErrorCategory", { enumerable: true, get: function () { return oauth_2.AuthErrorCategory; } });
Object.defineProperty(exports, "AuthError", { enumerable: true, get: function () { return oauth_2.AuthError; } });
function registerAuth(app) {
    try {
        app.registerComponent(component);
    }
    catch (e) {
        console.warn(e);
    }
}
exports.registerAuth = registerAuth;
