// @cloudbase/auth/dist/cjs/utilities.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useDefaultAdapter = exports.adapterForWxMp = exports.CloudbaseCache = exports.CloudbaseEventEmitter = exports.catchErrorsDecorator = exports.COMMUNITY_SITE_URL = exports.ERRORS = exports.throwError = exports.printWarn = void 0;
var util_1 = require("@cloudbase/utilities/dist/cjs/libs/util");
Object.defineProperty(exports, "printWarn", { enumerable: true, get: function () { return util_1.printWarn; } });
Object.defineProperty(exports, "throwError", { enumerable: true, get: function () { return util_1.throwError; } });
var constants_1 = require("@cloudbase/utilities/dist/cjs/constants");
Object.defineProperty(exports, "ERRORS", { enumerable: true, get: function () { return constants_1.ERRORS; } });
Object.defineProperty(exports, "COMMUNITY_SITE_URL", { enumerable: true, get: function () { return constants_1.COMMUNITY_SITE_URL; } });
var helpers_1 = require("@cloudbase/utilities/dist/cjs/helpers");
Object.defineProperty(exports, "catchErrorsDecorator", { enumerable: true, get: function () { return helpers_1.catchErrorsDecorator; } });
var events_1 = require("@cloudbase/utilities/dist/cjs/libs/events");
Object.defineProperty(exports, "CloudbaseEventEmitter", { enumerable: true, get: function () { return events_1.CloudbaseEventEmitter; } });
var cache_1 = require("@cloudbase/utilities/dist/cjs/libs/cache");
Object.defineProperty(exports, "CloudbaseCache", { enumerable: true, get: function () { return cache_1.CloudbaseCache; } });
var adapter_wx_mp_1 = require("@cloudbase/adapter-wx_mp");
Object.defineProperty(exports, "adapterForWxMp", { enumerable: true, get: function () { return __importDefault(adapter_wx_mp_1).default; } });
var adapters_1 = require("@cloudbase/utilities/dist/cjs/adapters");
Object.defineProperty(exports, "useDefaultAdapter", { enumerable: true, get: function () { return adapters_1.useDefaultAdapter; } });
