// @cloudbase/auth/dist/cjs/v1-compat.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyUserV1Compat = exports.AuthV1Compat = exports.AnonymousAuthProvider = exports.CustomAuthProvider = exports.WeixinAuthProvider = void 0;
var oauth_1 = require("@cloudbase/oauth");
var utilities_1 = require("./utilities");
var WeixinAuthProvider = (function () {
    function WeixinAuthProvider(_options) {
    }
    WeixinAuthProvider.prototype.signInWithRedirect = function () {
        throw new Error('[v1 兼容] WeixinAuthProvider.signInWithRedirect() 在当前版本中无法实现。'
            + ' 建议使用 auth.signInWithOAuth({ provider: \'providerId\' }) 替代。');
    };
    WeixinAuthProvider.prototype.getRedirectResult = function (_options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('[v1 兼容] WeixinAuthProvider.getRedirectResult() 在当前版本中无法实现。'
                    + ' 建议使用 auth.verifyOAuth({ code, state, provider }) 替代。');
            });
        });
    };
    WeixinAuthProvider.prototype.getLinkRedirectResult = function (_options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('[v1 兼容] WeixinAuthProvider.getLinkRedirectResult() 在当前版本中无法实现。'
                    + ' 建议使用 auth.linkIdentity({ provider: \'providerId\' }) 替代。');
            });
        });
    };
    return WeixinAuthProvider;
}());
exports.WeixinAuthProvider = WeixinAuthProvider;
var CustomAuthProvider = (function () {
    function CustomAuthProvider(options) {
        this.authInstance = options.authInstance;
    }
    CustomAuthProvider.prototype.signIn = function (ticket) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.authInstance.signInWithCustomTicket(function () { return Promise.resolve(ticket); })];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    return CustomAuthProvider;
}());
exports.CustomAuthProvider = CustomAuthProvider;
var AnonymousAuthProvider = (function () {
    function AnonymousAuthProvider(options) {
        this.authInstance = options.authInstance;
    }
    AnonymousAuthProvider.prototype.signIn = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.authInstance.signInAnonymously({})];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    return AnonymousAuthProvider;
}());
exports.AnonymousAuthProvider = AnonymousAuthProvider;
var AuthV1Compat = (function () {
    function AuthV1Compat() {
    }
    AuthV1Compat.prototype.weixinAuthProvider = function (options) {
        return new WeixinAuthProvider({
            authInstance: this,
            appid: options.appid,
            scope: options.scope,
        });
    };
    AuthV1Compat.prototype.customAuthProvider = function () {
        return new CustomAuthProvider({
            authInstance: this,
        });
    };
    AuthV1Compat.prototype.anonymousAuthProvider = function () {
        return new AnonymousAuthProvider({
            authInstance: this,
        });
    };
    AuthV1Compat.prototype.signUpWithEmailAndPassword = function (email, password) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof email !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'email must be a string');
                        }
                        if (typeof password !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'password must be a string');
                        }
                        return [4, this.signUp({
                                email: email,
                                password: password,
                            })];
                    case 1: return [2, _a.sent()];
                }
            });
        });
    };
    AuthV1Compat.prototype.signInWithEmailAndPassword = function (email, password) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof email !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'email must be a string');
                        }
                        if (typeof password !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'password must be a string');
                        }
                        return [4, this.signIn({
                                username: email,
                                password: password,
                            })];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState()];
                }
            });
        });
    };
    AuthV1Compat.prototype.sendPasswordResetEmail = function (email) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof email !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'email must be a string');
                        }
                        return [4, this.getVerification({ email: email })];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    AuthV1Compat.prototype.signInWithUsernameAndPassword = function (username, password) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof username !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'username must be a string');
                        }
                        if (typeof password !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'password must be a string');
                        }
                        return [4, this.signIn({
                                username: username,
                                password: password,
                            })];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState()];
                }
            });
        });
    };
    AuthV1Compat.prototype.sendPhoneCode = function (phoneNumber) {
        return __awaiter(this, void 0, void 0, function () {
            var formattedPhone;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof phoneNumber !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'phoneNumber must be a string');
                        }
                        formattedPhone = this.formatPhone(phoneNumber);
                        return [4, this.getVerification({ phone_number: formattedPhone })];
                    case 1:
                        _a.sent();
                        return [2, true];
                }
            });
        });
    };
    AuthV1Compat.prototype.signUpWithPhoneCode = function (phoneNumber, phoneCode, password) {
        return __awaiter(this, void 0, void 0, function () {
            var formattedPhone;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof phoneNumber !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'phoneNumber must be a string');
                        }
                        if (typeof phoneCode !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'phoneCode must be a string');
                        }
                        formattedPhone = this.formatPhone(phoneNumber);
                        return [4, this.signUp(__assign({ phone_number: formattedPhone, verification_code: phoneCode }, (password ? { password: password } : {})))];
                    case 1:
                        _a.sent();
                        return [2, this.createLoginState()];
                }
            });
        });
    };
    AuthV1Compat.prototype.signInWithPhoneCodeOrPassword = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var phoneNumber, phoneCode, password, formattedPhone;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        phoneNumber = params.phoneNumber, phoneCode = params.phoneCode, password = params.password;
                        if (typeof phoneNumber !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'phoneNumber must be a string');
                        }
                        formattedPhone = this.formatPhone(phoneNumber);
                        if (!password) return [3, 2];
                        return [4, this.signIn({
                                username: formattedPhone,
                                password: password,
                            })];
                    case 1:
                        _a.sent();
                        return [3, 5];
                    case 2:
                        if (!phoneCode) return [3, 4];
                        return [4, this.signIn({
                                username: formattedPhone,
                                verification_token: phoneCode,
                            })];
                    case 3:
                        _a.sent();
                        return [3, 5];
                    case 4:
                        (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'phoneCode or password must be provided');
                        _a.label = 5;
                    case 5: return [2, this.createLoginState()];
                }
            });
        });
    };
    AuthV1Compat.prototype.forceResetPwdByPhoneCode = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var phoneNumber, phoneCode, password, formattedPhone;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        phoneNumber = params.phoneNumber, phoneCode = params.phoneCode, password = params.password;
                        if (typeof phoneNumber !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'phoneNumber must be a string');
                        }
                        if (typeof phoneCode !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'phoneCode must be a string');
                        }
                        if (typeof password !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'password must be a string');
                        }
                        formattedPhone = this.formatPhone(phoneNumber);
                        return [4, this.resetPassword({
                                phone_number: formattedPhone,
                                new_password: password,
                                verification_token: phoneCode,
                            })];
                    case 1:
                        _a.sent();
                        return [4, this.signIn({
                                username: formattedPhone,
                                password: password,
                            })];
                    case 2:
                        _a.sent();
                        return [2, this.createLoginState()];
                }
            });
        });
    };
    AuthV1Compat.prototype.shouldRefreshAccessToken = function (_callback) {
        throw new Error('[v1 兼容] shouldRefreshAccessToken() 在当前版本中无法实现。'
            + ' 建议使用 auth.onAuthStateChange(callback) 来监听 TOKEN_REFRESHED 事件替代。');
    };
    AuthV1Compat.prototype.onLoginStateExpired = function (callback) {
        var _this = this;
        this.onLoginStateChanged(function (params) {
            if ((params === null || params === void 0 ? void 0 : params.eventType) === oauth_1.LOGIN_STATE_CHANGED_TYPE.CREDENTIALS_ERROR) {
                callback.call(_this);
            }
        });
    };
    AuthV1Compat.prototype.onAccessTokenRefreshed = function (callback) {
        var _this = this;
        this.onAuthStateChange(function (event) {
            if (event === oauth_1.AUTH_STATE_CHANGED_TYPE.TOKEN_REFRESHED) {
                callback.call(_this);
            }
        });
    };
    AuthV1Compat.prototype.onAnonymousConverted = function (callback) {
        var _this = this;
        this.onAuthStateChange(function (event) {
            if (event === oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_IN) {
                callback.call(_this);
            }
        });
    };
    AuthV1Compat.prototype.onLoginTypeChanged = function (callback) {
        var _this = this;
        this.onAuthStateChange(function (event) {
            if (event === oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_IN || event === oauth_1.AUTH_STATE_CHANGED_TYPE.SIGNED_OUT) {
                callback.call(_this);
            }
        });
    };
    return AuthV1Compat;
}());
exports.AuthV1Compat = AuthV1Compat;
function applyUserV1Compat(UserClass) {
    UserClass.prototype.linkWithTicket = function (_ticket) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('[v1 兼容] User.linkWithTicket() 在当前版本中无法实现。'
                    + ' 建议使用 auth.signInWithCustomTicket(() => Promise.resolve(ticket)) 替代。');
            });
        });
    };
    UserClass.prototype.linkWithRedirect = function (_provider) {
        throw new Error('[v1 兼容] User.linkWithRedirect() 在当前版本中无法实现。'
            + ' 建议使用 auth.linkIdentity({ provider: \'providerId\' }) 替代。');
    };
    UserClass.prototype.updateEmail = function (newEmail, _password) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (typeof newEmail !== 'string') {
                            (0, utilities_1.throwError)(utilities_1.ERRORS.INVALID_PARAMS, 'newEmail must be a string');
                        }
                        return [4, this.updateUserBasicInfo({
                                email: newEmail,
                            })];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    UserClass.prototype.linkWithPhoneNumber = function (_phoneNumber, _phoneCode) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('[v1 兼容] User.linkWithPhoneNumber() 在当前版本中无法实现。'
                    + ' 建议使用 auth.bindPhoneNumber({ phone_number, verification_token, sudo_token }) 替代。');
            });
        });
    };
    UserClass.prototype.updatePhoneNumber = function (_phoneNumber, _phoneCode) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('[v1 兼容] User.updatePhoneNumber() 在当前版本中无法实现。'
                    + ' 建议使用 auth.updateUser({ phone: newPhoneNumber }) 替代。');
            });
        });
    };
    UserClass.prototype.unlink = function (_loginType) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error('[v1 兼容] User.unlink() 在当前版本中无法实现。'
                    + ' 建议使用 auth.unlinkIdentity({ provider: loginType }) 替代。');
            });
        });
    };
}
exports.applyUserV1Compat = applyUserV1Compat;
