// @cloudbase/app/dist/esm/libs/adapter-node/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { StorageType, } from '@cloudbase/adapter-interface';
import { NodeRequest } from './request';
import { nodeTool } from './tool';
import { getSecretInfo } from './context';
import { parseQueryString } from './utils';
var _WS = null;
function getWS() {
    if (!_WS) {
        try {
            _WS = require('ws');
        }
        catch (e) {
            console.error('缺少依赖 ws，请执行以下命令安装：\n\n'
                + '  npm install ws\n\n'
                + '该依赖用于 Node 环境下的 WebSocket 连接。');
        }
    }
    return _WS;
}
export { NodeRequest } from './request';
export { parseQueryString, createWebStreamFromNodeReadableStream } from './utils';
function isMatch() {
    return typeof process !== 'undefined' && process.versions != null && process.versions.node != null;
}
var storage = (function () {
    var db = new Map();
    return {
        mode: 'sync',
        setItem: function (key, value) {
            db.set(key, value);
        },
        getItem: function (key) {
            return db.get(key);
        },
        removeItem: function (key) {
            db.delete(key);
        },
        clear: function () {
            db.clear();
        },
    };
})();
function genAdapter(options) {
    var adapter = {
        nodeTool: nodeTool,
        getSecretInfo: getSecretInfo,
        root: { globalThis: {} },
        reqClass: NodeRequest,
        wsClass: getWS(),
        sessionStorage: storage,
        localStorage: storage,
        primaryStorage: StorageType.session,
        captchaOptions: {
            openURIWithCallback: function (_url) {
                var EventBus = options.EventBus;
                var queryObj = {};
                var url = _url;
                var matched = _url.match(/^(data:.*?)(\?[^#\s]*)?$/);
                if (matched) {
                    url = matched[1];
                    var search = matched[2];
                    if (search) {
                        queryObj = parseQueryString(search);
                    }
                }
                var token = queryObj.token, restQueryObj = __rest(queryObj, ["token"]);
                if (/^data:/.test(url) && !token) {
                    return Promise.reject({
                        error: 'invalid_argument',
                        error_description: "invalie captcha data: ".concat(_url),
                    });
                }
                if (!token) {
                    return Promise.reject({
                        error: 'unimplemented',
                        error_description: 'need to impl captcha data',
                    });
                }
                return new Promise(function (resolve) {
                    EventBus.$emit('CAPTCHA_DATA_CHANGE', __assign(__assign({}, restQueryObj), { token: token, url: url }));
                    EventBus.$once('RESOLVE_CAPTCHA_DATA', function (res) {
                        resolve(res);
                    });
                });
            },
        },
    };
    return adapter;
}
var adapter = {
    genAdapter: genAdapter,
    isMatch: isMatch,
    runtime: 'node-adapter',
};
export default adapter;
