// @cloudbase/app/dist/esm/libs/adapter-node/metadata.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
export var kMetadataBaseUrl = 'http://metadata.tencentyun.com';
export var kAppIdPath = 'meta-data/app-id';
export var kSecurityCredentialsPath = 'meta-data/cam/security-credentials';
export var kMetadataVersions;
(function (kMetadataVersions) {
    kMetadataVersions["v20170919"] = "2017-09-19";
    kMetadataVersions["v1.0"] = "1.0";
    kMetadataVersions["latest"] = "latest";
})(kMetadataVersions || (kMetadataVersions = {}));
export function isAppId(appIdStr) {
    return /^[1-9][0-9]{4,64}$/gim.test(appIdStr);
}
export function lookup(path, options) {
    if (options === void 0) { options = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var url, controller, timer, resp;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    url = "".concat(kMetadataBaseUrl, "/").concat(kMetadataVersions.latest, "/").concat(path);
                    controller = new AbortController();
                    if (options.timeout) {
                        timer = setTimeout(function () { return controller.abort(); }, options.timeout);
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, , 5, 6]);
                    return [4, fetch(url, { signal: controller.signal })];
                case 2:
                    resp = _a.sent();
                    if (!(resp.status === 200)) return [3, 4];
                    return [4, resp.text()];
                case 3: return [2, _a.sent()];
                case 4: throw new Error("[ERROR] GET ".concat(url, " status: ").concat(resp.status));
                case 5:
                    if (timer)
                        clearTimeout(timer);
                    return [7];
                case 6: return [2];
            }
        });
    });
}
var metadataCache = {
    appId: undefined,
};
export function lookupAppId() {
    return __awaiter(this, void 0, void 0, function () {
        var appId, _e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!(metadataCache.appId === undefined)) return [3, 4];
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4, lookup(kAppIdPath, { timeout: 30 })];
                case 2:
                    appId = _a.sent();
                    if (isAppId(appId)) {
                        metadataCache.appId = appId;
                    }
                    else {
                        metadataCache.appId = '';
                    }
                    return [3, 4];
                case 3:
                    _e_1 = _a.sent();
                    return [3, 4];
                case 4: return [2, metadataCache.appId || ''];
            }
        });
    });
}
export function lookupCredentials(ruleName) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4, lookup("".concat(kSecurityCredentialsPath, "/").concat(ruleName), {
                        timeout: 200,
                    })];
                case 1: return [2, _a.sent()];
            }
        });
    });
}
