// @cloudbase/app/dist/cjs/libs/adapter-node/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SECRET_TYPE = exports.checkCustomUserIdRegex = exports.CLIENT_AUTH_PATH = exports.ADMIN_PATH = exports.ERROR = void 0;
exports.ERROR = {
    INVALID_PARAM: {
        code: 'INVALID_PARAM',
        message: 'invalid param',
    },
    SYS_ERR: {
        code: 'SYS_ERR',
        message: 'system error',
    },
    STORAGE_REQUEST_FAIL: {
        code: 'STORAGE_REQUEST_FAIL',
        message: 'storage request fail',
    },
    STORAGE_FILE_NONEXIST: {
        code: 'STORAGE_FILE_NONEXIST',
        message: 'storage file not exist',
    },
    TCB_CLS_UNOPEN: {
        code: 'TCB_CLS_UNOPEN',
        message: '需要先开通日志检索功能',
    },
    INVALID_CONTEXT: {
        code: 'INVALID_CONTEXT',
        message: '无效的context对象，请使用 云函数入口的context参数',
    },
};
exports.ADMIN_PATH = '/admin';
exports.CLIENT_AUTH_PATH = '/auth/v1/token/clientCredential';
exports.checkCustomUserIdRegex = /^[a-zA-Z0-9_\-#@~=*(){}[\]:.,<>+]{4,32}$/;
var SECRET_TYPE;
(function (SECRET_TYPE) {
    SECRET_TYPE["SECRET"] = "SECRET";
    SECRET_TYPE["SESSION_SECRET"] = "SESSION_SECRET";
})(SECRET_TYPE = exports.SECRET_TYPE || (exports.SECRET_TYPE = {}));
