// @cloudbase/app/dist/cjs/libs/adapter-node/context.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSecretInfo = exports.INIT_CONFIG = exports.getCloudbaseContext = exports.parseContext = void 0;
var constants_1 = require("./constants");
var utils_1 = require("./utils");
function parseContext(context) {
    if (typeof context !== 'object') {
        throw __assign(__assign({}, constants_1.ERROR.INVALID_CONTEXT), { message: 'context 必须为对象类型' });
    }
    var parseResult = {};
    var memory_limit_in_mb = context.memory_limit_in_mb, time_limit_in_ms = context.time_limit_in_ms, request_id = context.request_id, function_version = context.function_version, namespace = context.namespace, function_name = context.function_name, environ = context.environ, environment = context.environment;
    try {
        parseResult.memoryLimitInMb = memory_limit_in_mb;
        parseResult.timeLimitIns = time_limit_in_ms;
        parseResult.requestId = request_id;
        parseResult.functionVersion = function_version;
        parseResult.namespace = namespace;
        parseResult.functionName = function_name;
        if (environment) {
            parseResult.environment = JSON.parse(environment);
        }
        else {
            var parseEnviron = environ === null || environ === void 0 ? void 0 : environ.split(';');
            var parseEnvironObj = {};
            for (var i in parseEnviron) {
                if (parseEnviron[i].includes('=')) {
                    var equalIndex = parseEnviron[i].indexOf('=');
                    var key = parseEnviron[i].slice(0, equalIndex);
                    var value = parseEnviron[i].slice(equalIndex + 1);
                    if (value.indexOf(',') >= 0) {
                        value = value.split(',');
                    }
                    parseEnvironObj[key] = value;
                }
            }
            parseResult.environ = parseEnvironObj;
        }
    }
    catch (_err) {
        throw __assign({}, constants_1.ERROR.INVALID_CONTEXT);
    }
    return parseResult;
}
exports.parseContext = parseContext;
var getCloudbaseContext = function (context) {
    if ((0, utils_1.checkIsInScf)()) {
        if (!(0, utils_1.getEnv)('TENCENTCLOUD_REGION')) {
            console.error('[TCB][ERROR] missing `TENCENTCLOUD_REGION` environment');
        }
        if (!(0, utils_1.getEnv)('SCF_NAMESPACE')) {
            console.error('[TCB][ERROR] missing `SCF_NAMESPACE` environment');
        }
    }
    var envObj = (0, utils_1.getEnv)();
    var envFromProcessEnv = {
        TRIGGER_SRC: envObj.TRIGGER_SRC,
        _SCF_TCB_LOG: envObj._SCF_TCB_LOG,
        SCF_NAMESPACE: envObj.SCF_NAMESPACE,
        TENCENTCLOUD_RUNENV: envObj.TENCENTCLOUD_RUNENV,
        TENCENTCLOUD_SECRETID: envObj.TENCENTCLOUD_SECRETID,
        TENCENTCLOUD_SECRETKEY: envObj.TENCENTCLOUD_SECRETKEY,
        TENCENTCLOUD_SESSIONTOKEN: envObj.TENCENTCLOUD_SESSIONTOKEN,
        WX_CONTEXT_KEYS: envObj.WX_CONTEXT_KEYS,
        WX_TRIGGER_API_TOKEN_V0: envObj.WX_TRIGGER_API_TOKEN_V0,
        WX_CLIENTIP: envObj.WX_CLIENTIP,
        WX_CLIENTIPV6: envObj.WX_CLIENTIPV6,
        TCB_CONTEXT_KEYS: envObj.TCB_CONTEXT_KEYS,
        TCB_CONTEXT_CNFG: envObj.TCB_CONTEXT_CNFG,
        LOGINTYPE: envObj.LOGINTYPE,
    };
    var envFromContext = {};
    if (context) {
        var _a = parseContext(context), environment = _a.environment, environ = _a.environ;
        envFromContext = environment || environ || {};
    }
    var tcbContextKeys = envFromContext.TCB_CONTEXT_KEYS || envObj.TCB_CONTEXT_KEYS;
    var wxContextKeys = envFromContext.WX_CONTEXT_KEYS || envObj.WX_CONTEXT_KEYS;
    if (tcbContextKeys) {
        try {
            var tcbKeysList = tcbContextKeys.split(',');
            for (var _i = 0, tcbKeysList_1 = tcbKeysList; _i < tcbKeysList_1.length; _i++) {
                var item = tcbKeysList_1[_i];
                envFromProcessEnv[item] = envFromContext[item] || (0, utils_1.getEnv)(item);
            }
        }
        catch (_e) {
        }
    }
    if (wxContextKeys) {
        try {
            var wxKeysList = wxContextKeys.split(',');
            for (var _b = 0, wxKeysList_1 = wxKeysList; _b < wxKeysList_1.length; _b++) {
                var item = wxKeysList_1[_b];
                envFromProcessEnv[item] = envFromContext[item] || (0, utils_1.getEnv)(item);
            }
        }
        catch (_e) {
        }
    }
    var allContext = __assign(__assign({}, envFromProcessEnv), envFromContext);
    var finalContext = {};
    for (var key in allContext) {
        if (allContext[key] !== undefined) {
            finalContext[key] = allContext[key];
        }
    }
    return finalContext;
};
exports.getCloudbaseContext = getCloudbaseContext;
exports.INIT_CONFIG = {};
var getSecretInfo = function (config) {
    var _a, _b, _c, _d, _f, _g;
    exports.INIT_CONFIG = JSON.parse(JSON.stringify(config || {}));
    var _h = (0, exports.getCloudbaseContext)(), TCB_ENV = _h.TCB_ENV, SCF_NAMESPACE = _h.SCF_NAMESPACE;
    var env = (config === null || config === void 0 ? void 0 : config.env) || TCB_ENV || SCF_NAMESPACE || '';
    var defaultInfo = {
        env: env,
        secretId: '',
        secretKey: '',
        sessionToken: '',
        accessKey: '',
        secretType: '',
    };
    if (config === null || config === void 0 ? void 0 : config.accessKey) {
        return defaultInfo;
    }
    if (((_a = config === null || config === void 0 ? void 0 : config.auth) === null || _a === void 0 ? void 0 : _a.secretId) && ((_b = config === null || config === void 0 ? void 0 : config.auth) === null || _b === void 0 ? void 0 : _b.secretKey)) {
        return __assign(__assign({}, defaultInfo), { secretId: config.auth.secretId, secretKey: config.auth.secretKey, secretType: constants_1.SECRET_TYPE.SECRET });
    }
    var secretEnv = (0, utils_1.getEnv)();
    var TENCENTCLOUD_SECRETID = secretEnv.TENCENTCLOUD_SECRETID;
    var TENCENTCLOUD_SECRETKEY = secretEnv.TENCENTCLOUD_SECRETKEY;
    var TENCENTCLOUD_SESSIONTOKEN = secretEnv.TENCENTCLOUD_SESSIONTOKEN;
    var CLOUDBASE_APIKEY = secretEnv.CLOUDBASE_APIKEY;
    if (CLOUDBASE_APIKEY) {
        return __assign(__assign({}, defaultInfo), { accessKey: CLOUDBASE_APIKEY });
    }
    if (TENCENTCLOUD_SECRETID && TENCENTCLOUD_SECRETKEY) {
        return __assign(__assign({}, defaultInfo), { secretId: TENCENTCLOUD_SECRETID, secretKey: TENCENTCLOUD_SECRETKEY, sessionToken: TENCENTCLOUD_SESSIONTOKEN || '', secretType: constants_1.SECRET_TYPE.SESSION_SECRET });
    }
    if ((_c = config === null || config === void 0 ? void 0 : config.context) === null || _c === void 0 ? void 0 : _c.extendedContext) {
        var extendedContext = config.context.extendedContext;
        if (!env) {
            env = extendedContext.envId || '';
        }
        return __assign(__assign({}, defaultInfo), { secretId: ((_d = extendedContext === null || extendedContext === void 0 ? void 0 : extendedContext.tmpSecret) === null || _d === void 0 ? void 0 : _d.secretId) || '', secretKey: ((_f = extendedContext === null || extendedContext === void 0 ? void 0 : extendedContext.tmpSecret) === null || _f === void 0 ? void 0 : _f.secretKey) || '', sessionToken: ((_g = extendedContext === null || extendedContext === void 0 ? void 0 : extendedContext.tmpSecret) === null || _g === void 0 ? void 0 : _g.token) || '' });
    }
    return __assign({}, defaultInfo);
};
exports.getSecretInfo = getSecretInfo;
