// @cloudbase/app/dist/cjs/libs/adapter-node/request.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeRequest = void 0;
var adapter_interface_1 = require("@cloudbase/adapter-interface");
var constants_1 = require("./constants");
var context_1 = require("./context");
var utils_1 = require("./utils");
var common_1 = require("../../constants/common");
var signFn = null;
function getSign() {
    return __awaiter(this, void 0, void 0, function () {
        var mod, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!!signFn) return [3, 4];
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4, import('@cloudbase/signature-nodejs')];
                case 2:
                    mod = _a.sent();
                    signFn = mod.sign;
                    return [3, 4];
                case 3:
                    e_1 = _a.sent();
                    console.error('缺少依赖 @cloudbase/signature-nodejs，请执行以下命令安装：\n\n'
                        + '  npm install @cloudbase/signature-nodejs\n\n'
                        + '该依赖用于 Node 环境下的请求签名。');
                    return [3, 4];
                case 4: return [2, signFn];
            }
        });
    });
}
var NodeRequest = (function () {
    function NodeRequest(config) {
        var _this = this;
        this.getReqOptions = function (url, options) { return __awaiter(_this, void 0, void 0, function () {
            var urlObj, TCB_SOURCE, SOURCE, _a, _b, isAdminPath, UNSUPPORTED_V3_SIGN_PATH, token, _c, secretId, secretKey, sessionToken, secretType, method, body, headers, params, secretInfo, signedParams, sign, _d, authorization, timestamp;
            var _e, _f, _g, _h, _j, _k, _l, _m;
            return __generator(this, function (_o) {
                switch (_o.label) {
                    case 0:
                        urlObj = new URL(url);
                        TCB_SOURCE = (0, context_1.getCloudbaseContext)().TCB_SOURCE;
                        _b = (_a = "".concat(((_f = (_e = context_1.INIT_CONFIG.context) === null || _e === void 0 ? void 0 : _e.extendedContext) === null || _f === void 0 ? void 0 : _f.source) || TCB_SOURCE || '', ",")).concat;
                        return [4, (0, utils_1.getCurrRunEnvTag)()];
                    case 1:
                        SOURCE = _b.apply(_a, [_o.sent()]);
                        options.headers = __assign(__assign({}, options.headers), { 'User-Agent': "adapter-node/".concat(((_g = options.headers) === null || _g === void 0 ? void 0 : _g['X-SDK-Version']) || "@cloudbase/js-sdk/".concat((0, common_1.getSdkVersion)())), 'X-TCB-Source': SOURCE, 'X-Client-Timestamp': new Date().valueOf(), 'X-TCB-Region': context_1.INIT_CONFIG.region || (0, utils_1.getEnv)('TENCENTCLOUD_REGION') || '', Host: urlObj.host });
                        if ((_j = (_h = options.headers.Authorization) === null || _h === void 0 ? void 0 : _h.replace('Bearer', '')) === null || _j === void 0 ? void 0 : _j.trim()) {
                            return [2, options];
                        }
                        if (urlObj.pathname.startsWith('/auth/') && urlObj.pathname !== constants_1.CLIENT_AUTH_PATH) {
                            return [2, options];
                        }
                        isAdminPath = urlObj.pathname === constants_1.ADMIN_PATH;
                        UNSUPPORTED_V3_SIGN_PATH = ['/v1/model/', '/v1/rdb/', '/v1/ai/'];
                        if (!UNSUPPORTED_V3_SIGN_PATH.some(function (v) { return urlObj.pathname.startsWith(v); })) return [3, 3];
                        return [4, this.getClientCredential({ origin: urlObj.origin })];
                    case 2:
                        token = _o.sent();
                        options.headers.Authorization = "Bearer ".concat(token);
                        return [2, options];
                    case 3:
                        _c = ((_k = this.config) === null || _k === void 0 ? void 0 : _k.auth) || {}, secretId = _c.secretId, secretKey = _c.secretKey, sessionToken = _c.sessionToken;
                        secretType = (((_l = this.config) === null || _l === void 0 ? void 0 : _l.auth) || {}).secretType;
                        method = options.method, body = options.body;
                        headers = JSON.parse(JSON.stringify(options.headers));
                        params = typeof body === 'object' ? body : (0, utils_1.safeParseJson)(body) || {};
                        params.envName = (params === null || params === void 0 ? void 0 : params.env) || this.config.env || ((_m = (0, context_1.getSecretInfo)()) === null || _m === void 0 ? void 0 : _m.env) || '';
                        if (params.access_token) {
                            return [2, options];
                        }
                        if (method.toLowerCase() === 'get') {
                            params = undefined;
                        }
                        if (!secretId || !secretKey || secretType === 'SESSION_SECRET') {
                            secretInfo = (0, context_1.getSecretInfo)();
                            secretId = secretInfo.secretId;
                            secretKey = secretInfo.secretKey;
                            sessionToken = secretInfo.sessionToken;
                        }
                        if (!secretId || !secretKey) {
                            return [2, options];
                        }
                        delete headers.Authorization;
                        if (headers['content-type']) {
                            headers['content-type'] = headers['content-type'].toLowerCase();
                        }
                        else if (headers['Content-Type']) {
                            headers['Content-Type'] = headers['Content-Type'].toLowerCase();
                        }
                        if (isAdminPath && !!sessionToken) {
                            params.sessionToken = sessionToken;
                        }
                        signedParams = {
                            secretId: secretId,
                            secretKey: secretKey,
                            method: method,
                            url: url,
                            params: params,
                            headers: headers,
                            withSignedParams: isAdminPath,
                            timestamp: Math.floor(new Date().getTime() / 1000) - 1,
                            isCloudApi: !isAdminPath,
                        };
                        if (!headers['content-type'] && !headers['Content-Type']) {
                            return [2, options];
                        }
                        return [4, getSign()];
                    case 4:
                        sign = _o.sent();
                        if (!sign)
                            return [2, options];
                        _d = sign(signedParams), authorization = _d.authorization, timestamp = _d.timestamp;
                        if (typeof sessionToken === 'string' && sessionToken !== '') {
                            if (isAdminPath) {
                                headers.Authorization = "".concat(authorization);
                                headers['X-Timestamp'] = timestamp;
                                headers['X-Signature-Expires'] = 600;
                            }
                            else {
                                headers.Authorization = "".concat(authorization, ", Timestamp=").concat(timestamp, ", Token=").concat(sessionToken);
                            }
                        }
                        else {
                            headers.Authorization = "".concat(authorization, ", Timestamp=").concat(timestamp);
                            if (isAdminPath) {
                                headers['X-Timestamp'] = timestamp;
                                headers['X-Signature-Expires'] = 600;
                            }
                        }
                        return [2, __assign(__assign({}, options), { headers: headers, body: typeof params === 'object' ? JSON.stringify(params) : params })];
                }
            });
        }); };
        this.get = function (options) { return _this.request(__assign(__assign({}, options), { method: 'get' }), _this.restrictedMethods.includes('get')); };
        this.post = function (options) { return _this.request(__assign(__assign({}, options), { method: 'post' }), _this.restrictedMethods.includes('post')); };
        this.put = function (options) { return _this.request(__assign(__assign({}, options), { method: 'put' })); };
        this.upload = function (options) {
            var _a;
            var _data = options.data, file = options.file, name = options.name, method = options.method, _b = options.headers, headers = _b === void 0 ? {} : _b;
            if (file === undefined || name == undefined) {
                throw new Error('file and name is required');
            }
            var data = (0, utils_1.obj2StrRecord)(_data !== null && _data !== void 0 ? _data : {});
            var loweredMethod = method === null || method === void 0 ? void 0 : method.toLowerCase();
            var reqMethod = (_a = ['post', 'put'].find(function (m) { return m === loweredMethod; })) !== null && _a !== void 0 ? _a : 'put';
            var formData = new FormData();
            if (reqMethod === 'post') {
                Object.keys(data).forEach(function (key) {
                    formData.append(key, data[key]);
                });
                formData.append('key', name);
                formData.append('file', file);
                return _this.request(__assign(__assign({}, options), { data: formData, method: reqMethod }), _this.restrictedMethods.includes('upload'));
            }
            return _this.request(__assign(__assign({}, options), { method: 'put', headers: headers, body: file }), _this.restrictedMethods.includes('upload'));
        };
        this.download = function (options) { return __awaiter(_this, void 0, void 0, function () {
            var data, url, fileName, link;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4, this.get(__assign(__assign({}, options), { headers: {}, responseType: 'blob' }))];
                    case 1:
                        data = (_b.sent()).data;
                        url = window.URL.createObjectURL(new Blob([data]));
                        fileName = decodeURIComponent(new URL((_a = options === null || options === void 0 ? void 0 : options.url) !== null && _a !== void 0 ? _a : '').pathname.split('/').pop() || '');
                        link = document.createElement('a');
                        link.href = url;
                        link.setAttribute('download', fileName);
                        link.style.display = 'none';
                        document.body.appendChild(link);
                        link.click();
                        window.URL.revokeObjectURL(url);
                        document.body.removeChild(link);
                        return [2, new Promise(function (resolve) {
                                resolve({
                                    statusCode: 200,
                                    tempFilePath: options.url,
                                });
                            })];
                }
            });
        }); };
        this.fetch = function (options) { return __awaiter(_this, void 0, void 0, function () {
            var _a, enableAbort, _b, stream, signal, customReqOpts, url, abortController, timeout, timer, headers, fetchOptions, res, ret, _c, _d;
            var _e;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _a = options.enableAbort, enableAbort = _a === void 0 ? false : _a, _b = options.stream, stream = _b === void 0 ? false : _b, signal = options.signal, customReqOpts = options.customReqOpts;
                        url = this.getRealUrl(options.url, options.headers || {}, options.body);
                        abortController = new AbortController();
                        timeout = (customReqOpts === null || customReqOpts === void 0 ? void 0 : customReqOpts.timeout) || this.timeout;
                        if (signal) {
                            if (signal.aborted) {
                                abortController.abort();
                            }
                            else {
                                signal.addEventListener('abort', function () { return abortController.abort(); });
                            }
                        }
                        timer = undefined;
                        if (enableAbort || timeout) {
                            timer = setTimeout(function () {
                                var timeoutMsg = "\u8BF7\u6C42\u5728".concat(timeout / 1000, "s\u5185\u672A\u5B8C\u6210\uFF0C\u5DF2\u4E2D\u65AD");
                                console.warn(timeoutMsg);
                                abortController.abort(new Error(timeoutMsg));
                            }, timeout);
                        }
                        headers = options.headers ? (0, utils_1.headersInit2Indexable)(options.headers) : undefined;
                        return [4, this.getReqOptions(url, __assign(__assign({}, options), { headers: headers, body: options.body, signal: abortController.signal }))];
                    case 1:
                        fetchOptions = _f.sent();
                        return [4, fetch(url, fetchOptions)
                                .then(function (x) {
                                clearTimeout(timer);
                                return x;
                            })
                                .catch(function (x) {
                                clearTimeout(timer);
                                return Promise.reject(x);
                            })];
                    case 2:
                        res = _f.sent();
                        _e = {};
                        if (!(+res.status === 204)) return [3, 3];
                        _c = '';
                        return [3, 7];
                    case 3:
                        if (!stream) return [3, 4];
                        _d = res.body;
                        return [3, 6];
                    case 4: return [4, res.json()];
                    case 5:
                        _d = _f.sent();
                        _f.label = 6;
                    case 6:
                        _c = _d;
                        _f.label = 7;
                    case 7:
                        ret = (_e.data = _c,
                            _e.statusCode = res.status,
                            _e.header = res.headers,
                            _e);
                        return [2, ret];
                }
            });
        }); };
        this.request = function (options, enableAbort) {
            if (enableAbort === void 0) { enableAbort = false; }
            return __awaiter(_this, void 0, void 0, function () {
                var url, _a, _headers, data, responseType, withCredentials, body, _method, customReqOpts, headers, method, abortController, signal, timeout, payload, realUrl, timer, requestOptions, response, result_1, _b, _c, _d, e_2, _e, _f, headers_1;
                return __generator(this, function (_g) {
                    switch (_g.label) {
                        case 0:
                            url = options.url, _a = options.headers, _headers = _a === void 0 ? {} : _a, data = options.data, responseType = options.responseType, withCredentials = options.withCredentials, body = options.body, _method = options.method, customReqOpts = options.customReqOpts;
                            headers = (0, utils_1.obj2StrRecord)(_headers);
                            method = String(_method).toLowerCase() || 'get';
                            abortController = new AbortController();
                            signal = abortController.signal;
                            timeout = (customReqOpts === null || customReqOpts === void 0 ? void 0 : customReqOpts.timeout) || this.timeout;
                            if ((0, utils_1.isFormData)(data)) {
                                payload = data;
                            }
                            else if (headers['content-type'] === 'application/x-www-form-urlencoded') {
                                payload = (0, utils_1.toQueryString)(data !== null && data !== void 0 ? data : {});
                            }
                            else if (body) {
                                payload = body;
                            }
                            else {
                                payload = data ? JSON.stringify(data) : undefined;
                            }
                            realUrl = this.getRealUrl((0, adapter_interface_1.formatUrl)('https', url !== null && url !== void 0 ? url : '', method === 'get' ? data : {}), options.headers || {}, payload);
                            if (enableAbort || timeout) {
                                timer = setTimeout(function () {
                                    var timeoutMsg = "\u8BF7\u6C42\u5728".concat(timeout / 1000, "s\u5185\u672A\u5B8C\u6210\uFF0C\u5DF2\u4E2D\u65AD");
                                    console.warn(timeoutMsg);
                                    abortController.abort(new Error(timeoutMsg));
                                }, timeout);
                            }
                            return [4, this.getReqOptions(realUrl, {
                                    method: method,
                                    headers: headers,
                                    body: payload,
                                    credentials: withCredentials ? 'include' : 'same-origin',
                                    signal: signal,
                                })];
                        case 1:
                            requestOptions = _g.sent();
                            _g.label = 2;
                        case 2:
                            _g.trys.push([2, , 15, 16]);
                            return [4, fetch(realUrl, requestOptions)];
                        case 3:
                            response = _g.sent();
                            result_1 = {
                                header: {},
                                statusCode: response.status,
                            };
                            _g.label = 4;
                        case 4:
                            _g.trys.push([4, 9, , 14]);
                            _b = result_1;
                            if (!(responseType === 'blob')) return [3, 6];
                            return [4, response.blob()];
                        case 5:
                            _c = _g.sent();
                            return [3, 8];
                        case 6:
                            _d = utils_1.safeParseJson;
                            return [4, response.text()];
                        case 7:
                            _c = _d.apply(void 0, [_g.sent()]);
                            _g.label = 8;
                        case 8:
                            _b.data = _c;
                            return [3, 14];
                        case 9:
                            e_2 = _g.sent();
                            console.log('catch an error', e_2);
                            _e = result_1;
                            if (!(responseType === 'blob')) return [3, 11];
                            return [4, response.blob()];
                        case 10:
                            _f = _g.sent();
                            return [3, 13];
                        case 11: return [4, response.text()];
                        case 12:
                            _f = _g.sent();
                            _g.label = 13;
                        case 13:
                            _e.data = _f;
                            return [3, 14];
                        case 14:
                            headers_1 = response.headers;
                            headers_1.forEach(function (val, key) { return (result_1.header[key.toLowerCase()] = val); });
                            return [2, result_1];
                        case 15:
                            clearTimeout(timer);
                            return [7];
                        case 16: return [2];
                    }
                });
            });
        };
        var timeout = config.timeout, restrictedMethods = config.restrictedMethods;
        this.timeout = timeout || 15000;
        this.restrictedMethods = restrictedMethods || ['get', 'post', 'upload', 'download'];
        this.config = config;
    }
    NodeRequest.prototype.getClientCredential = function (opts) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4, this.fetch({
                            url: "".concat(opts.origin).concat(constants_1.CLIENT_AUTH_PATH),
                            method: 'POST',
                            headers: {
                                'content-type': 'application/json',
                            },
                            body: JSON.stringify({
                                grant_type: 'client_credentials',
                            }),
                        })];
                    case 1:
                        res = _b.sent();
                        return [2, (_a = res.data) === null || _a === void 0 ? void 0 : _a.access_token];
                }
            });
        });
    };
    NodeRequest.prototype.getRealUrl = function (url, headers, body) {
        var _a, _b, _c, _d;
        if ((_b = (_a = headers.Authorization) === null || _a === void 0 ? void 0 : _a.replace('Bearer', '')) === null || _b === void 0 ? void 0 : _b.trim()) {
            return url;
        }
        var params = typeof body === 'object' ? body : (0, utils_1.safeParseJson)(body) || '';
        if (params.access_token) {
            return url;
        }
        var urlObj = new URL(url);
        if (urlObj.pathname === '/web' && ((_c = this.config.auth) === null || _c === void 0 ? void 0 : _c.secretId) && ((_d = this.config.auth) === null || _d === void 0 ? void 0 : _d.secretKey)) {
            urlObj.pathname = constants_1.ADMIN_PATH;
        }
        return urlObj.toString();
    };
    return NodeRequest;
}());
exports.NodeRequest = NodeRequest;
