// @cloudbase/app/dist/cjs/libs/adapter-node/tool.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeTool = void 0;
var constants_1 = require("./constants");
var context_1 = require("./context");
var utils_1 = require("./utils");
var jwtSign = null;
function getJwtSign() {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var mod, e_1;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (!!jwtSign) return [3, 4];
                    _b.label = 1;
                case 1:
                    _b.trys.push([1, 3, , 4]);
                    return [4, import('jsonwebtoken')];
                case 2:
                    mod = _b.sent();
                    jwtSign = ((_a = mod.default) === null || _a === void 0 ? void 0 : _a.sign) || mod.sign;
                    return [3, 4];
                case 3:
                    e_1 = _b.sent();
                    console.error('缺少依赖 jsonwebtoken，请执行以下命令安装：\n\n'
                        + '  npm install jsonwebtoken\n\n'
                        + '该依赖用于 Node 环境下的自定义登录票据生成。');
                    return [3, 4];
                case 4: return [2, jwtSign];
            }
        });
    });
}
function getDefaultUserInfo() {
    var _a = (0, context_1.getCloudbaseContext)(), WX_OPENID = _a.WX_OPENID, WX_APPID = _a.WX_APPID, TCB_UUID = _a.TCB_UUID, TCB_CUSTOM_USER_ID = _a.TCB_CUSTOM_USER_ID, TCB_ISANONYMOUS_USER = _a.TCB_ISANONYMOUS_USER;
    return {
        openId: WX_OPENID || '',
        appId: WX_APPID || '',
        uid: TCB_UUID || '',
        customUserId: TCB_CUSTOM_USER_ID || '',
        isAnonymous: TCB_ISANONYMOUS_USER === 'true',
    };
}
function sendUserInfoRequest(app, params) {
    var _a, _b;
    return (_b = (_a = app === null || app === void 0 ? void 0 : app.request) === null || _a === void 0 ? void 0 : _a.send) === null || _b === void 0 ? void 0 : _b.call(_a, 'auth.getUserInfoForAdmin', params, {
        pathname: 'web',
        endPointMode: 'CLOUD_API',
    });
}
var nodeTool = function (app, config) {
    if (app.auth) {
        var auth_1 = {
            getUserInfo: function () {
                return getDefaultUserInfo();
            },
            getEndUserInfo: function (uid) {
                return __awaiter(this, void 0, void 0, function () {
                    var defaultUserInfo;
                    return __generator(this, function (_a) {
                        defaultUserInfo = getDefaultUserInfo();
                        if (uid === undefined) {
                            return [2, { userInfo: defaultUserInfo }];
                        }
                        (0, utils_1.validateUid)(uid);
                        return [2, sendUserInfoRequest(app, {
                                uuid: uid,
                                envName: config.env,
                            }).then(function (result) {
                                if (result.code) {
                                    return result;
                                }
                                return {
                                    userInfo: __assign(__assign({}, defaultUserInfo), result.data),
                                    requestId: result.requestId,
                                };
                            })];
                    });
                });
            },
            createTicket: function (uid, options) {
                if (options === void 0) { options = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var timestamp, credentials, env, _a, refresh, _b, expire, sign, token;
                    return __generator(this, function (_c) {
                        switch (_c.label) {
                            case 0:
                                (0, utils_1.validateUid)(uid);
                                timestamp = new Date().getTime();
                                credentials = (config.auth || {}).credentials;
                                env = config.env;
                                if (!env) {
                                    throw __assign(__assign({}, constants_1.ERROR.INVALID_PARAM), { message: 'no env in config' });
                                }
                                if (!(credentials === null || credentials === void 0 ? void 0 : credentials.env_id)) {
                                    throw __assign(__assign({}, constants_1.ERROR.INVALID_PARAM), { message: '当前私钥未包含env_id 信息， 请前往腾讯云云开发控制台，获取自定义登录最新私钥' });
                                }
                                if (credentials.env_id !== env) {
                                    throw __assign(__assign({}, constants_1.ERROR.INVALID_PARAM), { message: '当前私钥所属环境与 init 指定环境不一致！' });
                                }
                                _a = options.refresh, refresh = _a === void 0 ? 3600 * 1000 : _a, _b = options.expire, expire = _b === void 0 ? timestamp + 7 * 24 * 60 * 60 * 1000 : _b;
                                return [4, getJwtSign()];
                            case 1:
                                sign = _c.sent();
                                token = sign({
                                    alg: 'RS256',
                                    env: env,
                                    iat: timestamp,
                                    exp: timestamp + 10 * 60 * 1000,
                                    uid: uid,
                                    refresh: refresh,
                                    expire: expire,
                                }, credentials.private_key, {
                                    allowInsecureKeySizes: true,
                                    algorithm: 'RS256',
                                });
                                return [2, "".concat(credentials.private_key_id, "/@@/").concat(token)];
                        }
                    });
                });
            },
            queryUserInfo: function (query) {
                return __awaiter(this, void 0, void 0, function () {
                    var uid, platform, platformId;
                    return __generator(this, function (_a) {
                        uid = query.uid, platform = query.platform, platformId = query.platformId;
                        return [2, sendUserInfoRequest(app, {
                                uuid: uid,
                                platform: platform,
                                platformId: platformId,
                                envName: config.env,
                            }).then(function (result) {
                                if (result.code) {
                                    return result;
                                }
                                return {
                                    userInfo: __assign({}, result.data),
                                    requestId: result.requestId,
                                };
                            })];
                    });
                });
            },
            getClientIP: function () {
                var TCB_SOURCE_IP = (0, context_1.getCloudbaseContext)().TCB_SOURCE_IP;
                return TCB_SOURCE_IP || '';
            },
        };
        Object.keys(auth_1).forEach(function (key) {
            app.auth[key] = auth_1[key];
        });
    }
    app.sendTemplateNotification = function (params, opts) { return __awaiter(void 0, void 0, void 0, function () {
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4, ((_a = app === null || app === void 0 ? void 0 : app.callFunction) === null || _a === void 0 ? void 0 : _a.call(app, {
                        name: 'lowcode-datasource',
                        data: {
                            methodName: 'callWedaApi',
                            params: {
                                action: 'PushNotifyMsg',
                                data: {
                                    NotifyId: params.notifyId,
                                    Data: JSON.stringify(params.data),
                                    NotifyUsers: undefined,
                                    Url: params.url,
                                },
                            },
                            mode: 'c',
                        },
                    }, undefined, opts))];
                case 1: return [2, _b.sent()];
            }
        });
    }); };
    app.parseContext = context_1.parseContext;
};
exports.nodeTool = nodeTool;
