// @cloudbase/app/dist/cjs/libs/adapter-node/utils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWebStreamFromNodeReadableStream = exports.parseQueryString = exports.headersInit2Indexable = exports.obj2StrRecord = exports.safeParseJson = exports.validateUid = exports.getCurrRunEnvTag = exports.checkIsInTencentCloud = exports.isNonEmptyString = exports.checkIsInSumeru = exports.checkIsInCBR = exports.checkIsInScf = exports.getEnv = exports.toQueryString = exports.isFormData = void 0;
var utilities_1 = require("@cloudbase/utilities");
var web_streams_polyfill_1 = require("web-streams-polyfill");
var constants_1 = require("./constants");
var metadata_1 = require("./metadata");
exports.isFormData = (_a = utilities_1.utils, _a.isFormData), exports.toQueryString = _a.toQueryString;
function getEnv(key) {
    if (typeof process === 'undefined' || !process.env) {
        return key !== undefined ? undefined : {};
    }
    return key !== undefined ? process.env[key] : process.env;
}
exports.getEnv = getEnv;
var checkIsInScf = function () { return getEnv('TENCENTCLOUD_RUNENV') === 'SCF'; };
exports.checkIsInScf = checkIsInScf;
function checkIsInCBR() {
    return !!getEnv('CBR_ENV_ID');
}
exports.checkIsInCBR = checkIsInCBR;
var kSumeruEnvSet = new Set(['formal', 'pre', 'test']);
function checkIsInSumeru() {
    return kSumeruEnvSet.has(getEnv('SUMERU_ENV') || '');
}
exports.checkIsInSumeru = checkIsInSumeru;
function isNonEmptyString(str) {
    return typeof str === 'string' && str !== '';
}
exports.isNonEmptyString = isNonEmptyString;
function checkIsInTencentCloud() {
    return __awaiter(this, void 0, void 0, function () {
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (getEnv('TENCENTCLOUD') === 'true') {
                        return [2, true];
                    }
                    _a = isNonEmptyString;
                    return [4, (0, metadata_1.lookupAppId)()];
                case 1: return [2, _a.apply(void 0, [_b.sent()])];
            }
        });
    });
}
exports.checkIsInTencentCloud = checkIsInTencentCloud;
var getCurrRunEnvTag = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                if ((0, exports.checkIsInScf)()) {
                    return [2, 'scf'];
                }
                if (checkIsInCBR()) {
                    return [2, 'cbr'];
                }
                if (checkIsInSumeru()) {
                    return [2, 'sumeru'];
                }
                return [4, checkIsInTencentCloud()];
            case 1:
                if (_a.sent()) {
                    return [2, 'tencentcloud'];
                }
                return [2, 'unknown'];
        }
    });
}); };
exports.getCurrRunEnvTag = getCurrRunEnvTag;
var validateUid = function (uid) {
    if (typeof uid !== 'string') {
        throw __assign(__assign({}, constants_1.ERROR.INVALID_PARAM), { message: 'uid must be a string' });
    }
    if (!constants_1.checkCustomUserIdRegex.test(uid)) {
        throw __assign(__assign({}, constants_1.ERROR.INVALID_PARAM), { message: "Invalid uid: \"".concat(uid, "\"") });
    }
};
exports.validateUid = validateUid;
function safeParseJson(text) {
    if (!text || !text.trim()) {
        return null;
    }
    try {
        return JSON.parse(text);
    }
    catch (e) {
        console.warn('catch an error', { e: e, text: text });
        return text;
    }
}
exports.safeParseJson = safeParseJson;
function obj2StrRecord(obj) {
    return Object.entries(obj).reduce(function (acc, cur) {
        var key = cur[0], value = cur[1];
        acc[key] = String(value);
        return acc;
    }, {});
}
exports.obj2StrRecord = obj2StrRecord;
function headersInit2Indexable(h) {
    if (isHeaders(h)) {
        var ret_1 = {};
        h.forEach(function (val, key) {
            ret_1[key] = val;
        });
        return ret_1;
    }
    return h;
    function isHeaders(h) {
        try {
            return h instanceof Headers;
        }
        catch (_) {
            return false;
        }
    }
}
exports.headersInit2Indexable = headersInit2Indexable;
function parseQueryString(search) {
    var params = {};
    var cleanSearch = search.replace(/^[?#]/, '');
    if (!cleanSearch) {
        return params;
    }
    var pairs = cleanSearch.split('&');
    pairs.forEach(function (item) {
        var _a = item.split('='), key = _a[0], value = _a[1];
        key = decodeURIComponent(key);
        value = decodeURIComponent(value || '');
        if (key) {
            if (params[key]) {
                if (Array.isArray(params[key])) {
                    params[key].push(value);
                }
                else {
                    params[key] = [params[key], value];
                }
            }
            else {
                params[key] = value;
            }
        }
    });
    return params;
}
exports.parseQueryString = parseQueryString;
function createWebStreamFromNodeReadableStream(stream) {
    var asyncIterator = stream[Symbol.asyncIterator]();
    return new web_streams_polyfill_1.ReadableStream({
        pull: function (controller) {
            return __awaiter(this, void 0, void 0, function () {
                var next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4, asyncIterator.next()];
                        case 1:
                            next = _a.sent();
                            if (next.done) {
                                controller.close();
                            }
                            else {
                                controller.enqueue(next.value);
                            }
                            return [2];
                    }
                });
            });
        },
    });
}
exports.createWebStreamFromNodeReadableStream = createWebStreamFromNodeReadableStream;
