// @cloudbase/oauth/dist/esm/index.js
import { OAuth2Client } from './oauth2client/oauth2client';
import { Auth } from './auth/apis';
import { AUTH_API_PREFIX } from './auth/consts';
export { Auth } from './auth/apis';
export { AuthError, AuthErrorCategory } from './auth/auth-error';
import * as authModels_1 from './auth/models';
export { authModels_1 as authModels };
export { weAppJwtDecodeAll } from './utils/base64';
export { AUTH_API_PREFIX, OAUTH_TYPE, DEFAULT_NODE_ACCESS_SCOPE, LOGIN_STATE_CHANGED_TYPE, EVENTS, AUTH_STATE_CHANGED_TYPE } from './auth/consts';
export class CloudbaseOAuth {
    constructor(authOptions) {
        const { apiOrigin, apiPath = AUTH_API_PREFIX, clientId, env, storage, request, baseRequest, anonymousSignInFunc, wxCloud, adapter, onCredentialsError, headers, i18n, useWxCloud, eventBus, detectSessionInUrl, debug, } = authOptions;
        this.detectSessionInUrl = detectSessionInUrl ?? false;
        this.oauth2client = new OAuth2Client({
            apiOrigin,
            apiPath,
            clientId,
            env,
            storage,
            baseRequest: baseRequest || request,
            anonymousSignInFunc,
            wxCloud,
            onCredentialsError,
            headers: headers || {},
            i18n,
            useWxCloud,
            eventBus,
            debug,
        });
        this.authApi = new Auth({
            credentialsClient: this.oauth2client,
            ...authOptions,
            request: request ? this.oauth2client.request.bind(this.oauth2client) : undefined,
            adapter,
        });
        if (detectSessionInUrl) {
            this.oauth2client.setGetInitialSession(this.authApi.getInitialSession.bind(this.authApi));
        }
        else {
            try {
                const url = new URL(window.location.href);
                const code = url.searchParams.get('code');
                const state = url.searchParams.get('state');
                const cacheData = JSON.parse(sessionStorage.getItem(state) || 'null');
                if (code && state && cacheData?.hash) {
                    url.hash = cacheData.hash;
                    window.history.replaceState(null, '', url.toString());
                    window.location.replace(url.toString());
                }
            }
            catch (error) {
            }
        }
    }
    initializeSession(onUserObtained) {
        if (!this.detectSessionInUrl) {
            this.oauth2client.initialize(Promise.resolve({ error: null }));
            return;
        }
        this.oauth2client.setOnInitialSessionObtained(onUserObtained);
        this.oauth2client.initialize();
    }
}
