// @cloudbase/storage/dist/esm/supabase/errors.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var StorageError = (function (_super) {
    __extends(StorageError, _super);
    function StorageError(message) {
        var _this = _super.call(this, message) || this;
        _this.__isStorageError = true;
        _this.name = 'StorageError';
        return _this;
    }
    return StorageError;
}(Error));
export { StorageError };
export function isStorageError(error) {
    return typeof error === 'object' && error !== null && '__isStorageError' in error;
}
var StorageApiError = (function (_super) {
    __extends(StorageApiError, _super);
    function StorageApiError(message, status, statusCode) {
        var _this = _super.call(this, message) || this;
        _this.name = 'StorageApiError';
        _this.status = status;
        _this.statusCode = statusCode;
        return _this;
    }
    StorageApiError.prototype.toJSON = function () {
        return {
            name: this.name,
            message: this.message,
            status: this.status,
            statusCode: this.statusCode,
        };
    };
    return StorageApiError;
}(StorageError));
export { StorageApiError };
var StorageUnknownError = (function (_super) {
    __extends(StorageUnknownError, _super);
    function StorageUnknownError(message, originalError) {
        var _this = _super.call(this, message) || this;
        _this.name = 'StorageUnknownError';
        _this.originalError = originalError;
        return _this;
    }
    return StorageUnknownError;
}(StorageError));
export { StorageUnknownError };
