// @cloudbase/cloudrun/dist/cjs/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerCloudrun = exports.requestContainer = void 0;
var utilities_1 = require("@cloudbase/utilities");
var ERRORS = utilities_1.constants.ERRORS;
var COMPONENT_NAME = 'cloudrun';
var CloudbaseRun = (function () {
    function CloudbaseRun() {
    }
    CloudbaseRun.prototype.callContainer = function (options, customReqOpts) {
        return __awaiter(this, void 0, void 0, function () {
            var name, data, jsonData;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        name = options.name, data = options.data;
                        if (!name) {
                            throw new Error(JSON.stringify({
                                code: ERRORS.INVALID_PARAMS,
                                msg: "[".concat(COMPONENT_NAME, ".callContainer] invalid name"),
                            }));
                        }
                        try {
                            jsonData = data ? JSON.stringify(data) : '';
                        }
                        catch (e) {
                            throw new Error(JSON.stringify({
                                code: ERRORS.INVALID_PARAMS,
                                msg: "[".concat(COMPONENT_NAME, ".callContainer] invalid data"),
                            }));
                        }
                        return [4, requestContainer.call(this, __assign(__assign({}, options), { data: jsonData }), customReqOpts)];
                    case 1: return [2, _a.sent()];
                }
            });
        });
    };
    return CloudbaseRun;
}());
function requestContainer(options, customReqOpts) {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var name, data, _b, path, method, _c, header, _d, BASE_URL, PROTOCOL, endpoint, reqPath, isGetAndHead, dataParse_1, response;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    name = options.name, data = options.data, _b = options.path, path = _b === void 0 ? '' : _b, method = options.method, _c = options.header, header = _c === void 0 ? {} : _c;
                    _d = this.getEndPointWithKey('GATEWAY'), BASE_URL = _d.BASE_URL, PROTOCOL = _d.PROTOCOL;
                    endpoint = "".concat(PROTOCOL).concat(BASE_URL, "/cloudrun/").concat(name);
                    reqPath = path.startsWith('/') ? path : "/".concat(path);
                    isGetAndHead = ['GET', 'HEAD'].includes((_a = method === null || method === void 0 ? void 0 : method.toUpperCase) === null || _a === void 0 ? void 0 : _a.call(method));
                    if (isGetAndHead) {
                        try {
                            dataParse_1 = {};
                            try {
                                dataParse_1 = JSON.parse(data);
                            }
                            catch (error) {
                                dataParse_1 = data || {};
                            }
                            reqPath = "".concat(reqPath).concat(reqPath.includes('?') ? '&' : '?').concat(Object.keys(dataParse_1)
                                .map(function (key) { return "".concat(key, "=").concat(dataParse_1[key]); })
                                .join('&'));
                        }
                        catch (error) {
                        }
                    }
                    return [4, this.request.fetch(__assign(__assign({ method: method || 'POST', headers: Object.assign({}, {
                                'Content-Type': 'application/json; charset=utf-8',
                            }, header) }, (isGetAndHead ? {} : { body: data })), { url: "".concat(endpoint).concat(reqPath), customReqOpts: customReqOpts }))];
                case 1:
                    response = _e.sent();
                    if ((customReqOpts === null || customReqOpts === void 0 ? void 0 : customReqOpts.from) === 'node-sdk') {
                        return [2, response];
                    }
                    return [4, response.data];
                case 2: return [2, _e.sent()];
            }
        });
    });
}
exports.requestContainer = requestContainer;
var cloudbaseRun = new CloudbaseRun();
var component = {
    name: COMPONENT_NAME,
    entity: {
        callContainer: cloudbaseRun.callContainer,
    },
};
try {
    cloudbase.registerComponent(component);
}
catch (e) { }
function registerCloudrun(app) {
    try {
        app.registerComponent(component);
    }
    catch (e) {
        console.warn(e);
    }
}
exports.registerCloudrun = registerCloudrun;
try {
    window.registerCloudrun = registerCloudrun;
}
catch (e) { }
