// @cloudbase/js-sdk/database/dist/index.esm.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { Db } from '@cloudbase/database';
import { EJSON } from 'bson';
var COMPONENT_NAME = 'database';
var gateWayFunc = {
    filterRes: function (res) {
        delete res.header;
        delete res.statusCode;
        return res;
    },
    flattenResData: function (res) {
        if (typeof res.data === 'object' && res.data !== null) {
            var result = __assign(__assign({}, res), res.data);
            delete result.data;
            return result;
        }
        return res;
    },
    safeParseEJSON: function (data) {
        if (typeof data !== 'string')
            return data;
        try {
            return EJSON.parse(data);
        }
        catch (_a) {
            return data;
        }
    },
    buildQueryParams: function (params) {
        return Object.entries(params)
            .filter(function (_a) {
            var _ = _a[0], value = _a[1];
            return value !== undefined && value !== null && value !== '';
        })
            .map(function (_a) {
            var key = _a[0], value = _a[1];
            var stringValue = typeof value === 'object' ? JSON.stringify(value) : String(value);
            return "".concat(encodeURIComponent(key), "=").concat(encodeURIComponent(stringValue));
        })
            .join('&');
    },
    getUrlPrefix: function (data) {
        var instance = (data === null || data === void 0 ? void 0 : data.instance) || '(default)';
        var database = (data === null || data === void 0 ? void 0 : data.database) || '(default)';
        return "/database/instances/".concat(instance, "/databases/").concat(database);
    },
    'database.queryDocument': function (data) {
        var _a, _b, _c, _d, _e;
        return __awaiter(this, void 0, void 0, function () {
            var isQueryDoc, queryParams, url, res;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        isQueryDoc = !!((_a = data.query) === null || _a === void 0 ? void 0 : _a._id);
                        queryParams = gateWayFunc.buildQueryParams(__assign({ projection: data.projection, transactionId: data.transactionId }, (isQueryDoc
                            ? {}
                            : {
                                offset: (_b = data.offset) !== null && _b !== void 0 ? _b : 0,
                                limit: (_c = data.limit) !== null && _c !== void 0 ? _c : 100,
                                order: data.order,
                                query: data.query ? EJSON.stringify(data.query, { relaxed: false }) : undefined,
                            })));
                        url = "".concat(gateWayFunc.getUrlPrefix(data), "/collections/").concat(data.collectionName, "/documents?").concat(queryParams);
                        if (isQueryDoc) {
                            url = "".concat(gateWayFunc.getUrlPrefix(data), "/collections/").concat(data.collectionName, "/documents/").concat(((_d = data.query._id) === null || _d === void 0 ? void 0 : _d.$eq) || data.query._id, "?").concat(queryParams);
                        }
                        return [4, this.gateWay({
                                url: url,
                                method: 'GET',
                            })];
                    case 1:
                        res = _f.sent();
                        res.data = gateWayFunc.safeParseEJSON(JSON.stringify(res.data));
                        if (isQueryDoc && !res.code && !((_e = res.data) === null || _e === void 0 ? void 0 : _e.code)) {
                            return [2, __assign(__assign({}, res), { data: { list: [res.data] } })];
                        }
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.addDocument': function (data) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4, this.gateWay({
                            url: "".concat(gateWayFunc.getUrlPrefix(data), "/collections/").concat(data.collectionName, "/documents"),
                            method: 'POST',
                            data: __assign({ data: Array.isArray(data.data) ? data.data : [data.data] }, (data.transactionId && { transactionId: data.transactionId })),
                        })];
                    case 1:
                        res = _d.sent();
                        if (Array.isArray(data.data)) {
                            res = __assign(__assign({}, res), { data: __assign({ ids: ((_a = res.data) === null || _a === void 0 ? void 0 : _a.insertedIds) || [] }, res.data) });
                        }
                        else if ((_c = (_b = res.data) === null || _b === void 0 ? void 0 : _b.insertedIds) === null || _c === void 0 ? void 0 : _c[0]) {
                            res = __assign(__assign({}, res), { data: __assign({ _id: res.data.insertedIds[0] }, res.data) });
                        }
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.updateDocument': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.gateWay({
                            url: "".concat(gateWayFunc.getUrlPrefix(data), "/collections/").concat(data.collectionName, "/documents"),
                            method: 'PATCH',
                            data: __assign({ query: data.query, data: data.data, multi: !!data.multi, upsert: !!data.upsert, replaceMode: !!data.replaceMode }, (data.transactionId && { transactionId: data.transactionId })),
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.deleteDocument': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.gateWay({
                            url: "".concat(gateWayFunc.getUrlPrefix(data), "/collections/").concat(data.collectionName, "/documents/remove"),
                            method: 'POST',
                            data: __assign({ query: data.query, multi: !!data.multi }, (data.transactionId && { transactionId: data.transactionId })),
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.countDocument': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var queryParams, res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        queryParams = gateWayFunc.buildQueryParams({
                            count: true,
                            query: data.query ? EJSON.stringify(data.query, { relaxed: false }) : undefined,
                        });
                        return [4, this.gateWay({
                                url: "".concat(gateWayFunc.getUrlPrefix(data), "/collections/").concat(data.collectionName, "/documents?").concat(queryParams),
                                method: 'GET',
                            })];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.addCollection': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.gateWay({
                            url: "".concat(gateWayFunc.getUrlPrefix(data), "/collections"),
                            method: 'POST',
                            data: { collectionName: data.collectionName },
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.aggregate': function (data) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var pipeline, res;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        pipeline = data.stages.map(function (stage) {
                            var _a, _b;
                            try {
                                var value = typeof stage.stageValue === 'string' ? JSON.parse(stage.stageValue) : stage.stageValue;
                                return _a = {},
                                    _a[stage.stageKey] = EJSON.serialize(value),
                                    _a;
                            }
                            catch (_c) {
                                return _b = {}, _b[stage.stageKey] = stage.stageValue, _b;
                            }
                        });
                        return [4, this.gateWay({
                                url: "".concat(gateWayFunc.getUrlPrefix(data), "/collections/").concat(data.collectionName, "/documents/aggregations"),
                                method: 'POST',
                                data: { pipeline: pipeline },
                            })];
                    case 1:
                        res = _b.sent();
                        if (Array.isArray((_a = res.data) === null || _a === void 0 ? void 0 : _a.list)) {
                            res.data.list = JSON.stringify(res.data.list.map(function (item) { return JSON.stringify(item); }));
                        }
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.startTransaction': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.gateWay({
                            url: "".concat(gateWayFunc.getUrlPrefix(data), "/transactions"),
                            method: 'POST',
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.filterRes(gateWayFunc.flattenResData(res))];
                }
            });
        });
    },
    'database.commitTransaction': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.gateWay({
                            url: "".concat(gateWayFunc.getUrlPrefix(data), "/transactions/").concat(data.transactionId, "/commit"),
                            method: 'POST',
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.abortTransaction': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.gateWay({
                            url: "".concat(gateWayFunc.getUrlPrefix(data), "/transactions/").concat(data.transactionId, "/rollback"),
                            method: 'POST',
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
    'database.getInTransaction': function (data) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4, gateWayFunc['database.queryDocument'].call(this, data)];
                    case 1:
                        res = _b.sent();
                        if (typeof res.data === 'object' && res.data !== null) {
                            if ((_a = res.data) === null || _a === void 0 ? void 0 : _a.code) {
                                return [2, gateWayFunc.flattenResData(res)];
                            }
                            return [2, __assign(__assign({}, res), { data: JSON.stringify(res.data) })];
                        }
                        return [2, res];
                }
            });
        });
    },
    'database.updateDocInTransaction': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        data.data = gateWayFunc.safeParseEJSON(data.data);
                        return [4, gateWayFunc['database.updateDocument'].call(this, data)];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.flattenResData(res)];
                }
            });
        });
    },
    'database.deleteDocInTransaction': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, gateWayFunc['database.deleteDocument'].call(this, data)];
                    case 1:
                        res = _a.sent();
                        return [2, gateWayFunc.flattenResData(res)];
                }
            });
        });
    },
    'database.insertDocInTransaction': function (data) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var res, hasId;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        data.data = gateWayFunc.safeParseEJSON(data.data);
                        return [4, gateWayFunc['database.addDocument'].call(this, data)];
                    case 1:
                        res = _b.sent();
                        if (typeof res.data === 'object' && res.data !== null) {
                            hasId = !!((_a = res.data) === null || _a === void 0 ? void 0 : _a._id);
                            res = __assign(__assign(__assign({}, res), { inserted: hasId ? 1 : 0, ok: hasId ? 1 : 0 }), res.data);
                        }
                        else {
                            res = __assign({ inserted: 0, ok: 0 }, res);
                        }
                        delete res.data;
                        return [2, res];
                }
            });
        });
    },
    'database.runCommands': function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.gateWay({
                            url: "".concat(gateWayFunc.getUrlPrefix(data), "/commands"),
                            method: 'POST',
                            data: __assign({ commands: EJSON.serialize(data.commands) }, (data.transactionId && { transactionId: data.transactionId })),
                        })];
                    case 1:
                        res = _a.sent();
                        if (typeof res.data === 'object' && res.data !== null) {
                            res = __assign(__assign({}, res), res.data);
                            delete res.data;
                            delete res.request_id;
                        }
                        try {
                            res = EJSON.deserialize(res);
                        }
                        catch (_b) {
                        }
                        return [2, gateWayFunc.filterRes(res)];
                }
            });
        });
    },
};
function database(dbConfig) {
    var _a = this.platform, adapter = _a.adapter, runtime = _a.runtime;
    var Req = (function (_super) {
        __extends(Req, _super);
        function Req(config) {
            return _super.call(this, config) || this;
        }
        Req.prototype.send = function (action, data) {
            return __awaiter(this, void 0, void 0, function () {
                var res;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            res = null;
                            if (!(this.config.endPointMode === 'GATEWAY' && gateWayFunc[action])) return [3, 2];
                            return [4, gateWayFunc[action].call(this, data)];
                        case 1:
                            res = _a.sent();
                            return [3, 4];
                        case 2: return [4, _super.prototype.send.call(this, action, data)];
                        case 3:
                            res = _a.sent();
                            _a.label = 4;
                        case 4: return [2, res];
                    }
                });
            });
        };
        return Req;
    }(this.request.constructor));
    Db.reqClass = Req;
    Db.getAccessToken = this.authInstance ? this.authInstance.getAccessToken.bind(this.authInstance) : function () { return ''; };
    Db.runtime = runtime;
    if (this.wsClientClass) {
        Db.wsClass = adapter.wsClass;
        Db.wsClientClass = this.wsClientClass;
    }
    if (!Db.ws) {
        Db.ws = null;
    }
    var NewDb = (function (_super) {
        __extends(NewDb, _super);
        function NewDb(config) {
            return _super.call(this, config) || this;
        }
        NewDb.prototype.runCommands = function (params) {
            return __awaiter(this, void 0, void 0, function () {
                var request;
                return __generator(this, function (_a) {
                    request = Db.createRequest(this.config);
                    return [2, request.send('database.runCommands', params)];
                });
            });
        };
        return NewDb;
    }(Db));
    return new NewDb(__assign(__assign(__assign({}, this.config), { _fromApp: this }), dbConfig));
}
var component = {
    name: COMPONENT_NAME,
    entity: {
        database: database,
    },
};
try {
    cloudbase.registerComponent(component);
}
catch (e) { }
export function registerDatabase(app) {
    try {
        app.registerComponent(component);
    }
    catch (e) {
        console.warn(e);
    }
}
try {
    ;
    window.registerDatabase = registerDatabase;
}
catch (e) { }
